
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>CBMPy Module Reference &#8212; CBMPy 0.7.20 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.20',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="search" type="application/opensearchdescription+xml"
          title="Search within CBMPy 0.7.20 documentation"
          href="_static/opensearch.xml"/>
    <link rel="shortcut icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Introduction" href="manual_doc.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-cbmpy.CBCommon">
<span id="cbmpy-module-reference"></span><h1>CBMPy Module Reference<a class="headerlink" href="#module-cbmpy.CBCommon" title="Permalink to this headline">¶</a></h1>
<div class="section" id="cbmpy-cbcommon-module">
<h2>CBMPy: CBCommon module<a class="headerlink" href="#cbmpy-cbcommon-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBCommon.py 644 2018-03-14 21:40:47Z bgoli $)</p>
<dl class="class">
<dt id="cbmpy.CBCommon.ComboGen">
<em class="property">class </em><code class="descclassname">cbmpy.CBCommon.</code><code class="descname">ComboGen</code><a class="reference internal" href="_modules/cbmpy/CBCommon.html#ComboGen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.ComboGen" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate sets of unique combinations</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.checkChemFormula">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">checkChemFormula</code><span class="sig-paren">(</span><em>cf</em>, <em>quiet=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#checkChemFormula"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.checkChemFormula" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether a string conforms to a Chemical Formula C3Br5 etc, returns True/False. Please see the SBML
Level 3 specification and <a class="reference external" href="http://wikipedia.org/wiki/Hill_system">http://wikipedia.org/wiki/Hill_system</a> for more information.</p>
<blockquote>
<div><ul class="simple">
<li><em>cf</em> a string that contains a formula to check</li>
<li><em>quiet</em> [default=False] do not print error messages</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.checkId">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">checkId</code><span class="sig-paren">(</span><em>s</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#checkId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.checkId" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the validity of the string to see if it conforms to a C variable. Returns true/false</p>
<ul class="simple">
<li><em>s</em> a string</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.createAssociationDictFromNode">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">createAssociationDictFromNode</code><span class="sig-paren">(</span><em>node</em>, <em>out</em>, <em>model</em>, <em>useweakref=True</em>, <em>cntr=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#createAssociationDictFromNode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.createAssociationDictFromNode" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a GPR string ‘((g1 and g2) or g3)’ to a dictionary via a Python AST.
In future I will get rid of all the string elements and work only with AST’s.</p>
<blockquote>
<div><ul class="simple">
<li><em>node</em> a Python AST node (e.g. body)</li>
<li><em>out</em> a gpr dictionary</li>
<li><em>model</em> a CBMPy model instance</li>
<li><em>useweakref</em> [default=True] use a weakref as the gene object or alternatively the label</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.extractGeneIdsFromString">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">extractGeneIdsFromString</code><span class="sig-paren">(</span><em>g</em>, <em>return_clean_gpr=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#extractGeneIdsFromString"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.extractGeneIdsFromString" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract and return a list of gene names from a gene association string formulation</p>
<ul class="simple">
<li><em>g</em> a COBRA style gene association string</li>
<li><em>return_clean_gpr</em> [default=False] in addition to the list returns the “cleaned” GPR string</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.fixId">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">fixId</code><span class="sig-paren">(</span><em>s</em>, <em>replace=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#fixId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.fixId" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks a string (Sid) to see if it is a valid C style variable. first letter must be an underscore or letter,
the rest should be alphanumeric or underscore.</p>
<ul class="simple">
<li><em>s</em> the string to test</li>
<li><em>replace</em> [None] default is to leave out offensive character, otherwise replace with this one</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.func_getAssociationStrFromGprDict">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">func_getAssociationStrFromGprDict</code><span class="sig-paren">(</span><em>gprd</em>, <em>out</em>, <em>parent=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#func_getAssociationStrFromGprDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.func_getAssociationStrFromGprDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a old school GPR association string from a CBMPy gprDict, e.g. obtained from gpr.getTree()</p>
<blockquote>
<div><ul class="simple">
<li><em>gprd</em> the gprDictionary</li>
<li><em>out</em> the output string</li>
<li><em>parent</em> [default=’‘] the string representing the current nodes parent relationship, used for recursion</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.getGPRasDictFromString">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">getGPRasDictFromString</code><span class="sig-paren">(</span><em>node</em>, <em>out</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#getGPRasDictFromString"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.getGPRasDictFromString" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a GPR string ‘((g1 and g2) or g3)’ to a gprDict which is returned</p>
<blockquote>
<div><ul class="simple">
<li><em>node</em> a Python AST note (e.g. <cite>ast.parse(gprstring).body[0]</cite>)</li>
<li><em>out</em> a new dictionary that will be be created in place</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.parseGeneAssociation">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">parseGeneAssociation</code><span class="sig-paren">(</span><em>gs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#parseGeneAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.parseGeneAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a COBRA style gene association into a nested list.</p>
<blockquote>
<div><ul class="simple">
<li><em>gs</em> a string containing a gene association</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCommon.processSpeciesChargeChemFormulaAnnot">
<code class="descclassname">cbmpy.CBCommon.</code><code class="descname">processSpeciesChargeChemFormulaAnnot</code><span class="sig-paren">(</span><em>s</em>, <em>getFromName=False</em>, <em>overwriteChemFormula=False</em>, <em>overwriteCharge=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCommon.html#processSpeciesChargeChemFormulaAnnot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCommon.processSpeciesChargeChemFormulaAnnot" title="Permalink to this definition">¶</a></dt>
<dd><p>Disambiguate the chemical formula from either the Notes or the overloaded name</p>
<blockquote>
<div><ul class="simple">
<li><em>s</em> a species object</li>
<li><em>getFromName</em> [default=False] whether to try strip the chemical formula from the name (old COBRA style)</li>
<li><em>overwriteChemFormula</em> [default=False]</li>
<li><em>overwriteCharge</em> [default=False]</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBConfig"></span><div class="section" id="cbmpy-cbconfig-module">
<h2>CBMPy: CBConfig module<a class="headerlink" href="#cbmpy-cbconfig-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBConfig.py 637 2017-12-15 11:34:59Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBConfig.current_version">
<code class="descclassname">cbmpy.CBConfig.</code><code class="descname">current_version</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBConfig.html#current_version"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBConfig.current_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the current CBMPy version as a string</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBConfig.current_version_tuple">
<code class="descclassname">cbmpy.CBConfig.</code><code class="descname">current_version_tuple</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBConfig.html#current_version_tuple"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBConfig.current_version_tuple" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the current CBMPy version as a tuple (x, y, z)</p>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBCPLEX"></span><div class="section" id="cbmpy-cbcplex-module">
<h2>CBMPy: CBCPLEX module<a class="headerlink" href="#cbmpy-cbcplex-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBCPLEX.py 629 2017-10-24 22:01:14Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_FluxVariabilityAnalysis">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_FluxVariabilityAnalysis</code><span class="sig-paren">(</span><em>fba</em>, <em>selected_reactions=None</em>, <em>pre_opt=True</em>, <em>tol=None</em>, <em>objF2constr=True</em>, <em>rhs_sense='lower'</em>, <em>optPercentage=100.0</em>, <em>work_dir=None</em>, <em>quiet=True</em>, <em>debug=False</em>, <em>oldlpgen=False</em>, <em>markupmodel=True</em>, <em>default_on_fail=False</em>, <em>roundoff_span=10</em>, <em>method='o'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_FluxVariabilityAnalysis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_FluxVariabilityAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a flux variability analysis on an fba model:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an FBA model object</li>
<li><em>selected reactions</em> [default=None] means use all reactions otherwise use the reactions listed here</li>
<li><em>pre_opt</em> [default=True] attempt to presolve the FBA and report its results in the ouput, if this is disabled and <em>objF2constr</em> is True then the rid/value of the current active objective is used</li>
<li><em>tol</em>  [default=None] do not floor/ceiling the objective function constraint, otherwise round of to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value*(optPercentage/100.0)</li>
<li><em>work_dir</em> [default=None] the FVA working directory for temporary files default = cwd+fva</li>
<li><em>debug</em> [default=False] if True write out all the intermediate FVA LP’s into work_dir</li>
<li><em>quiet</em> [default=False] if enabled, supress CPLEX output</li>
<li><em>objF2constr</em> [default=True] add the model objective function as a constraint using rhs_sense etc. If
this is True with pre_opt=False then the id/value of the active objective is used to form the constraint</li>
<li><em>markupmodel</em> [default=True] add the values returned by the fva to the reaction.fva_min and reaction.fva_max</li>
<li><em>default_on_fail</em> [default=False] if <em>pre_opt</em> is enabled replace a failed minimum/maximum with the solution value</li>
<li><em>roundoff_span</em> [default=10] number of digits is round off (not individual min/max values)</li>
<li><em>method</em> [default=’o’] choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details<ul>
<li>‘o’: auto</li>
<li>‘p’: primal</li>
<li>‘d’: dual</li>
<li>‘b’: barrier (no crossover)</li>
<li>‘h’: barrier</li>
<li>‘s’: sifting</li>
<li>‘c’: concurrent</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>Returns an array with columns: Reaction, Reduced Costs, Variability Min, Variability Max, abs(Max-Min), MinStatus, MaxStatus and a list containing the row names.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_MinimizeNumActiveFluxes">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_MinimizeNumActiveFluxes</code><span class="sig-paren">(</span><em>fba</em>, <em>selected_reactions=None</em>, <em>pre_opt=True</em>, <em>tol=None</em>, <em>objF2constr=True</em>, <em>rhs_sense='lower'</em>, <em>optPercentage=100.0</em>, <em>work_dir=None</em>, <em>quiet=False</em>, <em>debug=False</em>, <em>objective_coefficients=None</em>, <em>return_lp_obj=False</em>, <em>populate=None</em>, <em>oldlpgen=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_MinimizeNumActiveFluxes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_MinimizeNumActiveFluxes" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimize the sum of active fluxes, updates the model with the values of the solution and returns the value
of the MILP objective function (not the model objective function which remains unchanged). If population mode is activated
output is as described below:</p>
<blockquote>
<div><dl class="docutils">
<dt>Min: sum(Bi)</dt>
<dd>Bi = 0 -&gt; Ci Ji = 0</dd>
<dt>Such that:</dt>
<dd>NJi = 0
Jbio = opt</dd>
<dt>where:</dt>
<dd>Binary Bi</dd>
</dl>
</div></blockquote>
<p>Arguments:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an FBA model object</li>
<li><em>selected reactions</em> [default=None] means use all reactions otherwise use the reactions listed here</li>
<li><em>pre_opt</em> [default=True] attempt to presolve the FBA and report its results in the ouput, if this is diabled and <em>objF2constr</em> is True then the vid/value of the current active objective is used</li>
<li><em>tol</em>  [default=None] do not floor/ceiling the objective function constraint, otherwise round of to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em>
Note this does not necessarily mean the upper or lower bound, although practically it will. If in doubt use <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value * (optPercentage/100.0)</li>
<li><em>work_dir</em> [default=None] the MSAF working directory for temporary files default = cwd+fva</li>
<li><em>debug</em> [default=False] if True write out all the intermediate MSAF LP’s into work_dir</li>
<li><em>quiet</em> [default=False] if enabled supress CPLEX output</li>
<li><em>objF2constr</em> [default=True] add the model objective function as a constraint using rhs_sense etc. If
this is True with pre_opt=False then the id/value of the active objective is used to form the constraint</li>
<li><em>objective_coefficients</em> [default=None] a dictionary of (reaction_id : float) pairs that provide the are introduced as objective coefficients to the absolute flux value. Note that the default value of the coefficient (non-specified) is +1.</li>
<li><em>return_lp_obj</em> [default=False] off by default when enabled it returns the CPLEX LP object</li>
<li><em>populate</em> [default=None] enable search algorithm to find multiple (sub)optimal solutions. Set with a tuple of (RELGAP=0.0, POPULATE_LIMIT=20, TIME_LIMIT=300) suggested values only.
- <em>RELGAP</em> [default=0.0] relative gap to optimal solution
- <em>POPULATE_LIMIT</em> [default=20] terminate when so many solutions have been found
- <em>TIME_LIMIT</em> [default=300] terminate search after so many seconds important with higher values of <em>POPULATION_LIMIT</em></li>
<li><em>with_reduced_costs</em> [default=’uncsaled’] can be ‘scaled’, ‘unscaled’ or anything else which is None</li>
</ul>
</div></blockquote>
<p>With outputs:</p>
<blockquote>
<div><ul class="simple">
<li><em>mincnt</em> the objective function value OR</li>
<li><em>mincnt, cpx</em> the objective function and cplex model OR</li>
<li><em>populate_data, mincnt</em> a population data set OR</li>
<li><em>populate_data, mincnt, cpx</em> both the cps object and population data set</li>
</ul>
</div></blockquote>
<p>depending on selected flags.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_MinimizeSumOfAbsFluxes">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_MinimizeSumOfAbsFluxes</code><span class="sig-paren">(</span><em>fba</em>, <em>selected_reactions=None</em>, <em>pre_opt=True</em>, <em>tol=None</em>, <em>objF2constr=True</em>, <em>rhs_sense='lower'</em>, <em>optPercentage=100.0</em>, <em>work_dir=None</em>, <em>quiet=False</em>, <em>debug=False</em>, <em>objective_coefficients=None</em>, <em>return_lp_obj=False</em>, <em>oldlpgen=False</em>, <em>with_reduced_costs=None</em>, <em>method='o'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_MinimizeSumOfAbsFluxes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_MinimizeSumOfAbsFluxes" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimize the sum of absolute fluxes sum(abs(J1) + abs(J2) + abs(J3) … abs(Jn)) by adding two constraints per flux
and a variable representing the absolute value:</p>
<blockquote>
<div><dl class="docutils">
<dt>Min: Ci abs_Ji</dt>
<dd>Ji - abs_Ji &lt;= 0
Ji + abs_Ji &gt;= 0</dd>
<dt>Such that:</dt>
<dd>NJi = 0
Jopt = opt</dd>
</dl>
</div></blockquote>
<p>returns the value of the flux minimization objective function (not the model objective function which remains unchanged from)</p>
<p>Arguments:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an FBA model object</li>
<li><em>selected reactions</em> [default=None] means use all reactions otherwise use the reactions listed here</li>
<li><em>pre_opt</em> [default=True] attempt to presolve the FBA and report its results in the ouput, if this is disabled and <em>objF2constr</em> is True then the vid/value of the current active objective is used</li>
<li><em>tol</em>  [default=None] do not floor/ceiling the objective function constraint, otherwise round of to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value*(optPercentage/100.0)</li>
<li><em>work_dir</em> [default=None] the MSAF working directory for temporary files default = cwd+fva</li>
<li><em>debug</em> [default=False] if True write out all the intermediate MSAF LP’s into work_dir</li>
<li><em>quiet</em> [default=False] if enabled supress CPLEX output</li>
<li><em>objF2constr</em> [default=True] add the model objective function as a constraint using rhs_sense etc. If
this is True with pre_opt=False then the id/value of the active objective is used to form the constraint</li>
<li><em>objective_coefficients</em> [default=None] a dictionary of (reaction_id : float) pairs that provide the are introduced as objective coefficients to the absolute flux value. Note that the default value of the coefficient (non-specified) is +1.</li>
<li><em>return_lp_obj</em> [default=False] off by default when enabled it returns the CPLEX LP object</li>
<li><em>with_reduced_costs</em> [default=None] if not None should be ‘scaled’ or ‘unscaled’</li>
<li><em>method</em> [default=’o’] choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details<ul>
<li>‘o’: auto</li>
<li>‘p’: primal</li>
<li>‘d’: dual</li>
<li>‘b’: barrier (no crossover)</li>
<li>‘h’: barrier</li>
<li>‘s’: sifting</li>
<li>‘c’: concurrent</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>With outputs:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an update instance of a CBModel. Note that the FBA model objective function value is the original value set as a constraint</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_MultiFluxVariabilityAnalysis">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_MultiFluxVariabilityAnalysis</code><span class="sig-paren">(</span><em>lp</em>, <em>selected_reactions=None</em>, <em>tol=1e-10</em>, <em>rhs_sense='lower'</em>, <em>optPercentage=100.0</em>, <em>work_dir=None</em>, <em>debug=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_MultiFluxVariabilityAnalysis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_MultiFluxVariabilityAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a flux variability analysis on a multistate LP</p>
<blockquote>
<div><ul class="simple">
<li><em>lp</em> a multistate LP</li>
<li><em>selected reactions</em> [default=None] means use all reactions otherwise use the reactions listed here</li>
<li><em>pre_opt</em> [default=True] attempt to presolve the FBA and report its results in the ouput</li>
<li><em>tol</em>  [default=1e-10] do floor/ceiling the objective function constraint, otherwise floor/ceil to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value*(optPercentage/100.0)</li>
<li><em>work_dir</em> [default=None] the FVA working directory for temporary files default = cwd+fva</li>
<li><em>debug</em> [default=False] if True write out all the intermediate FVA LP’s into work_dir</li>
<li><em>bypass</em> [default=False] bypass everything and only run the min/max on lp</li>
</ul>
</div></blockquote>
<p>and returns an array with columns:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Reaction</span><span class="p">,</span> <span class="n">Reduced</span> <span class="n">Costs</span><span class="p">,</span> <span class="n">Variability</span> <span class="n">Min</span><span class="p">,</span> <span class="n">Variability</span> <span class="n">Max</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">Max</span><span class="o">-</span><span class="n">Min</span><span class="p">),</span> <span class="n">MinStatus</span><span class="p">,</span> <span class="n">MaxStatus</span>
</pre></div>
</div>
<p>and a list containing the row names.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_SolveMILP">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_SolveMILP</code><span class="sig-paren">(</span><em>c</em>, <em>auto_mipgap=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_SolveMILP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_SolveMILP" title="Permalink to this definition">¶</a></dt>
<dd><p>Solve and MILP</p>
<blockquote>
<div><ul class="simple">
<li><em>auto_mipgap</em> auto decrease mipgap until mipgap == absmipgap</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_WriteFVAtoCSV">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_WriteFVAtoCSV</code><span class="sig-paren">(</span><em>pid</em>, <em>fva</em>, <em>names</em>, <em>Dir=None</em>, <em>fbaObj=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_WriteFVAtoCSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_WriteFVAtoCSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method has been refactored to <cite>CBWrite.WriteFVAtoCSV()</cite>.</p>
<blockquote>
<div><ul class="simple">
<li><em>pid</em> filename_base for the CSV output</li>
<li><em>fva</em> FluxVariabilityAnalysis() OUTPUT_ARRAY</li>
<li><em>names</em> FluxVariabilityAnalysis() OUTPUT_NAMES</li>
<li><em>Dir</em> [default=None] if set the output directory for the csv files</li>
<li><em>fbaObj</em> [default=None] if supplied adds extra model information into the output tables</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_analyzeModel">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_analyzeModel</code><span class="sig-paren">(</span><em>f</em>, <em>lpFname=None</em>, <em>return_lp_obj=False</em>, <em>with_reduced_costs='unscaled'</em>, <em>with_sensitivity=False</em>, <em>del_intermediate=False</em>, <em>build_n=True</em>, <em>quiet=False</em>, <em>oldlpgen=False</em>, <em>method='o'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_analyzeModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_analyzeModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Optimize a model and add the result of the optimization to the model object
(e.g. <cite>reaction.value</cite>, <cite>objectiveFunction.value</cite>). The stoichiometric
matrix is automatically generated. This is a common function available
in all solver interfaces. By default returns the objective function value</p>
<blockquote>
<div><ul class="simple">
<li><em>f</em> an instantiated PySCeSCBM model object</li>
<li><em>lpFname</em> [default=None] the name of the intermediate LP file. If not specified no LP file is produced</li>
<li><em>return_lp_obj</em> [default=False] off by default when enabled it returns the CPLEX LP object</li>
<li><em>with_reduced_costs</em> [default=’unscaled’] calculate and add reduced cost information to mode this can be: ‘unscaled’ or ‘scaled’
or anything else which is interpreted as ‘None’. Scaled means s_rcost = (r.reduced_cost*rval)/obj_value</li>
<li><em>with_sensitivity</em> [default=False] add solution sensitivity information (not yet implemented)</li>
<li><em>del_intermediate</em> [default=False] redundant except if output file is produced and deleted (not useful)</li>
<li><em>build_n</em> [default=True] generate stoichiometry from the reaction network (reactions/reagents/species)</li>
<li><em>quiet</em> [default=False] suppress cplex output</li>
<li><em>method</em> [default=’o’] choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details<ul>
<li>‘o’: auto</li>
<li>‘p’: primal</li>
<li>‘d’: dual</li>
<li>‘b’: barrier (no crossover)</li>
<li>‘h’: barrier</li>
<li>‘s’: sifting</li>
<li>‘c’: concurrent</li>
</ul>
</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_constructLPfromFBA">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_constructLPfromFBA</code><span class="sig-paren">(</span><em>fba</em>, <em>fname=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_constructLPfromFBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_constructLPfromFBA" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a CPLEX LP in memory.
- <em>fba</em> an FBA object
- <em>fname</em> optional filename if defined writes out the constructed lp</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_fixConSense">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_fixConSense</code><span class="sig-paren">(</span><em>operator</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_fixConSense"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_fixConSense" title="Permalink to this definition">¶</a></dt>
<dd><p>Fixes the sense of inequality operators, returns corrected sense symbol</p>
<ul class="simple">
<li><em>operator</em> the operator to check</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_func_GetCPXandPresolve">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_func_GetCPXandPresolve</code><span class="sig-paren">(</span><em>fba</em>, <em>pre_opt</em>, <em>objF2constr</em>, <em>quiet=False</em>, <em>oldlpgen=False</em>, <em>with_reduced_costs='unscaled'</em>, <em>method='o'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_func_GetCPXandPresolve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_func_GetCPXandPresolve" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a utility function that does a presolve for FVA, MSAF etc. Generates properly formatted
empty objects if pre_opt == False</p>
<blockquote>
<div><ul class="simple">
<li><em>pre_opt</em> a boolean</li>
<li><em>fba</em> a CBModel object</li>
<li><em>objF2constr</em> add objective function as constraint</li>
<li><em>quiet</em> [default=False] supress cplex output</li>
<li><em>with_reduced_costs</em> [default=’unscaled’] can be ‘scaled’ or ‘unscaled’</li>
<li><em>method</em> [default=’o’] choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details<ul>
<li>‘o’: auto</li>
<li>‘p’: primal</li>
<li>‘d’: dual</li>
<li>‘b’: barrier (no crossover)</li>
<li>‘h’: barrier</li>
<li>‘s’: sifting</li>
<li>‘c’: concurrent</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>Returns: pre_sol, pre_oid, pre_oval, OPTIMAL_PRESOLUTION, REDUCED_COSTS</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_func_SetObjectiveFunctionAsConstraint">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_func_SetObjectiveFunctionAsConstraint</code><span class="sig-paren">(</span><em>cpx</em>, <em>rhs_sense</em>, <em>oval</em>, <em>tol</em>, <em>optPercentage</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_func_SetObjectiveFunctionAsConstraint"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_func_SetObjectiveFunctionAsConstraint" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Take the objective function and “optimum” value and add it as a constraint</dt>
<dd><ul class="first last simple">
<li><em>cpx</em> a cplex object</li>
<li><em>oval</em> the objective value</li>
<li><em>tol</em>  [default=None] do not floor/ceiling the objective function constraint, otherwise round of to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value*(optPercentage/100.0)</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getCPLEXModelFromLP">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getCPLEXModelFromLP</code><span class="sig-paren">(</span><em>lptFile</em>, <em>Dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getCPLEXModelFromLP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getCPLEXModelFromLP" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a LPT (CPLEX format) file and return a CPLX LP model</p>
<blockquote>
<div><ul class="simple">
<li><em>lptfile</em> an CPLEX LP format file</li>
<li><em>Dir</em> an optional directory</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getDualValues">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getDualValues</code><span class="sig-paren">(</span><em>c</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getDualValues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getDualValues" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the get the dual values of the solution</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a CPLEX LP</li>
</ul>
</div></blockquote>
<p>Output is a dictionary of {name : value} pairs</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getModelFromLP">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getModelFromLP</code><span class="sig-paren">(</span><em>lptFile</em>, <em>Dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getModelFromLP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getModelFromLP" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a LPT (CPLEX format) file and return a CPLX LP model</p>
<blockquote>
<div><ul class="simple">
<li><em>lptfile</em> an CPLEX LP format file</li>
<li><em>Dir</em> an optional directory</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getModelFromObj">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getModelFromObj</code><span class="sig-paren">(</span><em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getModelFromObj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getModelFromObj" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a CPLEX object from a FBA model object (via LP file)</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getOptimalSolution">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getOptimalSolution</code><span class="sig-paren">(</span><em>c</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getOptimalSolution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getOptimalSolution" title="Permalink to this definition">¶</a></dt>
<dd><p>From a CPLX model extract a tuple of solution, ObjFuncName and ObjFuncVal</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getOptimalSolution2">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getOptimalSolution2</code><span class="sig-paren">(</span><em>c</em>, <em>names</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getOptimalSolution2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getOptimalSolution2" title="Permalink to this definition">¶</a></dt>
<dd><p>From a CPLX model extract a tuple of solution, ObjFuncName and ObjFuncVal</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getReducedCosts">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getReducedCosts</code><span class="sig-paren">(</span><em>c</em>, <em>scaled=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getReducedCosts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getReducedCosts" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract ReducedCosts from LP and return as a dictionary ‘Rid’ : reduced cost</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a cplex LP object</li>
<li><em>scaled</em> scale the reduced cost by the optimal flux value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getSensitivities">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getSensitivities</code><span class="sig-paren">(</span><em>c</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getSensitivities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getSensitivities" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the sensitivities of each constraint on the objective function with inpt</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a CPLEX LP</li>
</ul>
</div></blockquote>
<p>Output is a tuple of bound and objective sensitivities where the objective
sensitivity is described in the CPLEX reference manual as:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">...</span> <span class="n">the</span> <span class="n">objective</span> <span class="n">sensitivity</span> <span class="n">shows</span> <span class="n">each</span> <span class="n">variable</span><span class="p">,</span> <span class="n">its</span> <span class="n">reduced</span> <span class="n">cost</span> <span class="ow">and</span> <span class="n">the</span> <span class="nb">range</span> <span class="n">over</span>
<span class="n">which</span> <span class="n">its</span> <span class="n">objective</span> <span class="n">function</span> <span class="n">coefficient</span> <span class="n">can</span> <span class="n">vary</span> <span class="n">without</span> <span class="n">forcing</span> <span class="n">a</span> <span class="n">change</span>
<span class="ow">in</span> <span class="n">the</span> <span class="n">optimal</span> <span class="n">basis</span><span class="o">.</span> <span class="n">The</span> <span class="n">current</span> <span class="n">value</span> <span class="n">of</span> <span class="n">each</span> <span class="n">objective</span> <span class="n">coefficient</span> <span class="ow">is</span>
<span class="n">also</span> <span class="n">displayed</span> <span class="k">for</span> <span class="n">reference</span><span class="o">.</span>

<span class="o">-</span> <span class="o">*</span><span class="n">objective</span> <span class="n">coefficient</span> <span class="n">sensitivity</span><span class="o">*</span> <span class="p">{</span><span class="n">flux</span> <span class="p">:</span> <span class="p">(</span><span class="n">reduced_cost</span><span class="p">,</span> <span class="n">lower_obj_sensitivity</span><span class="p">,</span> <span class="n">coeff_value</span><span class="p">,</span> <span class="n">upper_obj_sensitivity</span><span class="p">)}</span>
<span class="o">-</span> <span class="o">*</span><span class="n">rhs</span> <span class="n">sensitivity</span><span class="o">*</span> <span class="p">{</span><span class="n">constraint</span> <span class="p">:</span> <span class="p">(</span><span class="n">low</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">high</span><span class="p">)}</span>
<span class="o">-</span> <span class="o">*</span><span class="n">bound</span> <span class="n">sensitivity</span> <span class="n">ranges</span><span class="o">*</span> <span class="p">{</span><span class="n">flux</span> <span class="p">:</span> <span class="p">(</span><span class="n">lb_low</span><span class="p">,</span> <span class="n">lb_high</span><span class="p">,</span> <span class="n">ub_low</span><span class="p">,</span> <span class="n">ub_high</span><span class="p">)}</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getShadowPrices">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getShadowPrices</code><span class="sig-paren">(</span><em>c</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getShadowPrices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getShadowPrices" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of shadow prices containing ‘Rid’ : (lb, rhs, ub)</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a cplex LP object</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_getSolutionStatus">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_getSolutionStatus</code><span class="sig-paren">(</span><em>c</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_getSolutionStatus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_getSolutionStatus" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns one of:</p>
<blockquote>
<div><ul class="simple">
<li><em>LPS_OPT</em>: solution is optimal;</li>
<li><em>LPS_FEAS</em>: solution is feasible;</li>
<li><em>LPS_INFEAS</em>: solution is infeasible;</li>
<li><em>LPS_NOFEAS</em>: problem has no feasible solution;</li>
<li><em>LPS_UNBND</em>: problem has unbounded solution;</li>
<li><em>LPS_UNDEF</em>: solution is undefined.</li>
<li><em>LPS_NONE</em>: no solution</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_runInputScan">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_runInputScan</code><span class="sig-paren">(</span><em>fba</em>, <em>exDict</em>, <em>wDir</em>, <em>input_lb=-10.0</em>, <em>input_ub=0.0</em>, <em>writeHformat=False</em>, <em>rationalLPout=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_runInputScan"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_runInputScan" title="Permalink to this definition">¶</a></dt>
<dd><p>scans all inputs</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_setFBAsolutionToModel">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_setFBAsolutionToModel</code><span class="sig-paren">(</span><em>fba</em>, <em>lp</em>, <em>with_reduced_costs='unscaled'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_setFBAsolutionToModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_setFBAsolutionToModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the FBA solution from a CPLEX solution to an FBA object</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> and fba object</li>
<li><em>lp</em> a CPLEX LP object</li>
<li><em>with_reduced_costs</em> [default=’unscaled’] calculate and add reduced cost information to mode this can be: ‘unscaled’ or ‘scaled’
or anything else which is interpreted as None. Scaled is: s_rcost = (r.reduced_cost*rval)/obj_value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_setMIPGapTolerance">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_setMIPGapTolerance</code><span class="sig-paren">(</span><em>c</em>, <em>tol</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_setMIPGapTolerance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_setMIPGapTolerance" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the the relative MIP gap tolerance</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_setObjective">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_setObjective</code><span class="sig-paren">(</span><em>c</em>, <em>pid</em>, <em>expr=None</em>, <em>sense='maximize'</em>, <em>reset=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_setObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_setObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new objective function note that there is a major memory leak in
<cite>c.variables.get_names()</cite> whch is used when reset=True. If this is a problem
use cplx_setObjective2 which takes <em>names</em> as an input:</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a CPLEX LP object</li>
<li><em>pid</em> the r_id of the flux to be optimized</li>
<li><em>expr</em> a list of (coefficient, flux) pairs</li>
<li><em>sense</em> ‘maximize’/’minimize’</li>
<li><em>reset</em> [default=True] reset all objective function coefficients to zero</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_setObjective2">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_setObjective2</code><span class="sig-paren">(</span><em>c</em>, <em>pid</em>, <em>names</em>, <em>expr=None</em>, <em>sense='maximize'</em>, <em>reset=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_setObjective2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_setObjective2" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new objective function. This is a workaround function to avoid the
e is a major memory leak in <cite>c.variables.get_names()</cite> whch is used
in cplx_setObjective()  when reset=True.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_setOutputStreams">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_setOutputStreams</code><span class="sig-paren">(</span><em>lp</em>, <em>mode='default'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_setOutputStreams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_setOutputStreams" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the noise level of the solver, mode can be one of:</p>
<blockquote>
<div><ul class="simple">
<li><em>None</em> silent i.e. no output</li>
<li><em>‘file’</em> set solver to silent and output logs to <em>CPLX_RESULT_STREAM_FILE</em> cplex_output.log</li>
<li><em>‘iostream’</em> set solver to silent and output logs to <em>CPLX_RESULT_STREAM_IO</em> csio</li>
<li><em>‘default’</em> or anything else noisy with full output closes STREAM_IO and STREAM_FILE (default)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_setSolutionStatusToModel">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_setSolutionStatusToModel</code><span class="sig-paren">(</span><em>m</em>, <em>lp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_setSolutionStatusToModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_setSolutionStatusToModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the lp solutions status to the CBMPy model</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_singleGeneScan">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_singleGeneScan</code><span class="sig-paren">(</span><em>fba</em>, <em>r_off_low=0.0</em>, <em>r_off_upp=0.0</em>, <em>optrnd=8</em>, <em>altout=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_singleGeneScan"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_singleGeneScan" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a single gene deletion scan</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> a model object</li>
<li><em>r_off_low</em> the lower bound of a deactivated reaction</li>
<li><em>r_off_upp</em> the upper bound of a deactivated reaction</li>
<li><em>optrnd</em> [default=8] round off the optimal value</li>
<li><em>altout</em> [default=False] by default return a list of gene:opt pairs, alternatively (True) return an extended result set including gene groups, optima and effect map</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_writeLPsolution">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_writeLPsolution</code><span class="sig-paren">(</span><em>fba_sol</em>, <em>objf_name</em>, <em>fname</em>, <em>Dir=None</em>, <em>separator='</em>, <em>'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_writeLPsolution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_writeLPsolution" title="Permalink to this definition">¶</a></dt>
<dd><p>This function writes the optimal solution, produced wth <cite>cplx_getOptimalSolution</cite> to file</p>
<blockquote>
<div><ul class="simple">
<li><em>fba_sol</em> a dictionary of Flux : value pairs</li>
<li><em>objf_name</em> the objective function flux id</li>
<li><em>fname</em> the output filename</li>
<li><em>Dir</em> [default=None] use directory if not None</li>
<li><em>separator</em> [default=’,’] the column separator</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.cplx_writeLPtoLPTfile">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">cplx_writeLPtoLPTfile</code><span class="sig-paren">(</span><em>c</em>, <em>filename</em>, <em>title=None</em>, <em>Dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#cplx_writeLPtoLPTfile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.cplx_writeLPtoLPTfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Write out a CPLEX model as an LP format file</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.getReducedCosts">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">getReducedCosts</code><span class="sig-paren">(</span><em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#getReducedCosts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.getReducedCosts" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a dictionary of reduced costs for each reaction/flux</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBCPLEX.setReducedCosts">
<code class="descclassname">cbmpy.CBCPLEX.</code><code class="descname">setReducedCosts</code><span class="sig-paren">(</span><em>fba</em>, <em>reduced_costs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBCPLEX.html#setReducedCosts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBCPLEX.setReducedCosts" title="Permalink to this definition">¶</a></dt>
<dd><p>For each reaction/flux, sets the attribute “reduced_cost” from a dictionary of
reduced costs</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba object</li>
<li><em>reduced_costs</em> a dictionary of {reaction : value} pairs</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBDataStruct"></span><div class="section" id="cbmpy-cbdatastruct-module">
<h2>CBMPy: CBDataStruct module<a class="headerlink" href="#cbmpy-cbdatastruct-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBDataStruct.py 575 2017-04-13 12:18:44Z bgoli $)</p>
<dl class="class">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation">
<em class="property">class </em><code class="descclassname">cbmpy.CBDataStruct.</code><code class="descname">MIRIAMannotation</code><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>The MIRIAMannotation class MIRIAM annotations: Biological Qualifiers</p>
<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.addIDorgURI">
<code class="descname">addIDorgURI</code><span class="sig-paren">(</span><em>qual</em>, <em>uri</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.addIDorgURI"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.addIDorgURI" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a URI directly into a qualifier collection:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>uri</em> the complete identifiers.org uri e.g. <a class="reference external" href="http://identifiers.org/chebi/CHEBI:58088">http://identifiers.org/chebi/CHEBI:58088</a></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.addMIRIAMannotation">
<code class="descname">addMIRIAMannotation</code><span class="sig-paren">(</span><em>qual</em>, <em>entity</em>, <em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.addMIRIAMannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.addMIRIAMannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a qualified MIRIAM annotation or entity:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>entity</em> a MIRIAM resource entity e.g. “ChEBI”</li>
<li><em>mid</em> the entity id e.g. CHEBI:17158</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.checkEntity">
<code class="descname">checkEntity</code><span class="sig-paren">(</span><em>entity</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.checkEntity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.checkEntity" title="Permalink to this definition">¶</a></dt>
<dd><p>Check an entity entry, this is a MIRIAM resource name: “chEBI”. The test is case insensitive and will correct the case
of wrongly capitalised entities automatically. If the entity is not recognised then a list of possible candidates
based on the first letters of the input is displayed.</p>
<ul class="simple">
<li><em>entity</em> a MIRIAM resource entity e.g. “ChEBI”</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.checkEntityPattern">
<code class="descname">checkEntityPattern</code><span class="sig-paren">(</span><em>entity</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.checkEntityPattern"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.checkEntityPattern" title="Permalink to this definition">¶</a></dt>
<dd><p>For an entity key compile the pattern to a regex, if necessary.</p>
<blockquote>
<div><ul class="simple">
<li><em>entity</em> a MIRIAM resource entity</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.checkId">
<code class="descname">checkId</code><span class="sig-paren">(</span><em>entity</em>, <em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.checkId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.checkId" title="Permalink to this definition">¶</a></dt>
<dd><p>Check that a entity id e.g. CHEBI:17158</p>
<blockquote>
<div><ul class="simple">
<li><em>mid</em> the entity id e.g. CHEBI:17158</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.deleteMIRIAMannotation">
<code class="descname">deleteMIRIAMannotation</code><span class="sig-paren">(</span><em>qual</em>, <em>entity</em>, <em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.deleteMIRIAMannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.deleteMIRIAMannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes a qualified MIRIAM annotation or entity:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>entity</em> a MIRIAM resource entity e.g. “ChEBI”</li>
<li><em>mid</em> the entity id e.g. CHEBI:17158</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.getAllMIRIAMUris">
<code class="descname">getAllMIRIAMUris</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.getAllMIRIAMUris"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.getAllMIRIAMUris" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary of qualifiers that contain ID.org URL’S</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.getAndViewUrisForQualifier">
<code class="descname">getAndViewUrisForQualifier</code><span class="sig-paren">(</span><em>qual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.getAndViewUrisForQualifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.getAndViewUrisForQualifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve all url’s associated with qualifier and attempt to open them all in a new browser tab</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> the qualifier e.g. “is” or “isEncoded”</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.getMIRIAMUrisForQualifier">
<code class="descname">getMIRIAMUrisForQualifier</code><span class="sig-paren">(</span><em>qual</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.getMIRIAMUrisForQualifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.getMIRIAMUrisForQualifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all list of urls associated with qualifier:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> the qualifier e.g. “is” or “isEncoded”</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.MIRIAMannotation.viewURL">
<code class="descname">viewURL</code><span class="sig-paren">(</span><em>url</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#MIRIAMannotation.viewURL"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.MIRIAMannotation.viewURL" title="Permalink to this definition">¶</a></dt>
<dd><p>This will try to open the URL in a new tab of the default webbrowser</p>
<blockquote>
<div><ul class="simple">
<li><em>url</em> the url</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBDataStruct.StructMatrix">
<em class="property">class </em><code class="descclassname">cbmpy.CBDataStruct.</code><code class="descname">StructMatrix</code><span class="sig-paren">(</span><em>array</em>, <em>ridx</em>, <em>cidx</em>, <em>row=None</em>, <em>col=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>This class is specifically designed to store structural matrix information
give it an array and row/col index permutations it can generate its own
row/col labels given the label src.</p>
<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.getColsByIdx">
<code class="descname">getColsByIdx</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.getColsByIdx"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.getColsByIdx" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the columns referenced by index (1,3,5)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.getColsByName">
<code class="descname">getColsByName</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.getColsByName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.getColsByName" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the columns referenced by label (‘s’,’x’,’d’)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.getIndexes">
<code class="descname">getIndexes</code><span class="sig-paren">(</span><em>axis='all'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.getIndexes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.getIndexes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the matrix indexes ([rows],[cols]) where axis=’row’/’col’/’all’</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.getLabels">
<code class="descname">getLabels</code><span class="sig-paren">(</span><em>axis='all'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.getLabels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.getLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the matrix labels ([rows],[cols]) where axis=’row’/’col’/’all’</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.getRowsByIdx">
<code class="descname">getRowsByIdx</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.getRowsByIdx"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.getRowsByIdx" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the rows referenced by index (1,3,5)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.getRowsByName">
<code class="descname">getRowsByName</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.getRowsByName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.getRowsByName" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the rows referenced by label (‘s’,’x’,’d’)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.setCol">
<code class="descname">setCol</code><span class="sig-paren">(</span><em>src</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.setCol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.setCol" title="Permalink to this definition">¶</a></dt>
<dd><p>Assuming that the col index array is a permutation (full/subset)
of a source label array by supplying that src to setCol
maps the row labels to cidx and creates self.col (col label list)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrix.setRow">
<code class="descname">setRow</code><span class="sig-paren">(</span><em>src</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrix.setRow"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrix.setRow" title="Permalink to this definition">¶</a></dt>
<dd><p>Assuming that the row index array is a permutation (full/subset)
of a source label array by supplying that source to setRow it
maps the row labels to ridx and creates self.row (row label list)</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBDataStruct.StructMatrixLP">
<em class="property">class </em><code class="descclassname">cbmpy.CBDataStruct.</code><code class="descname">StructMatrixLP</code><span class="sig-paren">(</span><em>array</em>, <em>ridx</em>, <em>cidx</em>, <em>row=None</em>, <em>col=None</em>, <em>rhs=None</em>, <em>operators=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrixLP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrixLP" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds some stuff to StructMatrix that makes it LP friendly</p>
<dl class="method">
<dt id="cbmpy.CBDataStruct.StructMatrixLP.getCopy">
<code class="descname">getCopy</code><span class="sig-paren">(</span><em>attr_str</em>, <em>deep=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBDataStruct.html#StructMatrixLP.getCopy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBDataStruct.StructMatrixLP.getCopy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of the attribute with name attr_str. Uses the copy module <cite>copy.copy</cite> or <cite>copy.deepcopy</cite></p>
<ul class="simple">
<li><em>attr_str</em> a string of the attribute name: ‘row’, ‘col’</li>
<li><em>deep</em> [default=False] try to do a deepcopy. Use with caution see copy module docstring for details</li>
</ul>
</dd></dl>

</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBGLPK"></span><div class="section" id="cbmpy-cbglpk-module">
<h2>CBMPy: CBGLPK module<a class="headerlink" href="#cbmpy-cbglpk-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBGLPK.py 645 2018-04-23 22:38:14Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBGLPK.getReducedCosts">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">getReducedCosts</code><span class="sig-paren">(</span><em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#getReducedCosts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.getReducedCosts" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a dictionary of reduced costs for each reaction/flux</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_FluxVariabilityAnalysis">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_FluxVariabilityAnalysis</code><span class="sig-paren">(</span><em>fba</em>, <em>selected_reactions=None</em>, <em>pre_opt=True</em>, <em>tol=None</em>, <em>objF2constr=True</em>, <em>rhs_sense='lower'</em>, <em>optPercentage=100.0</em>, <em>work_dir=None</em>, <em>quiet=True</em>, <em>debug=False</em>, <em>oldlpgen=False</em>, <em>markupmodel=True</em>, <em>default_on_fail=False</em>, <em>roundoff_span=10</em>, <em>method='s'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_FluxVariabilityAnalysis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_FluxVariabilityAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a flux variability analysis on an fba model:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an FBA model object</li>
<li><em>selected reactions</em> [default=None] means use all reactions otherwise use the reactions listed here</li>
<li><em>pre_opt</em> [default=True] attempt to presolve the FBA and report its results in the ouput, if this is disabled and <em>objF2constr</em> is True then the vid/value of the current active objective is used</li>
<li><em>tol</em>  [default=None] do not floor/ceiling the objective function constraint, otherwise round of to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value*(optPercentage/100.0)</li>
<li><em>work_dir</em> [default=None] the FVA working directory for temporary files default = cwd+fva</li>
<li><em>debug</em> [default=False] if True write out all the intermediate FVA LP’s into work_dir</li>
<li><em>quiet</em> [default=False] if enabled supress CPLEX output</li>
<li><em>objF2constr</em> [default=True] add the model objective function as a constraint using rhs_sense etc. If
this is True with pre_opt=False then the id/value of the active objective is used to form the constraint</li>
<li><em>markupmodel</em> [default=True] add the values returned by the fva to the reaction.fva_min and reaction.fva_max</li>
<li><em>default_on_fail</em> [default=False] if <em>pre_opt</em> is enabled replace a failed minimum/maximum with the solution value</li>
<li><em>roundoff_span</em> [default=10] number of digits is round off (not individual min/max values)</li>
<li><em>method</em> [default=’s’] select the GLPK solver method, see the GLPK documentation for details<ul>
<li>‘s’: simplex</li>
<li>‘i’: interior</li>
<li>‘e’: exact</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>Returns an array with columns Reaction, Reduced Costs, Variability Min, Variability Max, abs(Max-Min), MinStatus, MaxStatus and a list containing the row names.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_MinimizeSumOfAbsFluxes">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_MinimizeSumOfAbsFluxes</code><span class="sig-paren">(</span><em>fba</em>, <em>selected_reactions=None</em>, <em>pre_opt=True</em>, <em>tol=None</em>, <em>objF2constr=True</em>, <em>rhs_sense='lower'</em>, <em>optPercentage=100.0</em>, <em>work_dir=None</em>, <em>quiet=False</em>, <em>debug=False</em>, <em>objective_coefficients={}</em>, <em>return_lp_obj=False</em>, <em>oldlpgen=False</em>, <em>with_reduced_costs=None</em>, <em>method='s'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_MinimizeSumOfAbsFluxes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_MinimizeSumOfAbsFluxes" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimize the sum of absolute fluxes sum(abs(J1) + abs(J2) + abs(J3) … abs(Jn)) by adding two constraints per flux
and a variable representing the absolute value:</p>
<blockquote>
<div><dl class="docutils">
<dt>Min: Ci abs_Ji</dt>
<dd>Ji - abs_Ji &lt;= 0
Ji + abs_Ji &gt;= 0</dd>
<dt>Such that:</dt>
<dd>NJi = 0
Jopt = opt</dd>
</dl>
</div></blockquote>
<p>returns the value of the flux minimization objective function (not the model objective function which remains unchanged from)</p>
<p>Arguments:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an FBA model object</li>
<li><em>selected reactions</em> [default=None] means use all reactions otherwise use the reactions listed here</li>
<li><em>pre_opt</em> [default=True] attempt to presolve the FBA and report its results in the ouput, if this is disabled and <em>objF2constr</em> is True then the vid/value of the current active objective is used</li>
<li><em>tol</em>  [default=None] do not floor/ceiling the objective function constraint, otherwise round of to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value*(optPercentage/100.0)</li>
<li><em>work_dir</em> [default=None] the MSAF working directory for temporary files default = cwd+fva</li>
<li><em>debug</em> [default=False] if True write out all the intermediate MSAF LP’s into work_dir</li>
<li><em>quiet</em> [default=False] if enabled supress CPLEX output</li>
<li><em>objF2constr</em> [default=True] add the model objective function as a constraint using rhs_sense etc. If
this is True with pre_opt=False then the id/value of the active objective is used to form the constraint</li>
<li><em>objective_coefficients</em> [default={}] a dictionary of (reaction_id : float) pairs that provide the are introduced as objective coefficients to the absolute flux value. Note that the default value of the coefficient (non-specified) is +1.</li>
<li><em>return_lp_obj</em> [default=False] off by default when enabled it returns the CPLEX LP object</li>
<li><em>with_reduced_costs</em> [default=None] if not None should be ‘scaled’ or ‘unscaled’</li>
<li><em>method</em> [default=’s’] select the GLPK solver method, see the GLPK documentation for details<ul>
<li>‘s’: simplex</li>
<li>‘i’: interior</li>
<li>‘e’: exact</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>With outputs:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an update instance of a CBModel. Note that the FBA model objective function value is the original value set as a constraint</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_Solve">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_Solve</code><span class="sig-paren">(</span><em>lp</em>, <em>method='s'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_Solve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_Solve" title="Permalink to this definition">¶</a></dt>
<dd><p>Solve the LP and create a status attribute with the solution status</p>
<blockquote>
<div><ul class="simple">
<li><em>method</em> [default=’s’] ‘s’ = simplex, ‘i’ = interior, ‘e’ = exact</li>
</ul>
</div></blockquote>
<p>GLPK solver options can be set in the dictionary GLPK_CFG</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_analyzeModel">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_analyzeModel</code><span class="sig-paren">(</span><em>f</em>, <em>lpFname=None</em>, <em>return_lp_obj=False</em>, <em>with_reduced_costs='unscaled'</em>, <em>with_sensitivity=False</em>, <em>del_intermediate=False</em>, <em>build_n=True</em>, <em>quiet=False</em>, <em>oldlpgen=False</em>, <em>method='s'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_analyzeModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_analyzeModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Optimize a model and add the result of the optimization to the model object
(e.g. <cite>reaction.value</cite>, <cite>objectiveFunction.value</cite>). The stoichiometric
matrix is automatically generated. This is a common function available in all
solver interfaces. By default returns the objective function value</p>
<blockquote>
<div><ul class="simple">
<li><em>f</em> an instantiated PySCeSCBM model object</li>
<li><em>lpFname</em> [default=None] the name of the intermediate LP file saved when this has a string value.</li>
<li><em>return_lp_obj</em> [default=False] off by default when enabled it returns the PyGLPK LP object</li>
<li><em>with_reduced_costs</em> [default=’unscaled’] calculate and add reduced cost information to mode this can be: ‘unscaled’ or ‘scaled’
or anything else which is interpreted as ‘None’. Scaled means s_rcost = (r.reduced_cost*rval)/obj_value</li>
<li><em>with_sensitivity</em> [default=False] add solution sensitivity information (not yet implemented)</li>
<li><em>del_intermediate</em> [default=False] delete the intermediary files after updating model object, useful for server applications</li>
<li><em>build_n</em> [default=True] generate stoichiometry from the reaction network (reactions/reagents/species)</li>
<li><em>quiet</em> [default=False] suppress glpk output</li>
<li><em>method</em> [default=’s’] select the GLPK solver method, see the GLPK documentation for details<ul>
<li>‘s’: simplex</li>
<li>‘i’: interior</li>
<li>‘e’: exact</li>
</ul>
</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_constructLPfromFBA">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_constructLPfromFBA</code><span class="sig-paren">(</span><em>fba</em>, <em>fname=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_constructLPfromFBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_constructLPfromFBA" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a GLPK LP in memory.
- <em>fba</em> an FBA object
- <em>fname</em> optional filename if defined writes out the constructed lp</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_func_GetCPXandPresolve">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_func_GetCPXandPresolve</code><span class="sig-paren">(</span><em>fba</em>, <em>pre_opt</em>, <em>objF2constr</em>, <em>quiet=False</em>, <em>oldlpgen=True</em>, <em>with_reduced_costs='unscaled'</em>, <em>method='s'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_func_GetCPXandPresolve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_func_GetCPXandPresolve" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a utility function that does a presolve for FVA, MSAF etc. Generates properly formatted
empty objects if pre_opt == False</p>
<blockquote>
<div><ul class="simple">
<li><em>pre_opt</em> a boolean</li>
<li><em>fba</em> a CBModel object</li>
<li><em>objF2constr</em> add objective function as constraint</li>
<li><em>quiet</em> [default=False] supress cplex output</li>
<li><em>with_reduced_costs</em> [default=’unscaled’] can be ‘scaled’ or ‘unscaled’</li>
<li><em>method</em> [default=’s’] select the GLPK solver method, see the GLPK documentation for details<ul>
<li>‘s’: simplex</li>
<li>‘i’: interior</li>
<li>‘e’: exact</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>Returns: pre_sol, pre_oid, pre_oval, OPTIMAL_PRESOLUTION, REDUCED_COSTS</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_func_SetObjectiveFunctionAsConstraint">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_func_SetObjectiveFunctionAsConstraint</code><span class="sig-paren">(</span><em>cpx</em>, <em>rhs_sense</em>, <em>oval</em>, <em>tol</em>, <em>optPercentage</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_func_SetObjectiveFunctionAsConstraint"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_func_SetObjectiveFunctionAsConstraint" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Take the objective function and “optimum” value and add it as a constraint</dt>
<dd><ul class="first last simple">
<li><em>cpx</em> a cplex object</li>
<li><em>oval</em> the objective value</li>
<li><em>tol</em>  [default=None] do not floor/ceiling the objective function constraint, otherwise round of to <em>tol</em></li>
<li><em>rhs_sense</em> [default=’lower’] means objC &gt;= objVal the inequality to use for the objective constraint can also be <em>upper</em> or <em>equal</em></li>
<li><em>optPercentage</em> [default=100.0] means the percentage optimal value to use for the RHS of the objective constraint: optimal_value*(optPercentage/100.0)</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_getOptimalSolution">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_getOptimalSolution</code><span class="sig-paren">(</span><em>c</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_getOptimalSolution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_getOptimalSolution" title="Permalink to this definition">¶</a></dt>
<dd><p>From a GLPK model extract a tuple of solution, ObjFuncName and ObjFuncVal</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_getReducedCosts">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_getReducedCosts</code><span class="sig-paren">(</span><em>c</em>, <em>scaled=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_getReducedCosts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_getReducedCosts" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract ReducedCosts from LP and return as a dictionary ‘Rid’ : reduced cost</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a GLPK LP object</li>
<li><em>scaled</em> scale the reduced cost by the optimal flux value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_getSolutionStatus">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_getSolutionStatus</code><span class="sig-paren">(</span><em>lp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_getSolutionStatus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_getSolutionStatus" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns one of:</p>
<blockquote>
<div><ul class="simple">
<li><em>LPS_OPT</em>: solution is optimal;</li>
<li><em>LPS_FEAS</em>: solution is feasible;</li>
<li><em>LPS_INFEAS</em>: solution is infeasible;</li>
<li><em>LPS_NOFEAS</em>: problem has no feasible solution;</li>
<li><em>LPS_UNBND</em>: problem has unbounded solution;</li>
<li><em>LPS_UNDEF</em>: solution is undefined.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_setFBAsolutionToModel">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_setFBAsolutionToModel</code><span class="sig-paren">(</span><em>fba</em>, <em>lp</em>, <em>with_reduced_costs='unscaled'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_setFBAsolutionToModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_setFBAsolutionToModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the FBA solution from a CPLEX solution to an FBA object</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> and fba object</li>
<li><em>lp</em> a CPLEX LP object</li>
<li><em>with_reduced_costs</em> [default=’unscaled’] calculate and add reduced cost information to mode this can be: ‘unscaled’ or ‘scaled’
or anything else which is interpreted as None. Scaled is: s_rcost = (r.reduced_cost*rval)/obj_value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_setObjective">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_setObjective</code><span class="sig-paren">(</span><em>c</em>, <em>oid</em>, <em>expr=None</em>, <em>sense='maximize'</em>, <em>reset=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_setObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_setObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a new objective function note that there is a major memory leak in
<cite>c.variables.get_names()</cite> whch is used when reset=True. If this is a problem
use cplx_setObjective2 which takes <em>names</em> as an input:</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a GLPK LP object</li>
<li><em>oid</em> the r_id of the flux to be optimized</li>
<li><em>expr</em> a list of (coefficient, flux) pairs</li>
<li><em>sense</em> ‘maximize’/’minimize’</li>
<li><em>reset</em> [default=True] reset all objective function coefficients to zero</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_setSingleConstraint">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_setSingleConstraint</code><span class="sig-paren">(</span><em>c</em>, <em>cid</em>, <em>expr=[]</em>, <em>sense='E'</em>, <em>rhs=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_setSingleConstraint"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_setSingleConstraint" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets a new sigle constraint to a GLPK model</p>
<blockquote>
<div><ul class="simple">
<li><em>c</em> a GLPK instance</li>
<li><em>cid</em> the constraint id</li>
<li><em>expr</em> a list of (coefficient, name) pairs</li>
<li><em>sense</em> [default=’G’] LGE</li>
<li><em>rhs</em> [default=0.0] the right hand side</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_setSolutionStatusToModel">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_setSolutionStatusToModel</code><span class="sig-paren">(</span><em>m</em>, <em>lp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_setSolutionStatusToModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_setSolutionStatusToModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the lp solutions status to the CBMPy model</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.glpk_writeLPtoLPTfile">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">glpk_writeLPtoLPTfile</code><span class="sig-paren">(</span><em>c</em>, <em>filename</em>, <em>title=None</em>, <em>Dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#glpk_writeLPtoLPTfile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.glpk_writeLPtoLPTfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Write out a GLPK model as an LP format file</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBGLPK.setReducedCosts">
<code class="descclassname">cbmpy.CBGLPK.</code><code class="descname">setReducedCosts</code><span class="sig-paren">(</span><em>fba</em>, <em>reduced_costs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGLPK.html#setReducedCosts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGLPK.setReducedCosts" title="Permalink to this definition">¶</a></dt>
<dd><p>For each reaction/flux, sets the attribute “reduced_cost” from a dictionary of
reduced costs</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba object</li>
<li><em>reduced_costs</em> a dictionary of {reaction : value} pairs</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBGUI"></span><div class="section" id="cbmpy-cbgui-module">
<h2>CBMPy: CBGUI module<a class="headerlink" href="#cbmpy-cbgui-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBGUI.py 575 2017-04-13 12:18:44Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBGUI.loadCBGUI">
<code class="descclassname">cbmpy.CBGUI.</code><code class="descname">loadCBGUI</code><span class="sig-paren">(</span><em>mod</em>, <em>version=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBGUI.html#loadCBGUI"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBGUI.loadCBGUI" title="Permalink to this definition">¶</a></dt>
<dd><p>Load an FBA model instance into the quick editor to view or change basic model properties</p>
<blockquote>
<div><ul class="simple">
<li><em>mod</em> a PySCeS CBMPy model instance</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBModel"></span><div class="section" id="cbmpy-cbmodel-module">
<h2>CBMPy: CBModel module<a class="headerlink" href="#cbmpy-cbmodel-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBModel.py 644 2018-03-14 21:40:47Z bgoli $)</p>
<dl class="class">
<dt id="cbmpy.CBModel.Compartment">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Compartment</code><span class="sig-paren">(</span><em>cid</em>, <em>name=None</em>, <em>size=1</em>, <em>dimensions=3</em>, <em>volume=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment" title="Permalink to this definition">¶</a></dt>
<dd><p>A compartment</p>
<dl class="method">
<dt id="cbmpy.CBModel.Compartment.containsReactions">
<code class="descname">containsReactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment.containsReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment.containsReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Lists the species contained in this compartment</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Compartment.containsSpecies">
<code class="descname">containsSpecies</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment.containsSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment.containsSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Lists the species contained in this compartment</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Compartment.getDimensions">
<code class="descname">getDimensions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment.getDimensions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment.getDimensions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the compartment dimensions</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Compartment.getSize">
<code class="descname">getSize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment.getSize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment.getSize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the compartment size</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Compartment.setDimensions">
<code class="descname">setDimensions</code><span class="sig-paren">(</span><em>dimensions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment.setDimensions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment.setDimensions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the compartment dimensions</p>
<blockquote>
<div><ul class="simple">
<li><em>dimensions</em> set the new compartment dimensions</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Compartment.setId">
<code class="descname">setId</code><span class="sig-paren">(</span><em>fid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment.setId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment.setId" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object Id</p>
<blockquote>
<div><ul class="simple">
<li><em>fid</em> a valid c variable style id string</li>
</ul>
<p>Reimplements &#64;FBase.setId()</p>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Compartment.setSize">
<code class="descname">setSize</code><span class="sig-paren">(</span><em>size</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Compartment.setSize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Compartment.setSize" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the compartment size</p>
<blockquote>
<div><ul class="simple">
<li><em>size</em> the new compartment size</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Fbase">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Fbase</code><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class for CB Model objects</p>
<dl class="method">
<dt id="cbmpy.CBModel.Fbase.addMIRIAMannotation">
<code class="descname">addMIRIAMannotation</code><span class="sig-paren">(</span><em>qual</em>, <em>entity</em>, <em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.addMIRIAMannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.addMIRIAMannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a qualified MIRIAM annotation or entity:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>entity</em> a MIRIAM resource entity e.g. “ChEBI”</li>
<li><em>mid</em> the entity id e.g. CHEBI:17158 or fully qualifies url (if only_qual_uri)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.addMIRIAMuri">
<code class="descname">addMIRIAMuri</code><span class="sig-paren">(</span><em>qual</em>, <em>uri</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.addMIRIAMuri"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.addMIRIAMuri" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a qualified MIRIAM annotation or entity:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>uri</em> the fully qualified entity id e.g. <a class="reference external" href="http://identifiers.org/chebi/CHEBI:12345">http://identifiers.org/chebi/CHEBI:12345</a> (no validity checking is done)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.clone">
<code class="descname">clone</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.clone"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.clone" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a clone of this object. Cloning performs a deepcop on the object which will also clone
any objects that exist as attributes of this object, in other words an independent copy of the
original. If this is not the desired behaviour override this method when subclassing or implement
your own.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.deleteAnnotation">
<code class="descname">deleteAnnotation</code><span class="sig-paren">(</span><em>key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.deleteAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.deleteAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Unsets (deltes) an objects annotation with key</p>
<blockquote>
<div><ul class="simple">
<li><em>key</em> the annotation key</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.deleteMIRIAMannotation">
<code class="descname">deleteMIRIAMannotation</code><span class="sig-paren">(</span><em>qual</em>, <em>entity</em>, <em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.deleteMIRIAMannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.deleteMIRIAMannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes a qualified MIRIAM annotation or entity:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>entity</em> a MIRIAM resource entity e.g. “ChEBI”</li>
<li><em>mid</em> the entity id e.g. CHEBI:17158</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getAnnotation">
<code class="descname">getAnnotation</code><span class="sig-paren">(</span><em>key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the object annotation associated with:</p>
<blockquote>
<div><ul class="simple">
<li><em>key</em> the annotation key</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getAnnotations">
<code class="descname">getAnnotations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getAnnotations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the object annotation dictionary</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getCompartmentId">
<code class="descname">getCompartmentId</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getCompartmentId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getCompartmentId" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the compartment id where this element is located</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getId">
<code class="descname">getId</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getId" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the object ID.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getMIRIAMannotations">
<code class="descname">getMIRIAMannotations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getMIRIAMannotations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getMIRIAMannotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of all MIRIAM annotations associated with this object
or None of there are none defined.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getMetaId">
<code class="descname">getMetaId</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getMetaId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getMetaId" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the object metaId.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getName">
<code class="descname">getName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getName" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the object name.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getNotes">
<code class="descname">getNotes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getNotes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getNotes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the object’s notes</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getPid">
<code class="descname">getPid</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getPid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getPid" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the object ID.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.getSBOterm">
<code class="descname">getSBOterm</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.getSBOterm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.getSBOterm" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the SBO term for this object.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.hasAnnotation">
<code class="descname">hasAnnotation</code><span class="sig-paren">(</span><em>key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.hasAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.hasAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a boolean representing the presence/absence of the key in the objext annotation</p>
<blockquote>
<div><ul class="simple">
<li><em>key</em> the annotation key</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.serialize">
<code class="descname">serialize</code><span class="sig-paren">(</span><em>protocol=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.serialize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.serialize" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize object, returns a string by default</p>
<blockquote>
<div><ul class="simple">
<li><dl class="first docutils">
<dt><em>protocol</em> [default=0] serialize to a string or binary if required,</dt>
<dd>see pickle module documentation for details</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p># Reimplemented in Model</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.serializeToDisk">
<code class="descname">serializeToDisk</code><span class="sig-paren">(</span><em>filename</em>, <em>protocol=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.serializeToDisk"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.serializeToDisk" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize to disk using pickle protocol:</p>
<blockquote>
<div><ul class="simple">
<li><em>filename</em> the name of the output file</li>
<li><dl class="first docutils">
<dt><em>protocol</em> [default=2] serialize to a string or binary if required,</dt>
<dd>see pickle module documentation for details</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p># Reimplemented in Model</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setAnnotation">
<code class="descname">setAnnotation</code><span class="sig-paren">(</span><em>key</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Set an objects annotation as a key : value pair.</p>
<blockquote>
<div><ul class="simple">
<li><em>key</em> the annotation key</li>
<li><em>value</em> the annotation value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setCompartmentId">
<code class="descname">setCompartmentId</code><span class="sig-paren">(</span><em>compartment</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setCompartmentId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setCompartmentId" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the compartment id where this element is located</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setId">
<code class="descname">setId</code><span class="sig-paren">(</span><em>fid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setId" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object Id</p>
<blockquote>
<div><ul class="simple">
<li><em>fid</em> a valid c variable style id string</li>
</ul>
<p>Reimplemented by &#64;Reaction, &#64;Species, &#64;Compartment, &#64;Gene</p>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setMetaId">
<code class="descname">setMetaId</code><span class="sig-paren">(</span><em>mid=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setMetaId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setMetaId" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object Id</p>
<blockquote>
<div><ul class="simple">
<li><em>mid</em> [default=None] a valid c variable style metaid string, if None it will be set as meta+id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setName">
<code class="descname">setName</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setName" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the object name:</p>
<blockquote>
<div><ul class="simple">
<li><em>name</em> the name string</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setNotes">
<code class="descname">setNotes</code><span class="sig-paren">(</span><em>notes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setNotes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setNotes" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object’s notes:</p>
<blockquote>
<div><ul class="simple">
<li><em>notes</em> the note string, should preferably be (X)HTML for SBML</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setPid">
<code class="descname">setPid</code><span class="sig-paren">(</span><em>fid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setPid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setPid" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object Id</p>
<blockquote>
<div><ul class="simple">
<li><em>fid</em> a valid c variable style id string</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Fbase.setSBOterm">
<code class="descname">setSBOterm</code><span class="sig-paren">(</span><em>sbo</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Fbase.setSBOterm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Fbase.setSBOterm" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the SBO term for this object.</p>
<blockquote>
<div><ul class="simple">
<li><em>sbo</em> the SBOterm with format: SBO:nnnnnnn”</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.FluxBound">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">FluxBound</code><span class="sig-paren">(</span><em>fid</em>, <em>reaction</em>, <em>operation</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#FluxBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.FluxBound" title="Permalink to this definition">¶</a></dt>
<dd><p>A reaction fluxbound</p>
<dl class="method">
<dt id="cbmpy.CBModel.FluxBound.getType">
<code class="descname">getType</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#FluxBound.getType"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.FluxBound.getType" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <em>type</em> of FluxBound: ‘lower’, ‘upper’, ‘equality’ or None</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.FluxBound.getValue">
<code class="descname">getValue</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#FluxBound.getValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.FluxBound.getValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the current value of the attribute (input/solution)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.FluxBound.setReactionId">
<code class="descname">setReactionId</code><span class="sig-paren">(</span><em>react</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#FluxBound.setReactionId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.FluxBound.setReactionId" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the reaction attribute of the FluxBound</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.FluxBound.setValue">
<code class="descname">setValue</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#FluxBound.setValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.FluxBound.setValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the attribute ‘’value’‘</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.FluxObjective">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">FluxObjective</code><span class="sig-paren">(</span><em>pid</em>, <em>reaction</em>, <em>coefficient=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#FluxObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.FluxObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>A weighted flux that appears in an objective function</p>
<p>NOTE: reaction is a string containing a reaction id</p>
</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Gene">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Gene</code><span class="sig-paren">(</span><em>gid</em>, <em>label=None</em>, <em>active=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Contains all the information about a gene (or gene+protein construct depending on your philosophy)</p>
<dl class="method">
<dt id="cbmpy.CBModel.Gene.getLabel">
<code class="descname">getLabel</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene.getLabel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene.getLabel" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the gene label</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Gene.isActive">
<code class="descname">isActive</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene.isActive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene.isActive" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns whether the gene is active or not</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Gene.resetActivity">
<code class="descname">resetActivity</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene.resetActivity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene.resetActivity" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset the gene to its default activity state</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Gene.setActive">
<code class="descname">setActive</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene.setActive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene.setActive" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the gene to be active</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Gene.setId">
<code class="descname">setId</code><span class="sig-paren">(</span><em>fid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene.setId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene.setId" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object Id</p>
<blockquote>
<div><ul class="simple">
<li><em>fid</em> a valid c variable style id string</li>
</ul>
<p>Reimplements Fbase method</p>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Gene.setInactive">
<code class="descname">setInactive</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene.setInactive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene.setInactive" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the gene to be inactive</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Gene.setLabel">
<code class="descname">setLabel</code><span class="sig-paren">(</span><em>label</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Gene.setLabel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Gene.setLabel" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the gene label, checks that the new label is unique</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.GeneProteinAssociation">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">GeneProteinAssociation</code><span class="sig-paren">(</span><em>gpid</em>, <em>protein</em>, <em>use_compiled=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>This class associates genes to proteins.</p>
<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.addAssociation">
<code class="descname">addAssociation</code><span class="sig-paren">(</span><em>assoc</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.addAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.addAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a gene/protein association expression</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.addGeneref">
<code class="descname">addGeneref</code><span class="sig-paren">(</span><em>geneid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.addGeneref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.addGeneref" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a gene reference to the list of gene references</p>
<blockquote>
<div><ul class="simple">
<li><em>geneid</em> a valid model Gene id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.createAssociationAndGeneRefsFromString">
<code class="descname">createAssociationAndGeneRefsFromString</code><span class="sig-paren">(</span><em>assoc</em>, <em>altlabels=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.createAssociationAndGeneRefsFromString"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.createAssociationAndGeneRefsFromString" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate the gene/protein association and add the genes necessary to evaluate it
Note that this GPR should be added to a model with cmod.addGPRAssociation() before calling this method</p>
<blockquote>
<div><ul class="simple">
<li><em>assoc</em> the COBRA style gene protein association</li>
<li><em>altlabels</em> [default=None] a dictionary containing a label&lt;–&gt;id mapping</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.createAssociationAndGeneRefsFromTree">
<code class="descname">createAssociationAndGeneRefsFromTree</code><span class="sig-paren">(</span><em>gprtree</em>, <em>altlabels=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.createAssociationAndGeneRefsFromTree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.createAssociationAndGeneRefsFromTree" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate the GPR tree and add the genes necessary to evaluate it
Note that this GPR should be added to a model with cmod.addGPRAssociation() before calling this method</p>
<blockquote>
<div><ul class="simple">
<li><em>gprtree</em> the CBMPy GPR tree data structure</li>
<li><em>altlabels</em> [default=None] a dictionary containing a label&lt;–&gt;id mapping</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.deleteGeneFromAssociation">
<code class="descname">deleteGeneFromAssociation</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.deleteGeneFromAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.deleteGeneFromAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes a gene id from the gene association. <em>WARNING</em> this process is irreversible!!</p>
<ul class="simple">
<li><em>gid</em> a valid gene identifier (not label)</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.deleteGeneref">
<code class="descname">deleteGeneref</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.deleteGeneref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.deleteGeneref" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes a gene reference</p>
<ul class="simple">
<li><em>geneid</em> a valid model Gene id</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.evalAssociation">
<code class="descname">evalAssociation</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.evalAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.evalAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an integer value representing the logical associations or None.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getActiveGenes">
<code class="descname">getActiveGenes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getActiveGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getActiveGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of active gene objects</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getAssociationStr">
<code class="descname">getAssociationStr</code><span class="sig-paren">(</span><em>use_labels=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getAssociationStr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getAssociationStr" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the gene association string, alternatively return string with labels</p>
<ul class="simple">
<li><em>use_lablels</em> [default=False] return the gene association string with labels rather than geneId’s (FBCv2 issue)</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getGene">
<code class="descname">getGene</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getGene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getGene" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a gene object with id</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getGeneIds">
<code class="descname">getGeneIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getGeneIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getGeneIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of gene id’s</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getGeneLabels">
<code class="descname">getGeneLabels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getGeneLabels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getGeneLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of gene labels associated with this GPRass</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getGenes">
<code class="descname">getGenes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of gene objects associated with this GPRass</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getProtein">
<code class="descname">getProtein</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getProtein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getProtein" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the protein associated with this set of genes</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.getTree">
<code class="descname">getTree</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.getTree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.getTree" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the dictionary/tree representation of the GPR</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.isProteinActive">
<code class="descname">isProteinActive</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.isProteinActive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.isProteinActive" title="Permalink to this definition">¶</a></dt>
<dd><p>This returns a boolean which indicates the result of evaluating the gene association. If the result is positive
then the protein is expressed and <em>True</em> is returned, otherwise if the expression evaluates to a value of 0 then
the protein is not expressed and  <em>False</em> is returned.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.setAllGenesActive">
<code class="descname">setAllGenesActive</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.setAllGenesActive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.setAllGenesActive" title="Permalink to this definition">¶</a></dt>
<dd><p>Activate all genes in association</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.setAllGenesInactive">
<code class="descname">setAllGenesInactive</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.setAllGenesInactive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.setAllGenesInactive" title="Permalink to this definition">¶</a></dt>
<dd><p>Deactivates all genes in association</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.setGeneActive">
<code class="descname">setGeneActive</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.setGeneActive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.setGeneActive" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a gene to be inactive</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.setGeneInactive">
<code class="descname">setGeneInactive</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.setGeneInactive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.setGeneInactive" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a gene to be inactive</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.setProtein">
<code class="descname">setProtein</code><span class="sig-paren">(</span><em>protein</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.setProtein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.setProtein" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the protein associated with this set of genes</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.GeneProteinAssociation.setTree">
<code class="descname">setTree</code><span class="sig-paren">(</span><em>tree</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#GeneProteinAssociation.setTree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GeneProteinAssociation.setTree" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a GPR dictionary/tree representation to the GPR.</p>
<blockquote>
<div><ul class="simple">
<li><em>tree</em> a dictionary representation of a GPR.</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Group">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Group</code><span class="sig-paren">(</span><em>pid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group" title="Permalink to this definition">¶</a></dt>
<dd><p>Container for SBML groups</p>
<dl class="method">
<dt id="cbmpy.CBModel.Group.addMember">
<code class="descname">addMember</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.addMember"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.addMember" title="Permalink to this definition">¶</a></dt>
<dd><p>Add member CBMPy object(s) to the group</p>
<blockquote>
<div><ul class="simple">
<li><em>obj</em> either a single, tuple or list of CBMPy objects</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.addSharedMIRIAMannotation">
<code class="descname">addSharedMIRIAMannotation</code><span class="sig-paren">(</span><em>qual</em>, <em>entity</em>, <em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.addSharedMIRIAMannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.addSharedMIRIAMannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a qualified MIRIAM annotation or entity to the list of members (all) rather than the group itself:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>entity</em> a MIRIAM resource entity e.g. “ChEBI”</li>
<li><em>mid</em> the entity id e.g. CHEBI:17158 or fully qualifies url (if only_qual_uri)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.assignAllSharedPropertiesToMembers">
<code class="descname">assignAllSharedPropertiesToMembers</code><span class="sig-paren">(</span><em>overwrite=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.assignAllSharedPropertiesToMembers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.assignAllSharedPropertiesToMembers" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns all group shared properties (notes, annotations, MIRIAM annotations, SBO) to the group members.</p>
<blockquote>
<div><ul class="simple">
<li><em>overwrite</em> [default=False] overwrite the target notes if they are defined</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.assignSharedAnnotationToMembers">
<code class="descname">assignSharedAnnotationToMembers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.assignSharedAnnotationToMembers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.assignSharedAnnotationToMembers" title="Permalink to this definition">¶</a></dt>
<dd><p>This function merges or updates the group member objects annotations with the group shared annotation.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.assignSharedMIRIAMannotationToMembers">
<code class="descname">assignSharedMIRIAMannotationToMembers</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.assignSharedMIRIAMannotationToMembers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.assignSharedMIRIAMannotationToMembers" title="Permalink to this definition">¶</a></dt>
<dd><p>This function merges or updates the group member objects MIRIAM annotations with the group shared MIRIAM annotation.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.assignSharedNotesToMembers">
<code class="descname">assignSharedNotesToMembers</code><span class="sig-paren">(</span><em>overwrite=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.assignSharedNotesToMembers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.assignSharedNotesToMembers" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns the group shared notes to the group members.</p>
<blockquote>
<div><ul class="simple">
<li><em>overwrite</em> [default=False] overwrite the target notes if they are defined</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.assignSharedSBOtermsToMembers">
<code class="descname">assignSharedSBOtermsToMembers</code><span class="sig-paren">(</span><em>overwrite=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.assignSharedSBOtermsToMembers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.assignSharedSBOtermsToMembers" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns the group shared member SBO term to the group members.</p>
<blockquote>
<div><ul class="simple">
<li><em>overwrite</em> [default=False] overwrite the target SBO term if it is defined</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.clone">
<code class="descname">clone</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.clone"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.clone" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a clone of this object. Note the for Groups this is a shallow copy, in that the reference
objects themselves are not cloned only the group (and attributes)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.deleteMember">
<code class="descname">deleteMember</code><span class="sig-paren">(</span><em>oid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.deleteMember"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.deleteMember" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes a group member with group id.</p>
<blockquote>
<div><ul class="simple">
<li><em>oid</em> group member id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.getKind">
<code class="descname">getKind</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.getKind"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.getKind" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the group kind</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.getMemberIDs">
<code class="descname">getMemberIDs</code><span class="sig-paren">(</span><em>as_set=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.getMemberIDs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.getMemberIDs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the ids of the member objects.</p>
<blockquote>
<div><ul class="simple">
<li><em>as_set</em> return id’s as a set rather than a list</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.getMembers">
<code class="descname">getMembers</code><span class="sig-paren">(</span><em>as_set=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.getMembers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.getMembers" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the member objects of the group.</p>
<blockquote>
<div><ul class="simple">
<li><em>as_set</em> return objects as a set rather than a list</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.getSharedAnnotations">
<code class="descname">getSharedAnnotations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.getSharedAnnotations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.getSharedAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary of the shared member annotations (rather than the group attribute).</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.getSharedMIRIAMannotations">
<code class="descname">getSharedMIRIAMannotations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.getSharedMIRIAMannotations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.getSharedMIRIAMannotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary of the shared member MIRIAM annotations (rather than the group attribute).</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.getSharedNotes">
<code class="descname">getSharedNotes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.getSharedNotes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.getSharedNotes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the shared member notes (rather than the group attribute).</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.getSharedSBOterm">
<code class="descname">getSharedSBOterm</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.getSharedSBOterm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.getSharedSBOterm" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the shared member SBO term (rather than the group attribute).</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.hasMember">
<code class="descname">hasMember</code><span class="sig-paren">(</span><em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.hasMember"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.hasMember" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a boolean indicating whether a member is in the group.</p>
<blockquote>
<div><ul class="simple">
<li><em>mid</em> the id to check</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.setKind">
<code class="descname">setKind</code><span class="sig-paren">(</span><em>kind</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.setKind"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.setKind" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the kind or type of the group, this must be one of: ‘collection’, ‘partonomy’, ‘classification’.</p>
<blockquote>
<div><ul class="simple">
<li><em>kind</em> the kind</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.setSharedAnnotation">
<code class="descname">setSharedAnnotation</code><span class="sig-paren">(</span><em>key</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.setSharedAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.setSharedAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the list of members (all) annotation as a key : value pair.</p>
<blockquote>
<div><ul class="simple">
<li><em>key</em> the annotation key</li>
<li><em>value</em> the annotation value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.setSharedNotes">
<code class="descname">setSharedNotes</code><span class="sig-paren">(</span><em>notes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.setSharedNotes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.setSharedNotes" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the group of objects notes attribute (all):</p>
<blockquote>
<div><ul class="simple">
<li><em>notes</em> the note string, should preferably be (X)HTML for SBML</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Group.setSharedSBOterm">
<code class="descname">setSharedSBOterm</code><span class="sig-paren">(</span><em>sbo</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Group.setSharedSBOterm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Group.setSharedSBOterm" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the SBO term for the the members of the group (all).</p>
<blockquote>
<div><ul class="simple">
<li><em>sbo</em> the SBOterm with format: “SBO:&lt;7 digit integer&gt;”</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.GroupMemberAttributes">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">GroupMemberAttributes</code><a class="reference internal" href="_modules/cbmpy/CBModel.html#GroupMemberAttributes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.GroupMemberAttributes" title="Permalink to this definition">¶</a></dt>
<dd><p>Contains the shared attributes of the group members (equivalent to SBML annotation on ListOfMembers)</p>
</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Model">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Model</code><span class="sig-paren">(</span><em>pid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model" title="Permalink to this definition">¶</a></dt>
<dd><p>Container for constraint based model, adds methods for manipulating:</p>
<blockquote>
<div><ul class="simple">
<li>objectives</li>
<li>constraints</li>
<li>reactions</li>
<li>species</li>
<li>compartments</li>
<li>groups</li>
<li>parameters</li>
<li>N a structmatrix object</li>
</ul>
</div></blockquote>
<dl class="method">
<dt id="cbmpy.CBModel.Model.addCompartment">
<code class="descname">addCompartment</code><span class="sig-paren">(</span><em>comp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addCompartment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addCompartment" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an instantiated Compartment object to the CBM model</p>
<ul class="simple">
<li><em>comp</em> an instance of the Compartment class</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addFluxBound">
<code class="descname">addFluxBound</code><span class="sig-paren">(</span><em>fluxbound</em>, <em>fbexists=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addFluxBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addFluxBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an instantiated FluxBound object to the FBA model</p>
<ul class="simple">
<li><em>fluxbound</em> an instance of the FluxBound class</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addGPRAssociation">
<code class="descname">addGPRAssociation</code><span class="sig-paren">(</span><em>gpr</em>, <em>update_idx=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addGPRAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addGPRAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a GeneProteinAssociation instance to the model</p>
<blockquote>
<div><ul class="simple">
<li><em>gpr</em> an instantiated GeneProteinAssociation object</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addGene">
<code class="descname">addGene</code><span class="sig-paren">(</span><em>gene</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addGene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addGene" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an instantiated Gene object to the FBA model</p>
<ul class="simple">
<li><em>gene</em> an instance of the G class</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addGroup">
<code class="descname">addGroup</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an instantiated group object to the model</p>
<blockquote>
<div><ul class="simple">
<li><em>obj</em> the Group instance</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addMIRIAMannotation">
<code class="descname">addMIRIAMannotation</code><span class="sig-paren">(</span><em>qual</em>, <em>entity</em>, <em>mid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addMIRIAMannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addMIRIAMannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a qualified MIRIAM annotation or entity:</p>
<blockquote>
<div><ul class="simple">
<li><em>qual</em> a Biomodels biological qualifier e.g. “is” “isEncodedBy”</li>
<li><em>entity</em> a MIRIAM resource entity e.g. “ChEBI”</li>
<li><em>mid</em> the entity id e.g. CHEBI:17158</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addModelCreator">
<code class="descname">addModelCreator</code><span class="sig-paren">(</span><em>firstname</em>, <em>lastname</em>, <em>organisation=None</em>, <em>email=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addModelCreator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addModelCreator" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a model creator to the list of model creators, only the first and fmaily names are mandatory:</p>
<blockquote>
<div><ul class="simple">
<li><em>firstname</em></li>
<li><em>lastname</em></li>
<li><em>organisation</em> [default=None]</li>
<li><em>email</em>  [default=None]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addObjective">
<code class="descname">addObjective</code><span class="sig-paren">(</span><em>obj</em>, <em>active=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an instantiated Objective object to the FBA model</p>
<ul class="simple">
<li><em>obj</em> an instance of the Objective class</li>
<li><em>active</em> [default=False] flag this objective as the active objective (fba.activeObjIdx)</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addParameter">
<code class="descname">addParameter</code><span class="sig-paren">(</span><em>par</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addParameter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addParameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an instantiated Parameter object to the model</p>
<ul class="simple">
<li><em>par</em> an instance of the Parameter class</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addReaction">
<code class="descname">addReaction</code><span class="sig-paren">(</span><em>reaction</em>, <em>create_default_bounds=False</em>, <em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a reaction object to the model</p>
<ul class="simple">
<li><em>reaction</em> an instance of the Reaction class</li>
<li><em>create_default_bounds</em> create default reaction bounds, irreversible 0 &lt;= J &lt;= INF, reversable -INF &lt;= J &lt;= INF</li>
<li><em>silent</em> [default=False] if enabled this disables the printing of information messages</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addSpecies">
<code class="descname">addSpecies</code><span class="sig-paren">(</span><em>species</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an instantiated Species object to the FBA model</p>
<ul class="simple">
<li><em>species</em> an instance of the Species class</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.addUserConstraint">
<code class="descname">addUserConstraint</code><span class="sig-paren">(</span><em>pid</em>, <em>fluxes=None</em>, <em>operator='='</em>, <em>rhs=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.addUserConstraint"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.addUserConstraint" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a user defined constraint to FBA model, this is additional to the automatically determined Stoichiometric constraints.</p>
<blockquote>
<div><ul class="simple">
<li><em>pid</em> user constraint name/id, use <cite>None</cite> for auto-assign</li>
<li><em>fluxes</em> a list of (coefficient, reaction id) pairs where coefficient is a float</li>
<li><em>operator</em> is one of = &gt; &lt; &gt;= &lt;=</li>
<li><em>rhs</em> a float</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.buildStoichMatrix">
<code class="descname">buildStoichMatrix</code><span class="sig-paren">(</span><em>matrix_type='numpy'</em>, <em>only_return=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.buildStoichMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.buildStoichMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Build the stoichiometric matrix N and additional constraint matrix CN (if required)</p>
<blockquote>
<div><ul class="simple">
<li><em>matrix_type</em> [default=’numpy’] the type of matrix to use to generate constraints<ul>
<li><em>numpy</em> a NumPy matrix default</li>
<li><em>sympy</em> a SymPy symbolic matrix, if available note the denominator limit can be set in <code class="docutils literal"><span class="pre">CBModel.__CBCONFIG__['SYMPY_DENOM_LIMIT']</span> <span class="pre">=</span> <span class="pre">10**12</span></code></li>
<li><em>scipy_csr</em> create using NumPy but store as SciPy csr_sparse</li>
</ul>
</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><em>only_return</em> [default=False] <strong>IMPORTANT</strong> only returns the stoichiometric matrix and constraint matrix (if required),
does not update the model</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.changeAllFluxBoundsWithValue">
<code class="descname">changeAllFluxBoundsWithValue</code><span class="sig-paren">(</span><em>old</em>, <em>new</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.changeAllFluxBoundsWithValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.changeAllFluxBoundsWithValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Replaces all flux bounds with value “old” with a new value “new”:</p>
<blockquote>
<div><ul class="simple">
<li><em>old</em> value</li>
<li><em>new</em> value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.clone">
<code class="descname">clone</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.clone"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.clone" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a clone of this object.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createCompartment">
<code class="descname">createCompartment</code><span class="sig-paren">(</span><em>cid</em>, <em>name=None</em>, <em>size=1</em>, <em>dimensions=3</em>, <em>volume=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createCompartment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createCompartment" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new compartment and add it to the model if the id does not exist</p>
<ul class="simple">
<li><em>cid</em> compartment id</li>
<li><em>name</em> [None] compartment name</li>
<li><em>size</em> [1] compartment size</li>
<li><em>dimensions</em> [3] compartment size dimensions</li>
<li><em>volume</em> [None] compartment volume</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createGeneAssociationsFromAnnotations">
<code class="descname">createGeneAssociationsFromAnnotations</code><span class="sig-paren">(</span><em>annotation_key='GENE ASSOCIATION'</em>, <em>replace_existing=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createGeneAssociationsFromAnnotations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createGeneAssociationsFromAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Add genes to the model using the definitions stored in the annotation key. If this fails it tries some standard annotation
keys: GENE ASSOCIATION, GENE_ASSOCIATION, gene_association, gene association.</p>
<blockquote>
<div><ul class="simple">
<li><em>annotation_key</em> the annotation dictionary key that holds the gene association for the protein/enzyme</li>
<li><em>replace_existing</em> [default=True] replace existing annotations, otherwise only new ones are added</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createGeneProteinAssociation">
<code class="descname">createGeneProteinAssociation</code><span class="sig-paren">(</span><em>protein</em>, <em>assoc</em>, <em>gid=None</em>, <em>name=None</em>, <em>gene_pattern=None</em>, <em>update_idx=True</em>, <em>altlabels=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createGeneProteinAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createGeneProteinAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Create and add a gene protein relationship to the model, note genes are mapped on protein objects which may or may not be reactions</p>
<blockquote>
<div><ul class="simple">
<li><em>protein</em> in this case the reaction</li>
<li><em>assoc</em> the COBRA style gene protein association</li>
<li><em>gid</em> the unique id</li>
<li><em>name</em> the optional name</li>
<li><em>gene_pattern</em> deprecated, not needed anymore</li>
<li><em>update_idx</em> update the model gene index, not used</li>
<li><em>altlabels</em> [default=None] alternative labels for genes, default uses geneIds</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createGeneProteinAssociationFromTree">
<code class="descname">createGeneProteinAssociationFromTree</code><span class="sig-paren">(</span><em>protein</em>, <em>gprtree</em>, <em>gid=None</em>, <em>name=None</em>, <em>gene_pattern=None</em>, <em>update_idx=True</em>, <em>altlabels=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createGeneProteinAssociationFromTree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createGeneProteinAssociationFromTree" title="Permalink to this definition">¶</a></dt>
<dd><p>Create and add a gene protein relationship to the model, note genes are mapped on protein objects which may or may not be reactions</p>
<blockquote>
<div><ul class="simple">
<li><em>protein</em> in this case the reaction</li>
<li><em>gprtree</em> the CBMPy GPR dictionary tree</li>
<li><em>gid</em> the unique id</li>
<li><em>name</em> the optional name</li>
<li><em>gene_pattern</em> deprecated, not needed anymore</li>
<li><em>update_idx</em> update the model gene index, not used</li>
<li><em>altlabels</em> [default=None] alternative labels for genes, default uses geneIds</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createGroup">
<code class="descname">createGroup</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an empty group with</p>
<blockquote>
<div><ul class="simple">
<li><em>gid</em> the unique group id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createObjectiveFunction">
<code class="descname">createObjectiveFunction</code><span class="sig-paren">(</span><em>rid</em>, <em>coefficient=1</em>, <em>osense='maximize'</em>, <em>active=True</em>, <em>delete_current_obj=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createObjectiveFunction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createObjectiveFunction" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a single variable objective function:</p>
<blockquote>
<div><ul class="simple">
<li><strong>rid</strong> The</li>
<li><strong>coefficient</strong> [default=1]</li>
<li><strong>osense</strong> [default=’maximize’]</li>
<li><strong>active</strong> [default=True]</li>
<li><strong>delete_current_obj</strong> [default=True]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createReaction">
<code class="descname">createReaction</code><span class="sig-paren">(</span><em>rid</em>, <em>name=None</em>, <em>reversible=True</em>, <em>create_default_bounds=True</em>, <em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new blank reaction and add it to the model:</p>
<blockquote>
<div><ul class="simple">
<li><em>id</em> the unique reaction ID</li>
<li><em>name</em> the reaction name</li>
<li><em>reversible</em> [default=True] the reaction reversibility. True is reversible, False is irreversible</li>
<li><em>create_default_bounds</em> create default reaction bounds, irreversible 0 &lt;= J &lt;= INF, reversable -INF &lt;= J &lt;= INF</li>
<li><em>silent</em> [default=False] if enabled this disables the printing of information messages</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createReactionBounds">
<code class="descname">createReactionBounds</code><span class="sig-paren">(</span><em>reaction</em>, <em>lb_value</em>, <em>ub_value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createReactionBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createReactionBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new lower bound for a reaction: value &lt;= reaction</p>
<blockquote>
<div><ul class="simple">
<li><strong>reaction</strong> the reaction id</li>
<li><strong>lb_value</strong> the value of the lower bound</li>
<li><strong>ub_value</strong> the value of the upper bound</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createReactionLowerBound">
<code class="descname">createReactionLowerBound</code><span class="sig-paren">(</span><em>reaction</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createReactionLowerBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createReactionLowerBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new lower bound for a reaction: value &lt;= reaction</p>
<blockquote>
<div><ul class="simple">
<li><strong>reaction</strong> the reaction id</li>
<li><strong>value</strong> the value of the bound</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createReactionReagent">
<code class="descname">createReactionReagent</code><span class="sig-paren">(</span><em>reaction</em>, <em>metabolite</em>, <em>coefficient</em>, <em>silent=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createReactionReagent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createReactionReagent" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a reagent to an existing reaction, both reaction and metabolites must exist</p>
<blockquote>
<div><ul class="simple">
<li><em>reaction</em> a reaction id</li>
<li><em>metabolite</em> a species/metabolite id</li>
<li><em>coefficient</em> the reagent coefficient</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createReactionUpperBound">
<code class="descname">createReactionUpperBound</code><span class="sig-paren">(</span><em>reaction</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createReactionUpperBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createReactionUpperBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new upper bound for a reaction: reaction &lt;= value</p>
<blockquote>
<div><ul class="simple">
<li><strong>reaction</strong> the reaction id</li>
<li><strong>value</strong> the value of the bound</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createSingleGeneEffectMap">
<code class="descname">createSingleGeneEffectMap</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createSingleGeneEffectMap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createSingleGeneEffectMap" title="Permalink to this definition">¶</a></dt>
<dd><p>This takes a model and analyses the logical gene expression patterns. This only needs to be done once,
the result is a dictionary that has boolean effect patterns as keys and the (list of) genes that give rise to
those patterns as values. This map is used by the single gene deletion method for further analysis.</p>
<p>Note this dictionary can also be stored and retrieved separately as long as the model structure is not changed i.e.
the gene associations themselves or order of reactions (stored as the special entry ‘keyJ’).</p>
<p>Stored as self.__single_gene_effect_map__</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.createSpecies">
<code class="descname">createSpecies</code><span class="sig-paren">(</span><em>sid</em>, <em>boundary=False</em>, <em>name=''</em>, <em>value=nan</em>, <em>compartment=None</em>, <em>charge=None</em>, <em>chemFormula=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.createSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.createSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new species and add it to the model:</p>
<blockquote>
<div><ul class="simple">
<li><strong>id</strong> the unique species id</li>
<li><strong>boundary</strong> [default=False] whether the species is a variable (False) or is a boundary parameter (fixed)</li>
<li><strong>name</strong> [default=’‘] the species name</li>
<li><strong>value</strong> [default=nan] the value <em>not currently used</em></li>
<li><strong>compartment</strong> [default=None] the compartment the species is located in</li>
<li><strong>charge</strong> [default=None] the species charge</li>
<li><strong>chemFormula</strong> [default=None] the chemical formula</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteAllFluxBoundsWithValue">
<code class="descname">deleteAllFluxBoundsWithValue</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteAllFluxBoundsWithValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteAllFluxBoundsWithValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete all flux bounds which have a specified value:</p>
<blockquote>
<div><ul class="simple">
<li><em>value</em> the value of the flux bound(s) to delete</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteBoundsForReactionId">
<code class="descname">deleteBoundsForReactionId</code><span class="sig-paren">(</span><em>rid</em>, <em>lower=True</em>, <em>upper=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteBoundsForReactionId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteBoundsForReactionId" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete bounds connected to reaction, rid</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a valid reaction id</li>
<li><em>upper</em> [default=True] delete the upper bound</li>
<li><em>lower</em> [default=True] delete the lower bound</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteGPRAssociation">
<code class="descname">deleteGPRAssociation</code><span class="sig-paren">(</span><em>gprid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteGPRAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteGPRAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete a GPR association with id</p>
<ul class="simple">
<li><em>gprid</em> the GPR association id</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteGene">
<code class="descname">deleteGene</code><span class="sig-paren">(</span><em>gid</em>, <em>also_delete_gpr=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteGene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteGene" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes the gene object with gid. Note if you want to delete a gene by label (locus tag etc)
use the deleteGeneByLabel() function.</p>
<ul class="simple">
<li><em>gid</em> the gene Id</li>
<li><em>also_delete_gpr</em> [default=True] automatically delete GPR’s that contain no gene references</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteGeneByLabel">
<code class="descname">deleteGeneByLabel</code><span class="sig-paren">(</span><em>label</em>, <em>also_delete_gpr=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteGeneByLabel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteGeneByLabel" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes the gene object with label (b2003 etc).</p>
<ul class="simple">
<li><em>label</em> the gene with label to be deleted</li>
<li><em>also_delete_gpr</em> [default=True] automatically delete GPR’s that contain no gene references</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteGroup">
<code class="descname">deleteGroup</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete a group with</p>
<blockquote>
<div><ul class="simple">
<li><em>gid</em> the unique group id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteNonReactingSpecies">
<code class="descname">deleteNonReactingSpecies</code><span class="sig-paren">(</span><em>simulate=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteNonReactingSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteNonReactingSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all species that are not reagents (do not to take part in a reaction).
<em>Warning</em> this deletion is permanent and greedy (not selective). Returns a list of (would be) deleted species</p>
<blockquote>
<div><ul class="simple">
<li><em>simulate</em> [default=True] only return a list of the speciesId’s that would have been deleted if False</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteObjective">
<code class="descname">deleteObjective</code><span class="sig-paren">(</span><em>objective_id</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete objective function:</p>
<blockquote>
<div><em>objective_id</em> the id of the objective function. If objective_id is given  as ‘active’ then the active objective is deleted.</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteReactionAndBounds">
<code class="descname">deleteReactionAndBounds</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteReactionAndBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteReactionAndBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete all reaction and bounds connected to reaction</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a valid reaction id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.deleteSpecies">
<code class="descname">deleteSpecies</code><span class="sig-paren">(</span><em>sid</em>, <em>also_delete=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.deleteSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.deleteSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes a species object with id</p>
<ul class="simple">
<li><em>sid</em> the species id</li>
<li><em>also_delete</em> [default=None] only delete the species</li>
</ul>
<p>– ‘reagents’ delete the species from the reactions it participates in as a <strong>reagent</strong>
– ‘reactions’ deletes the <strong>reactions</strong> that the species participates in</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.emptyUndelete">
<code class="descname">emptyUndelete</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.emptyUndelete"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.emptyUndelete" title="Permalink to this definition">¶</a></dt>
<dd><p>Empties the undelete cache</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.exportFVAdata">
<code class="descname">exportFVAdata</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.exportFVAdata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.exportFVAdata" title="Permalink to this definition">¶</a></dt>
<dd><p>Export the fva data as an array and list of reaction id’s</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.exportUserConstraints">
<code class="descname">exportUserConstraints</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.exportUserConstraints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.exportUserConstraints" title="Permalink to this definition">¶</a></dt>
<dd><p>Exports user constraints in json</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.findFluxesForConnectedSpecies">
<code class="descname">findFluxesForConnectedSpecies</code><span class="sig-paren">(</span><em>metab</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.findFluxesForConnectedSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.findFluxesForConnectedSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent of</p>
<blockquote>
<div><ul class="simple">
<li><em>metab</em> the metabolite name</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getActiveObjective">
<code class="descname">getActiveObjective</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getActiveObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getActiveObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the active objective object.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getActiveObjectiveReactionIds">
<code class="descname">getActiveObjectiveReactionIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getActiveObjectiveReactionIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getActiveObjectiveReactionIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the active objective flux objective reaction id’s</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getActiveObjectiveStoichiometry">
<code class="descname">getActiveObjectiveStoichiometry</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getActiveObjectiveStoichiometry"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getActiveObjectiveStoichiometry" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of (coefficient, flux_objective) tuples</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getAllFluxBounds">
<code class="descname">getAllFluxBounds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getAllFluxBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getAllFluxBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of all flux bounds [id:value]</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getAllGeneActivities">
<code class="descname">getAllGeneActivities</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getAllGeneActivities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getAllGeneActivities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of genes (if defined) and whether they are active or not</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getAllGeneProteinAssociations">
<code class="descname">getAllGeneProteinAssociations</code><span class="sig-paren">(</span><em>use_labels=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getAllGeneProteinAssociations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getAllGeneProteinAssociations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of genes associated with each protein</p>
<blockquote>
<div><ul class="simple">
<li><em>use_labels</em> use V2 gene labels rather than ID’s</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getAllProteinActivities">
<code class="descname">getAllProteinActivities</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getAllProteinActivities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getAllProteinActivities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of reactions (if genes and GPR’s are defined) and whether they are active or not</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getAllProteinGeneAssociations">
<code class="descname">getAllProteinGeneAssociations</code><span class="sig-paren">(</span><em>use_labels=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getAllProteinGeneAssociations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getAllProteinGeneAssociations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of the proteins associated with each gene</p>
<blockquote>
<div><ul class="simple">
<li><em>use_labels</em> use V2 gene labels rather than ID’s</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getBoundarySpeciesIds">
<code class="descname">getBoundarySpeciesIds</code><span class="sig-paren">(</span><em>rid=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getBoundarySpeciesIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getBoundarySpeciesIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all boundary species associated with reaction</p>
<blockquote>
<div><ul class="simple">
<li>rid [default=None] by default return all boundary species in a model, alternatively a string containing a reaction id or list of reaction id’s</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getCompartment">
<code class="descname">getCompartment</code><span class="sig-paren">(</span><em>cid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getCompartment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getCompartment" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a compartment object with <em>cid</em></p>
<blockquote>
<div><ul class="simple">
<li><em>cid</em> compartment ID</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getCompartmentIds">
<code class="descname">getCompartmentIds</code><span class="sig-paren">(</span><em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getCompartmentIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getCompartmentIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of compartment Ids, applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getDescription">
<code class="descname">getDescription</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getDescription"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the model description which was stored in the SBML &lt;notes&gt; field</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getExchangeReactionIds">
<code class="descname">getExchangeReactionIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getExchangeReactionIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getExchangeReactionIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns id’s of reactions where the ‘is_exchange’ attribute set to True. This is by default
reactions that contain a boundary species.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getExchangeReactions">
<code class="descname">getExchangeReactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getExchangeReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getExchangeReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns reaction instances where the ‘is_exchange’ attribute set to True. This is by default
reactions that contain a boundary species.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getFluxBoundByID">
<code class="descname">getFluxBoundByID</code><span class="sig-paren">(</span><em>fid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getFluxBoundByID"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getFluxBoundByID" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a FluxBound with id</p>
<blockquote>
<div><ul class="simple">
<li><em>fid</em> the fluxBound ID</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getFluxBoundByReactionID">
<code class="descname">getFluxBoundByReactionID</code><span class="sig-paren">(</span><em>rid</em>, <em>bound</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getFluxBoundByReactionID"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getFluxBoundByReactionID" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a FluxBound instance</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reaction ID</li>
<li><em>bound</em> the bound: ‘upper’, ‘lower’, ‘equality’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getFluxBoundIds">
<code class="descname">getFluxBoundIds</code><span class="sig-paren">(</span><em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getFluxBoundIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getFluxBoundIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of fluxbound Ids, applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getFluxBoundsByReactionID">
<code class="descname">getFluxBoundsByReactionID</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getFluxBoundsByReactionID"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getFluxBoundsByReactionID" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all FluxBound instances connected to a reactionId as a tuple of valid
(lower, upper, None) or (None, None, equality) or alternatively invalid (lower, upper, equality).</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reaction ID</li>
</ul>
<p><em>under evaluation</em></p>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getFluxesAssociatedWithSpecies">
<code class="descname">getFluxesAssociatedWithSpecies</code><span class="sig-paren">(</span><em>metab</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getFluxesAssociatedWithSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getFluxesAssociatedWithSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent in</p>
<blockquote>
<div><ul class="simple">
<li><em>metab</em> the metabolite name</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGPRIdAssociatedWithGeneId">
<code class="descname">getGPRIdAssociatedWithGeneId</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGPRIdAssociatedWithGeneId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGPRIdAssociatedWithGeneId" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the GPR(s) associated with the gene id:</p>
<blockquote>
<div><ul class="simple">
<li><em>gid</em> a gene id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGPRIdAssociatedWithGeneLabel">
<code class="descname">getGPRIdAssociatedWithGeneLabel</code><span class="sig-paren">(</span><em>label</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGPRIdAssociatedWithGeneLabel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGPRIdAssociatedWithGeneLabel" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the GPR Id’s associated with the gene label:</p>
<blockquote>
<div><ul class="simple">
<li><em>label</em> a gene label</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGPRassociation">
<code class="descname">getGPRassociation</code><span class="sig-paren">(</span><em>gpr_id</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGPRassociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGPRassociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a gene protein association object that has the identifier:</p>
<blockquote>
<div><ul class="simple">
<li><em>gpr_id</em> the gene protein identifier</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGPRforReaction">
<code class="descname">getGPRforReaction</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGPRforReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGPRforReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the GPR associated with the reaction id:</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a reaction id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGPRforReactionAsDict">
<code class="descname">getGPRforReactionAsDict</code><span class="sig-paren">(</span><em>rid</em>, <em>useweakref=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGPRforReactionAsDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGPRforReactionAsDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the GPR associated with the reaction id as a nested dictionary structure:</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a reaction id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGene">
<code class="descname">getGene</code><span class="sig-paren">(</span><em>g_id</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGene" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a gene object that has the identifier:</p>
<blockquote>
<div><ul class="simple">
<li><em>gid</em> the gene identifier</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGeneByLabel">
<code class="descname">getGeneByLabel</code><span class="sig-paren">(</span><em>label</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGeneByLabel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGeneByLabel" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a gene label return the corresponding Gene object</p>
<blockquote>
<div><ul class="simple">
<li><em>label</em></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGeneIdFromLabel">
<code class="descname">getGeneIdFromLabel</code><span class="sig-paren">(</span><em>label</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGeneIdFromLabel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGeneIdFromLabel" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a gene label it returns the corresponding Gene id or None</p>
<blockquote>
<div><ul class="simple">
<li><em>label</em></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGeneIds">
<code class="descname">getGeneIds</code><span class="sig-paren">(</span><em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGeneIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGeneIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of gene Ids, applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGeneLabels">
<code class="descname">getGeneLabels</code><span class="sig-paren">(</span><em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGeneLabels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGeneLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of gene labels (locus tags), applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGroup">
<code class="descname">getGroup</code><span class="sig-paren">(</span><em>gid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a group with</p>
<blockquote>
<div><ul class="simple">
<li><em>gid</em> the unique group id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGroupIds">
<code class="descname">getGroupIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGroupIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGroupIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete all group ids</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getGroupMembership">
<code class="descname">getGroupMembership</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getGroupMembership"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getGroupMembership" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns group membership of items in groups. Returns {object_id: [‘group_id1’, ‘group_id2’]}</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getIrreversibleReactionIds">
<code class="descname">getIrreversibleReactionIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getIrreversibleReactionIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getIrreversibleReactionIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of irreversible reaction Id’s</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getModelCreators">
<code class="descname">getModelCreators</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getModelCreators"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getModelCreators" title="Permalink to this definition">¶</a></dt>
<dd><p>Return model creator information</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getObjFuncValue">
<code class="descname">getObjFuncValue</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getObjFuncValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getObjFuncValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the objective function value</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getObjectiveIds">
<code class="descname">getObjectiveIds</code><span class="sig-paren">(</span><em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getObjectiveIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getObjectiveIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of objective function Ids, applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getParameter">
<code class="descname">getParameter</code><span class="sig-paren">(</span><em>pid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getParameter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getParameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a parameter object with pid</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReaction">
<code class="descname">getReaction</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a reaction object with <em>id</em></p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> reaction ID</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionActivity">
<code class="descname">getReactionActivity</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionActivity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionActivity" title="Permalink to this definition">¶</a></dt>
<dd><p>If there is a GPR and genes associated with the reaction ID then return either active=True or inactive=False
Note if there is no gene associated information then this will return active.</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a reaction id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionBounds">
<code class="descname">getReactionBounds</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the bounds of a reaction, returns a tuple of rid, lowerbound value, upperbound value and equality value (None means bound does not exist).</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reaction ID</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionIds">
<code class="descname">getReactionIds</code><span class="sig-paren">(</span><em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of reaction Ids, applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionIdsAssociatedWithSpecies">
<code class="descname">getReactionIdsAssociatedWithSpecies</code><span class="sig-paren">(</span><em>metab</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionIdsAssociatedWithSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionIdsAssociatedWithSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent in</p>
<blockquote>
<div><ul class="simple">
<li><em>metab</em> the metabolite name</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionLowerBound">
<code class="descname">getReactionLowerBound</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionLowerBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionLowerBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the lower bound of a reaction (it it exists) or None</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reaction ID</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionNames">
<code class="descname">getReactionNames</code><span class="sig-paren">(</span><em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionNames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionNames" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of reaction names, applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the name</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionUpperBound">
<code class="descname">getReactionUpperBound</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionUpperBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionUpperBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the upper bound of a reaction (it it exists) or None</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reaction ID</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReactionValues">
<code class="descname">getReactionValues</code><span class="sig-paren">(</span><em>only_exchange=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReactionValues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReactionValues" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of ReactionID : ReactionValue pairs:</p>
<blockquote>
<div><ul class="simple">
<li><em>only_exchange</em> [default=False] only return the reactions labelled as exchange</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getReversibleReactionIds">
<code class="descname">getReversibleReactionIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getReversibleReactionIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getReversibleReactionIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of reversible reaction Id’s</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getSolutionVector">
<code class="descname">getSolutionVector</code><span class="sig-paren">(</span><em>names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getSolutionVector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getSolutionVector" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a vector of solution values</p>
<blockquote>
<div><ul class="simple">
<li><em>names</em> [default=False] if True return a solution vector and list of names</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getSpecies">
<code class="descname">getSpecies</code><span class="sig-paren">(</span><em>sid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a species object with <em>sid</em></p>
<blockquote>
<div><ul class="simple">
<li><em>sid</em> a specied ID</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.getSpeciesIds">
<code class="descname">getSpeciesIds</code><span class="sig-paren">(</span><em>substring=None</em>, <em>non_boundary=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.getSpeciesIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.getSpeciesIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of species Ids, applies a substring search if substring is defined</p>
<blockquote>
<div><ul class="simple">
<li><em>substring</em> search for this pattern anywhere in the id</li>
<li><em>non_boundary</em> [default=False] return only non-boundary species, i.e., variable metabolites that appear in the stoichiometric matrix. The default is to return all metabolites boundary and variable.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.importUserConstraints">
<code class="descname">importUserConstraints</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.importUserConstraints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.importUserConstraints" title="Permalink to this definition">¶</a></dt>
<dd><p>Exports user constraints in json</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.renameObjectIds">
<code class="descname">renameObjectIds</code><span class="sig-paren">(</span><em>prefix=None</em>, <em>suffix=None</em>, <em>target='all'</em>, <em>ignore=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.renameObjectIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.renameObjectIds" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is designed for target=”all” other use may result in inconsistent models. Update: “species” and “reactions”
should also work as advertised, please check results.</p>
<blockquote>
<div><ul class="simple">
<li><em>prefix</em> [None] if supplied add as a prefix</li>
<li><em>suffix</em> [None] if supplied add as a suffix</li>
<li><em>target</em> [‘all’] specify what class of objects to rename</li>
</ul>
<blockquote>
<div><ul class="simple">
<li>‘species’</li>
<li>‘reactions’</li>
<li>‘bounds’</li>
<li>‘objectives’</li>
<li>‘all’</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><em>ignore</em> [default=None] a list of id’s to ignore</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.resetAllGenes">
<code class="descname">resetAllGenes</code><span class="sig-paren">(</span><em>update_reactions=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.resetAllGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.resetAllGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets all genes to their default activity state (normally on)</p>
<ul class="simple">
<li><em>update_reactions</em> [default=False] update the associated reactions fluxbounds from the gene deletion bounds if they exist</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.resetAllInactiveGPRBounds">
<code class="descname">resetAllInactiveGPRBounds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.resetAllInactiveGPRBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.resetAllInactiveGPRBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets all reaction bounds modified by the <code class="docutils literal"><span class="pre">cmod.setAllInactiveGeneReactionBounds()</span></code> method to their previous values</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.serialize">
<code class="descname">serialize</code><span class="sig-paren">(</span><em>protocol=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.serialize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.serialize" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize object, returns a string by default</p>
<blockquote>
<div><ul class="simple">
<li><dl class="first docutils">
<dt><em>protocol</em> [default=0] serialize to a string or binary if required,</dt>
<dd>see pickle module documentation for details</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p># overloaded in CBModel</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.serializeToDisk">
<code class="descname">serializeToDisk</code><span class="sig-paren">(</span><em>filename</em>, <em>protocol=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.serializeToDisk"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.serializeToDisk" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize to disk using pickle protocol:</p>
<blockquote>
<div><ul class="simple">
<li><em>filename</em> the name of the output file</li>
<li><dl class="first docutils">
<dt><em>protocol</em> [default=2] serialize to a string or binary if required,</dt>
<dd>see pickle module documentation for details</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p># overloaded in CBModel</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setAllFluxBounds">
<code class="descname">setAllFluxBounds</code><span class="sig-paren">(</span><em>bounds</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setAllFluxBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setAllFluxBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>DEPRECATED! use setFluxBoundsFromDict()</p>
<p>Sets all the fluxbounds present in bounds</p>
<blockquote>
<div><ul class="simple">
<li><em>bounds</em> a dictionary of [fluxbound_id : value] pairs (not per reaction!!!)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setAllInactiveGPRBounds">
<code class="descname">setAllInactiveGPRBounds</code><span class="sig-paren">(</span><em>lower=0.0</em>, <em>upper=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setAllInactiveGPRBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setAllInactiveGPRBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Set all reactions that are inactive (as determined by gene and gpr evaluation) to bounds:</p>
<blockquote>
<div><ul class="simple">
<li><em>lower</em> [default=0.0] the new lower bound</li>
<li><em>upper</em> [default=0.0] the new upper bound</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setAllProteinActivities">
<code class="descname">setAllProteinActivities</code><span class="sig-paren">(</span><em>activites</em>, <em>lower=0.0</em>, <em>upper=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setAllProteinActivities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setAllProteinActivities" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dictionary of activities [rid : boolean] pairs set all the corresponding reactions:</p>
<blockquote>
<div><ul class="simple">
<li><em>activities</em> a dictionary of [rid : boolean] pairs</li>
<li><em>lower</em> [default=0.0] the lower bound of the deactivated flux</li>
<li><em>upper</em> [default=0.0] the upper bound of the deactivated flux</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setBoundValueByName">
<code class="descname">setBoundValueByName</code><span class="sig-paren">(</span><em>rid</em>, <em>value</em>, <em>bound</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setBoundValueByName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setBoundValueByName" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated use setReactionBound</p>
<dl class="docutils">
<dt>Set a reaction bound</dt>
<dd><ul class="first last simple">
<li><em>rid</em> the reactions id</li>
<li><em>value</em> the new value</li>
<li><em>bound</em> this is either ‘lower’ or ‘upper’</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setCreatedDate">
<code class="descname">setCreatedDate</code><span class="sig-paren">(</span><em>date=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setCreatedDate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setCreatedDate" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the model created date tuple(year, month, day, hour, minute, second)</p>
<ul class="simple">
<li><em>date</em> [default=None] default is now (automatic) otherwise (year, month, day, hour, minute, second) e.g. (2012, 09, 24, 13, 34, 00)</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setDescription">
<code class="descname">setDescription</code><span class="sig-paren">(</span><em>html</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setDescription"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the model description which translates into the SBML &lt;notes&gt; field.</p>
<blockquote>
<div><ul class="simple">
<li><em>html</em> any valid html or the empty string to clear ‘’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setFluxBoundsFromDict">
<code class="descname">setFluxBoundsFromDict</code><span class="sig-paren">(</span><em>bounds</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setFluxBoundsFromDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setFluxBoundsFromDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets all the fluxbounds present in bounds</p>
<blockquote>
<div><ul class="simple">
<li><em>bounds</em> a dictionary of [fluxbound_id : value] pairs (not per reaction!!!)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setGeneActive">
<code class="descname">setGeneActive</code><span class="sig-paren">(</span><em>g_id</em>, <em>update_reactions=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setGeneActive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setGeneActive" title="Permalink to this definition">¶</a></dt>
<dd><p>Effectively restores a gene by setting it’s active flag</p>
<blockquote>
<div><ul class="simple">
<li><em>g_id</em> a gene ID</li>
<li><em>update_reactions</em> [default=False] update the associated reactions fluxbounds from the gene deletion bounds if they exist</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setGeneInactive">
<code class="descname">setGeneInactive</code><span class="sig-paren">(</span><em>g_id</em>, <em>update_reactions=False</em>, <em>lower=0.0</em>, <em>upper=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setGeneInactive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setGeneInactive" title="Permalink to this definition">¶</a></dt>
<dd><p>Effectively deletes a gene by setting it’s inactive flag while optionally updating the GPR associated reactions</p>
<blockquote>
<div><ul class="simple">
<li><em>g_id</em> a gene ID</li>
<li><em>update_reactions</em> [default=False] update the associated reactions fluxbounds</li>
<li><em>lower</em> [default=0.0] the deactivated reaction lower bound</li>
<li><em>upper</em> [default=0.0] the deactivated reaction upper bound</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setModifiedDate">
<code class="descname">setModifiedDate</code><span class="sig-paren">(</span><em>date=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setModifiedDate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setModifiedDate" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the model modification date: tuple(year, month, day, hour, minute, second)</p>
<ul class="simple">
<li><em>date</em> [default=None] default is now (automatic) otherwise (year, month, day, hour, minute, second) e.g. (2012, 09, 24, 13, 34, 00)</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setObjectiveFlux">
<code class="descname">setObjectiveFlux</code><span class="sig-paren">(</span><em>rid</em>, <em>coefficient=1</em>, <em>osense='maximize'</em>, <em>delete_objflx=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setObjectiveFlux"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setObjectiveFlux" title="Permalink to this definition">¶</a></dt>
<dd><p>Set single target reaction flux for the current active objective function.</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a string containing a reaction id</li>
<li><em>coefficient</em> [default=1] an objective flux coefficient</li>
<li><em>osense</em> the optimization sense must be <strong>maximize</strong> or <strong>minimize</strong></li>
<li><em>delete_objflx</em> [default=True] delete all existing fluxObjectives in the active objective function</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setPrefix">
<code class="descname">setPrefix</code><span class="sig-paren">(</span><em>prefix</em>, <em>target</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setPrefix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setPrefix" title="Permalink to this definition">¶</a></dt>
<dd><p>This is alpha stuff, target can be:</p>
<blockquote>
<div><ul class="simple">
<li>‘species’</li>
<li>‘reactions’</li>
<li>‘constraints’</li>
<li>‘objectives’</li>
<li>‘all’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setReactionBound">
<code class="descname">setReactionBound</code><span class="sig-paren">(</span><em>rid</em>, <em>value</em>, <em>bound</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setReactionBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setReactionBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a reaction bound</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reactions id</li>
<li><em>value</em> the new value</li>
<li><em>bound</em> this is either ‘lower’ or ‘upper’, or ‘equal’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setReactionBounds">
<code class="descname">setReactionBounds</code><span class="sig-paren">(</span><em>rid</em>, <em>lower</em>, <em>upper</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setReactionBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setReactionBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Set both the upper and lower bound of a reaction:</p>
<ul class="simple">
<li><em>rid</em> the good old reaction id</li>
<li><em>lower</em> the lower bound value</li>
<li><em>upper</em> the upper bound value</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setReactionLowerBound">
<code class="descname">setReactionLowerBound</code><span class="sig-paren">(</span><em>rid</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setReactionLowerBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setReactionLowerBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a reactions lower bound (if it exists)</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reactions id</li>
<li><em>value</em> the new value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setReactionUpperBound">
<code class="descname">setReactionUpperBound</code><span class="sig-paren">(</span><em>rid</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setReactionUpperBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setReactionUpperBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a reactions upper bound (if it exists)</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> the reaction id</li>
<li><em>value</em> the new value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.setSuffix">
<code class="descname">setSuffix</code><span class="sig-paren">(</span><em>suffix</em>, <em>target</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.setSuffix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.setSuffix" title="Permalink to this definition">¶</a></dt>
<dd><p>This is alpha stuff, target can be:</p>
<blockquote>
<div><ul class="simple">
<li>‘species’</li>
<li>‘reactions’</li>
<li>‘constraints’</li>
<li>‘objectives’</li>
<li>‘all’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.sortReactionsById">
<code class="descname">sortReactionsById</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.sortReactionsById"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.sortReactionsById" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorts the reactions by Reaction.id uses the python string sort</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.sortSpeciesById">
<code class="descname">sortSpeciesById</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.sortSpeciesById"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.sortSpeciesById" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorts the reaction list by Reaction.id uses the python string sort</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.splitEqualityFluxBounds">
<code class="descname">splitEqualityFluxBounds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.splitEqualityFluxBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.splitEqualityFluxBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits any equalit flux bounds into lower and upper bounds.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.testGeneProteinAssociations">
<code class="descname">testGeneProteinAssociations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.testGeneProteinAssociations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.testGeneProteinAssociations" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will test the GeneProtein associations and return a list of protein, association pairs</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Model.updateNetwork">
<code class="descname">updateNetwork</code><span class="sig-paren">(</span><em>lower=0.0</em>, <em>upper=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Model.updateNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Model.updateNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the reaction network based on gene activity. If reaction is deactivated then lower and upper bounds are used</p>
<blockquote>
<div><ul class="simple">
<li><em>lower</em> [default=0.0] deactivated lower bound</li>
<li><em>upper</em> [default=0.0] deactivated upper bound</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Objective">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Objective</code><span class="sig-paren">(</span><em>pid</em>, <em>operation</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective" title="Permalink to this definition">¶</a></dt>
<dd><p>An objective function</p>
<dl class="method">
<dt id="cbmpy.CBModel.Objective.addFluxObjective">
<code class="descname">addFluxObjective</code><span class="sig-paren">(</span><em>fobj</em>, <em>override=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.addFluxObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.addFluxObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a FluxObjective instance to the Objective</p>
<blockquote>
<div><ul class="simple">
<li><em>fobj</em> the FluxObjective object</li>
<li><em>override</em> [default=False] override pushing the global id map, this should never be used</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.createFluxObjectives">
<code class="descname">createFluxObjectives</code><span class="sig-paren">(</span><em>fluxlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.createFluxObjectives"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.createFluxObjectives" title="Permalink to this definition">¶</a></dt>
<dd><p>Create and add flux objective objects to this objective function.</p>
<blockquote>
<div><ul class="simple">
<li><em>fluxlist</em> a list of one or more (‘coefficient’, ‘rid’) pairs</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.deleteAllFluxObjectives">
<code class="descname">deleteAllFluxObjectives</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.deleteAllFluxObjectives"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.deleteAllFluxObjectives" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete all flux objectives</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getFluxObjective">
<code class="descname">getFluxObjective</code><span class="sig-paren">(</span><em>foid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getFluxObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getFluxObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the flux objective with id.</p>
<blockquote>
<div><ul class="simple">
<li><em>foid</em> the flux objective id returns either an object or a list if there are multiply defined flux objectives</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getFluxObjectiveData">
<code class="descname">getFluxObjectiveData</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getFluxObjectiveData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getFluxObjectiveData" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of ObjectiveFunction components as (coefficient, flux) pairs</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getFluxObjectiveForReaction">
<code class="descname">getFluxObjectiveForReaction</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getFluxObjectiveForReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getFluxObjectiveForReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the FluxObjective associated with the suplied rid. If there is more than fluxObjective associated with a reaction (illegal)
then a list of fluxObjectives is returned.</p>
<blockquote>
<div><em>rid</em> a reaction id</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getFluxObjectiveIDs">
<code class="descname">getFluxObjectiveIDs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getFluxObjectiveIDs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getFluxObjectiveIDs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of ObjectiveFlux ids, for the reaction id’s use <em>getFluxObjectiveReactions()</em>
or for coefficient, fluxobjective pairs use <em>getFluxObjectiveData()</em></p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getFluxObjectiveReactions">
<code class="descname">getFluxObjectiveReactions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getFluxObjectiveReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getFluxObjectiveReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of reactions that are used as FluxObjectives</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getFluxObjectives">
<code class="descname">getFluxObjectives</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getFluxObjectives"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getFluxObjectives" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the list of FluxObjective objects.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getOperation">
<code class="descname">getOperation</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getOperation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getOperation" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the operation or sense of the objective</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.getValue">
<code class="descname">getValue</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.getValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.getValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the current value of the attribute (input/solution)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.setOperation">
<code class="descname">setOperation</code><span class="sig-paren">(</span><em>operation</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.setOperation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.setOperation" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the objective operation (sense)</p>
<blockquote>
<div><ul class="simple">
<li><em>operation</em> [default=’maximize’] one of ‘maximize’, ‘maximise’, ‘max’, ‘minimize’, ‘minimise’, ‘min’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Objective.setValue">
<code class="descname">setValue</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Objective.setValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Objective.setValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the attribute ‘’value’‘</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Parameter">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Parameter</code><span class="sig-paren">(</span><em>pid</em>, <em>value</em>, <em>name=None</em>, <em>constant=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Parameter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Parameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Holds parameter information</p>
<dl class="method">
<dt id="cbmpy.CBModel.Parameter.addAssociation">
<code class="descname">addAssociation</code><span class="sig-paren">(</span><em>assoc</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Parameter.addAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Parameter.addAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an object ID to associate with this object</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Parameter.deleteAssociation">
<code class="descname">deleteAssociation</code><span class="sig-paren">(</span><em>assoc</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Parameter.deleteAssociation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Parameter.deleteAssociation" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete the object id associated with this object</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Parameter.getAssociations">
<code class="descname">getAssociations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Parameter.getAssociations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Parameter.getAssociations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Object ID’s associated with this parameter</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Parameter.getValue">
<code class="descname">getValue</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Parameter.getValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Parameter.getValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the current value of the attribute (input/solution)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Parameter.setValue">
<code class="descname">setValue</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Parameter.setValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Parameter.setValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the attribute ‘’value’‘</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Reaction">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Reaction</code><span class="sig-paren">(</span><em>pid</em>, <em>name=None</em>, <em>reversible=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Holds reaction information</p>
<dl class="method">
<dt id="cbmpy.CBModel.Reaction.addReagent">
<code class="descname">addReagent</code><span class="sig-paren">(</span><em>reag</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.addReagent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.addReagent" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an instantiated Reagent object to the reaction</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.changeId">
<code class="descname">changeId</code><span class="sig-paren">(</span><em>pid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.changeId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.changeId" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the Id of the reaction and updates associated FluxBounds</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.createReagent">
<code class="descname">createReagent</code><span class="sig-paren">(</span><em>metabolite</em>, <em>coefficient</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.createReagent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.createReagent" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new reagent and add it to the reaction:</p>
<blockquote>
<div><ul class="simple">
<li><strong>metabolite</strong> the metabolite name</li>
<li><strong>coefficient</strong> the</li>
</ul>
<p class="attribution">&mdash;negative coefficient is a substrate
– positive coefficient is a product</p>
</div></blockquote>
<p>Will fail if a species reference already exists</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.deactivateReaction">
<code class="descname">deactivateReaction</code><span class="sig-paren">(</span><em>lower=0.0</em>, <em>upper=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.deactivateReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.deactivateReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Deactivates a reaction by setting its bounds to lower and upper. Restore with reactivateReaction()</p>
<blockquote>
<div><ul class="simple">
<li><em>lower</em> [default=0.0] bound</li>
<li><em>upper</em> [default=0.0] bound</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.deleteReagentWithSpeciesRef">
<code class="descname">deleteReagentWithSpeciesRef</code><span class="sig-paren">(</span><em>sid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.deleteReagentWithSpeciesRef"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.deleteReagentWithSpeciesRef" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete a reagent (or reagents) that refers to the species id:</p>
<blockquote>
<div><ul class="simple">
<li><em>sid</em> a species/metabolite id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getEquation">
<code class="descname">getEquation</code><span class="sig-paren">(</span><em>reverse_symb='='</em>, <em>irreverse_symb='&gt;'</em>, <em>use_names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getEquation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getEquation" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a pretty printed string containing the reaction equation</p>
<blockquote>
<div><ul class="simple">
<li><em>reverse_symb</em> [default = ‘=’] the symbol to use for reversible reactions</li>
<li><em>irreverse_symb</em> [default = ‘&gt;’] the symbol to use for irreversible reactions</li>
<li><em>use_names</em> [defualt = False] use species names rather than id’s</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getFVAdata">
<code class="descname">getFVAdata</code><span class="sig-paren">(</span><em>roundnum=None</em>, <em>silent=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getFVAdata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getFVAdata" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the data generated by CBSolver.FluxVariabilityAnalysis() for this reaction as a tuple of
(Flux, FVAmin, FVAmax, span) where span is abs(FVAmax - FVAmin). FVAmin or FVAmax is None this indicates no solution
to that particular optimization (infeasible).</p>
<blockquote>
<div><ul class="simple">
<li><em>roundnum</em> [default=None] the integer number of roundoff decimals the default is no rounding</li>
<li><em>silent</em> [default=True] supress output to stdout</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getLowerBound">
<code class="descname">getLowerBound</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getLowerBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getLowerBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the value of the reactions lower bound</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getProductIds">
<code class="descname">getProductIds</code><span class="sig-paren">(</span><em>use_names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getProductIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getProductIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of the reaction products, species identifiers</p>
<ul class="simple">
<li><em>use_names</em> [default = False] use species names rather than id’s</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getReagent">
<code class="descname">getReagent</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getReagent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getReagent" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the one or more reagent objects which have <em>rid</em>:</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a reagent <em>rid</em></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getReagentObjIds">
<code class="descname">getReagentObjIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getReagentObjIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getReagentObjIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of the reagent id’s. For the name of the reagents/metabolites use <em>&lt;reaction&gt;.getSpeciesIds()</em></p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getReagentRefs">
<code class="descname">getReagentRefs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getReagentRefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getReagentRefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of the reagents/metabolites</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getReagentWithSpeciesRef">
<code class="descname">getReagentWithSpeciesRef</code><span class="sig-paren">(</span><em>sid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getReagentWithSpeciesRef"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getReagentWithSpeciesRef" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the reagent object which refers to the <em>species</em> id. If there are multiple reagents that
refer to the same species a list is returned.</p>
<blockquote>
<div><ul class="simple">
<li><em>sid</em> the species/metabolite id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getSpeciesIds">
<code class="descname">getSpeciesIds</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getSpeciesIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getSpeciesIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of the reagents/metabolites</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getSpeciesObj">
<code class="descname">getSpeciesObj</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getSpeciesObj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getSpeciesObj" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of the species objects that are reagents</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getStoichiometry">
<code class="descname">getStoichiometry</code><span class="sig-paren">(</span><em>use_names=False</em>, <em>altout=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getStoichiometry"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getStoichiometry" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of (coefficient, species) pairs for this reaction</p>
<ul class="simple">
<li><em>use_names</em> [default = False] use species names rather than id’s</li>
<li><em>altout</em> [default = False] returns a dictionary [DEPRECATED]</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getSubstrateIds">
<code class="descname">getSubstrateIds</code><span class="sig-paren">(</span><em>use_names=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getSubstrateIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getSubstrateIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of the reaction substrates, species identifiers</p>
<ul class="simple">
<li><em>use_names</em> [defualt = False] use species names rather than id’s</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getUpperBound">
<code class="descname">getUpperBound</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getUpperBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getUpperBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the value of the reactions upper bound</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.getValue">
<code class="descname">getValue</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.getValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.getValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the current value of the flux.</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.reactivateReaction">
<code class="descname">reactivateReaction</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.reactivateReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.reactivateReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Activates a reaction deactivated with deactivateReaction</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.setId">
<code class="descname">setId</code><span class="sig-paren">(</span><em>fid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.setId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.setId" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object Id</p>
<blockquote>
<div><ul class="simple">
<li><em>fid</em> a valid c variable style id string</li>
</ul>
<p>Reimplements &#64;FBase.setId()</p>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.setLowerBound">
<code class="descname">setLowerBound</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.setLowerBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.setLowerBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the value of the reactions lower bound</p>
<blockquote>
<div><ul class="simple">
<li><em>value</em> a floating point value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.setStoichCoefficient">
<code class="descname">setStoichCoefficient</code><span class="sig-paren">(</span><em>sid</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.setStoichCoefficient"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.setStoichCoefficient" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the stoichiometric coefficient of a reagent that refers to a metabolite. Note <em>negative coefficients</em> are <em>substrates</em>
while <em>positive</em> ones are <em>products</em>. At this point zero coefficients are not allowed</p>
<ul class="simple">
<li><em>sid</em> the species/metabolite id</li>
<li><em>value</em> a floating point value != 0</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.setUpperBound">
<code class="descname">setUpperBound</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.setUpperBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.setUpperBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the value of the reactions upper bound</p>
<blockquote>
<div><ul class="simple">
<li><em>value</em> a floating point value</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reaction.setValue">
<code class="descname">setValue</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reaction.setValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reaction.setValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the attribute <em>value</em> in this case the flux.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Reagent">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Reagent</code><span class="sig-paren">(</span><em>reid</em>, <em>species_ref</em>, <em>coef</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reagent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reagent" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Has a reactive species id and stoichiometric coefficient:</dt>
<dd><ul class="first last simple">
<li>negative = substrate</li>
<li>positive = product</li>
<li>species_ref a reference to a species obj</li>
</ul>
</dd>
</dl>
<dl class="method">
<dt id="cbmpy.CBModel.Reagent.getCoefficient">
<code class="descname">getCoefficient</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reagent.getCoefficient"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reagent.getCoefficient" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reagent coefficient</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reagent.getRole">
<code class="descname">getRole</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reagent.getRole"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reagent.getRole" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reagents role, “substrate”, “product” or None</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reagent.getSpecies">
<code class="descname">getSpecies</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reagent.getSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reagent.getSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the metabolite/species that the reagent reference refers to</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reagent.setCoefficient">
<code class="descname">setCoefficient</code><span class="sig-paren">(</span><em>coef</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reagent.setCoefficient"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reagent.setCoefficient" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the reagent coefficient and role, negative coefficients are substrates and positive ones are products</p>
<blockquote>
<div><ul class="simple">
<li><em>coeff</em> the new coefficient</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Reagent.setSpecies">
<code class="descname">setSpecies</code><span class="sig-paren">(</span><em>spe</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Reagent.setSpecies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Reagent.setSpecies" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the metabolite/species that the reagent reference refers to</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBModel.Species">
<em class="property">class </em><code class="descclassname">cbmpy.CBModel.</code><code class="descname">Species</code><span class="sig-paren">(</span><em>pid</em>, <em>boundary=False</em>, <em>name=None</em>, <em>value=nan</em>, <em>compartment=None</em>, <em>charge=None</em>, <em>chemFormula=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species" title="Permalink to this definition">¶</a></dt>
<dd><p>Holds species/metabolite information</p>
<dl class="method">
<dt id="cbmpy.CBModel.Species.getCharge">
<code class="descname">getCharge</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.getCharge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.getCharge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the species charge</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.getChemFormula">
<code class="descname">getChemFormula</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.getChemFormula"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.getChemFormula" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the species chemical formula</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.getReagentOf">
<code class="descname">getReagentOf</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.getReagentOf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.getReagentOf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of reaction id’s that this metabolite occurs in</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.getValue">
<code class="descname">getValue</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.getValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.getValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the current value of the attribute (input/solution)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.isReagentOf">
<code class="descname">isReagentOf</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.isReagentOf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.isReagentOf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dynamically generated list of reactions that this species occurs as a reagent</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.setBoundary">
<code class="descname">setBoundary</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.setBoundary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.setBoundary" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the species so it is a boundary metabolite or fixed which does not occur in the stoichiometric matrix N</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.setCharge">
<code class="descname">setCharge</code><span class="sig-paren">(</span><em>charge</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.setCharge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.setCharge" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the species charge:</p>
<blockquote>
<div><ul class="simple">
<li><em>charge</em> a signed double but generally a signed int is used</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.setChemFormula">
<code class="descname">setChemFormula</code><span class="sig-paren">(</span><em>cf</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.setChemFormula"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.setChemFormula" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the species chemical formula</p>
<blockquote>
<div><ul class="simple">
<li><em>cf</em> a chemical formula e.g. CH3NO2</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.setId">
<code class="descname">setId</code><span class="sig-paren">(</span><em>fid</em>, <em>allow_rename=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.setId"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.setId" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the object Id</p>
<blockquote>
<div><ul class="simple">
<li><em>fid</em> a valid c variable style id string</li>
</ul>
<p>Reimplements &#64;FBase.setId()</p>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.setReagentOf">
<code class="descname">setReagentOf</code><span class="sig-paren">(</span><em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.setReagentOf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.setReagentOf" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds the supplied reaction id to the reagent_of list (if it isn’t one already)</p>
<blockquote>
<div><ul class="simple">
<li><em>rid</em> a valid reaction id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.setValue">
<code class="descname">setValue</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.setValue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.setValue" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the attribute ‘’value’‘</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBModel.Species.unsetBoundary">
<code class="descname">unsetBoundary</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBModel.html#Species.unsetBoundary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBModel.Species.unsetBoundary" title="Permalink to this definition">¶</a></dt>
<dd><p>Unsets the species boundary attribute so that the metabolite is free and therefore occurs in the stoichiometric matrix N</p>
</dd></dl>

</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBModelTools"></span><div class="section" id="cbmpy-cbmodeltools-module">
<h2>CBMPy: CBModelTools module<a class="headerlink" href="#cbmpy-cbmodeltools-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBModelTools.py 575 2017-04-13 12:18:44Z bgoli $)</p>
</div>
<span class="target" id="module-cbmpy.CBMultiCore"></span><div class="section" id="cbmpy-cbmulticore-module">
<h2>CBMPy: CBMultiCore module<a class="headerlink" href="#cbmpy-cbmulticore-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBMultiCore.py 575 2017-04-13 12:18:44Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBMultiCore.grouper">
<code class="descclassname">cbmpy.CBMultiCore.</code><code class="descname">grouper</code><span class="sig-paren">(</span><em>3</em>, <em>'abcdefg'</em>, <em>'x') --&gt; ('a'</em>, <em>'b'</em>, <em>'c')</em>, <em>('d'</em>, <em>'e'</em>, <em>'f')</em>, <em>('g'</em>, <em>'x'</em>, <em>'x'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBMultiCore.html#grouper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBMultiCore.grouper" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="cbmpy.CBMultiCore.runMultiCoreFVA">
<code class="descclassname">cbmpy.CBMultiCore.</code><code class="descname">runMultiCoreFVA</code><span class="sig-paren">(</span><em>fba</em>, <em>selected_reactions=None</em>, <em>pre_opt=True</em>, <em>tol=None</em>, <em>objF2constr=True</em>, <em>rhs_sense='lower'</em>, <em>optPercentage=100.0</em>, <em>work_dir=None</em>, <em>quiet=True</em>, <em>debug=False</em>, <em>oldlpgen=False</em>, <em>markupmodel=True</em>, <em>procs=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBMultiCore.html#runMultiCoreFVA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBMultiCore.runMultiCoreFVA" title="Permalink to this definition">¶</a></dt>
<dd><p>Run a multicore FVA where:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> is an fba model instance</li>
<li><em>procs</em> [default=2] number of processing threads (optimum seems to be about the number of physical cores)</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBMultiEnv"></span><div class="section" id="cbmpy-cbmultienv-module">
<h2>CBMPy: CBMultiEnv module<a class="headerlink" href="#cbmpy-cbmultienv-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBMultiEnv.py 575 2017-04-13 12:18:44Z bgoli $)</p>
</div>
<span class="target" id="module-cbmpy.CBNetDB"></span><div class="section" id="cbmpy-cbnetdb-module">
<h2>CBMPy: CBNetDB module<a class="headerlink" href="#cbmpy-cbnetdb-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBNetDB.py 629 2017-10-24 22:01:14Z bgoli $)</p>
<dl class="class">
<dt id="cbmpy.CBNetDB.DBTools">
<em class="property">class </em><code class="descclassname">cbmpy.CBNetDB.</code><code class="descname">DBTools</code><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools" title="Permalink to this definition">¶</a></dt>
<dd><p>Tools to work with SQLite DB’s (optimized, no SQL required).</p>
<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.checkEntryInColumn">
<code class="descname">checkEntryInColumn</code><span class="sig-paren">(</span><em>table</em>, <em>col</em>, <em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.checkEntryInColumn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.checkEntryInColumn" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if an entry exists in a table</p>
<ul class="simple">
<li><em>table</em> the table name</li>
<li><em>col</em> the column name</li>
<li><em>rid</em> the row to search for</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.closeDB">
<code class="descname">closeDB</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.closeDB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.closeDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Close the DB connection and reset the DBTools instance (can be reconnected)</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.connectSQLiteDB">
<code class="descname">connectSQLiteDB</code><span class="sig-paren">(</span><em>db_name</em>, <em>work_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.connectSQLiteDB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.connectSQLiteDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Connect to a sqlite database.</p>
<ul class="simple">
<li><em>db_name</em> the name of the sqlite database</li>
<li><em>work_dir</em> the optional database path</li>
</ul>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.createDBTable">
<code class="descname">createDBTable</code><span class="sig-paren">(</span><em>table</em>, <em>sqlcols</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.createDBTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.createDBTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a database table if it does not exist:</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the table name</li>
<li><em>sqlcols</em> a list containing the SQL definitions of the table columns: &lt;id&gt; &lt;type&gt; for example <cite>[‘gene TEXT PRIMARY KEY’, ‘aa_seq TEXT’, ‘nuc_seq TEXT’, ‘aa_len INT’, ‘nuc_len INT’]</cite></li>
</ul>
</div></blockquote>
<p>Effectively writes CREATE TABLE “table” (&lt;id&gt; &lt;type&gt;, gene TEXT PRIMARY KEY, aa_seq TEXT, nuc_seq TEXT, aa_len INT, nuc_len INT) % table</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.dumpTableToCSV">
<code class="descname">dumpTableToCSV</code><span class="sig-paren">(</span><em>table</em>, <em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.dumpTableToCSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.dumpTableToCSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a table as tab separated txt file</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the table to export</li>
<li><em>filename</em> the filename of the table dump</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.dumpTableToTxt">
<code class="descname">dumpTableToTxt</code><span class="sig-paren">(</span><em>table</em>, <em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.dumpTableToTxt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.dumpTableToTxt" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a table as tab separated txt file</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the table to export</li>
<li><em>filename</em> the filename of the table dump</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.executeSQL">
<code class="descname">executeSQL</code><span class="sig-paren">(</span><em>sql</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.executeSQL"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.executeSQL" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute a SQL command:</p>
<blockquote>
<div><ul class="simple">
<li><em>sql</em> a string containing a SQL command</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.fetchAll">
<code class="descname">fetchAll</code><span class="sig-paren">(</span><em>sql</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.fetchAll"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.fetchAll" title="Permalink to this definition">¶</a></dt>
<dd><p>Raw SQL query e.g. ‘SELECT id FROM gene WHERE gene=”G”’</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.getCell">
<code class="descname">getCell</code><span class="sig-paren">(</span><em>table</em>, <em>col</em>, <em>rid</em>, <em>cell</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.getCell"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.getCell" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the table cell which correspond to rid in column. Returns the value or None</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the database table</li>
<li><em>col</em> the column id</li>
<li><em>rid</em> the row index id</li>
<li><em>cell</em> the column of the cell you want tp extract</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.getColumns">
<code class="descname">getColumns</code><span class="sig-paren">(</span><em>table</em>, <em>cols</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.getColumns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.getColumns" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch the contents of one or more columns of data in a table</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the database table</li>
<li><em>cols</em> a list of one or more column id’s</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.getRow">
<code class="descname">getRow</code><span class="sig-paren">(</span><em>table</em>, <em>col</em>, <em>rid</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.getRow"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.getRow" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the table row(s) which correspond to rid in column. Returns the row(s) as a list, if the column is the primary key
this is always a single entry.</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the database table</li>
<li><em>col</em> the column id</li>
<li><em>rid</em> the row index id</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.getTable">
<code class="descname">getTable</code><span class="sig-paren">(</span><em>table</em>, <em>colOut=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.getTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.getTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an entire database table</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the table name</li>
<li><em>colOut</em> optionally return a tuple of (data,ColNames)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.insertData">
<code class="descname">insertData</code><span class="sig-paren">(</span><em>table</em>, <em>data</em>, <em>commit=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.insertData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.insertData" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Insert data into a table: “INSERT INTO %s (?, ?, ?, ?, ?) VALUES (?, ?, ?, ?, ?)” % table,</dt>
<dd><blockquote class="first">
<div>(?, ?, ?, ?, ?)) )</div></blockquote>
<ul class="last simple">
<li><em>table</em> the DB table name</li>
<li><em>data</em> a dictionary of {id:value} pairs</li>
<li><em>commit</em> whether to commit the data insertions</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.DBTools.updateData">
<code class="descname">updateData</code><span class="sig-paren">(</span><em>table</em>, <em>col</em>, <em>rid</em>, <em>data</em>, <em>commit=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#DBTools.updateData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.DBTools.updateData" title="Permalink to this definition">¶</a></dt>
<dd><p>Update already defined data</p>
<blockquote>
<div><ul class="simple">
<li><em>table</em> the table name</li>
<li><em>col</em> the column name</li>
<li><em>rid</em> the row id to update</li>
<li><em>data</em> a dictionary of {id:value} pairs</li>
<li><em>commit</em> whether to commit the data updates</li>
</ul>
<p>UPDATE COMPANY SET ADDRESS = ‘Texas’ WHERE ID = 6;</p>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBNetDB.KeGGSequenceTools">
<em class="property">class </em><code class="descclassname">cbmpy.CBNetDB.</code><code class="descname">KeGGSequenceTools</code><span class="sig-paren">(</span><em>url</em>, <em>db_name</em>, <em>work_dir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#KeGGSequenceTools"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.KeGGSequenceTools" title="Permalink to this definition">¶</a></dt>
<dd><p>Using the KeGG connector this class provides tools to construct an organims specific sequence database</p>
</dd></dl>

<dl class="class">
<dt id="cbmpy.CBNetDB.KeGGTools">
<em class="property">class </em><code class="descclassname">cbmpy.CBNetDB.</code><code class="descname">KeGGTools</code><span class="sig-paren">(</span><em>url</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#KeGGTools"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.KeGGTools" title="Permalink to this definition">¶</a></dt>
<dd><p>Class that holds useful methods for querying KeGG via a SUDS provided soap client</p>
<dl class="method">
<dt id="cbmpy.CBNetDB.KeGGTools.fetchSeqfromKeGG">
<code class="descname">fetchSeqfromKeGG</code><span class="sig-paren">(</span><em>k_gene</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#KeGGTools.fetchSeqfromKeGG"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.KeGGTools.fetchSeqfromKeGG" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a gene name try and retrieve the gene and amino acid sequence</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBNetDB.MIRIAMTools">
<em class="property">class </em><code class="descclassname">cbmpy.CBNetDB.</code><code class="descname">MIRIAMTools</code><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#MIRIAMTools"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.MIRIAMTools" title="Permalink to this definition">¶</a></dt>
<dd><p>Tools dealing with MIRIAM annotations</p>
</dd></dl>

<dl class="class">
<dt id="cbmpy.CBNetDB.RESTClient">
<em class="property">class </em><code class="descclassname">cbmpy.CBNetDB.</code><code class="descname">RESTClient</code><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#RESTClient"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.RESTClient" title="Permalink to this definition">¶</a></dt>
<dd><p>Class that provides the basis for application specific connectors to REST web services</p>
<dl class="method">
<dt id="cbmpy.CBNetDB.RESTClient.Close">
<code class="descname">Close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#RESTClient.Close"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.RESTClient.Close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close the currently active connection</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.RESTClient.Connect">
<code class="descname">Connect</code><span class="sig-paren">(</span><em>root</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#RESTClient.Connect"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.RESTClient.Connect" title="Permalink to this definition">¶</a></dt>
<dd><p>Establish HTTP connection to</p>
<blockquote>
<div><ul class="simple">
<li><em>root</em> the site root “www.google.com”</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.RESTClient.Get">
<code class="descname">Get</code><span class="sig-paren">(</span><em>query</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#RESTClient.Get"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.RESTClient.Get" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform an http GET using:</p>
<blockquote>
<div><ul class="simple">
<li><em>query</em> e.g.</li>
<li><em>reply_mode</em> [default=’‘] this is the reply mode</li>
</ul>
</div></blockquote>
<p>For example “/semanticSBML/annotate/search.xml?q=ATP”</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.RESTClient.GetLog">
<code class="descname">GetLog</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#RESTClient.GetLog"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.RESTClient.GetLog" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the logged history</p>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.RESTClient.Log">
<code class="descname">Log</code><span class="sig-paren">(</span><em>txt</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#RESTClient.Log"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.RESTClient.Log" title="Permalink to this definition">¶</a></dt>
<dd><p>Add txt to logfile history</p>
<blockquote>
<div><ul class="simple">
<li><em>txt</em> a string</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="cbmpy.CBNetDB.SemanticSBML">
<em class="property">class </em><code class="descclassname">cbmpy.CBNetDB.</code><code class="descname">SemanticSBML</code><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#SemanticSBML"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.SemanticSBML" title="Permalink to this definition">¶</a></dt>
<dd><p>REST client for connecting to SemanticSBML services</p>
<dl class="method">
<dt id="cbmpy.CBNetDB.SemanticSBML.parseXMLtoText">
<code class="descname">parseXMLtoText</code><span class="sig-paren">(</span><em>xml</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#SemanticSBML.parseXMLtoText"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.SemanticSBML.parseXMLtoText" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the xml output by quickLookup() into a list of URL</p>
<blockquote>
<div><ul class="simple">
<li><em>xml</em> XML returns from SemanticSBML</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.SemanticSBML.quickLookup">
<code class="descname">quickLookup</code><span class="sig-paren">(</span><em>txt</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#SemanticSBML.quickLookup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.SemanticSBML.quickLookup" title="Permalink to this definition">¶</a></dt>
<dd><p>Do a quick lookpup for txt using SemanticSBML (connectic if required) and return results. Returns
a list of identifiers.org id’s in descending priority (as return)</p>
<blockquote>
<div><ul class="simple">
<li><em>txt</em> the string to lookup</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="cbmpy.CBNetDB.SemanticSBML.viewDataInWebrowser">
<code class="descname">viewDataInWebrowser</code><span class="sig-paren">(</span><em>maxres=10</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBNetDB.html#SemanticSBML.viewDataInWebrowser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBNetDB.SemanticSBML.viewDataInWebrowser" title="Permalink to this definition">¶</a></dt>
<dd><p>Attempt to view #maxres results returned by SemanticSBML in the default browser</p>
<blockquote>
<div><ul class="simple">
<li><em>maxres</em> default maximum number of results to display.</li>
</ul>
</div></blockquote>
</dd></dl>

</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBPlot"></span><div class="section" id="cbmpy-cbplot-module">
<h2>CBMPy: CBPlot module<a class="headerlink" href="#cbmpy-cbplot-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBPlot.py 575 2017-04-13 12:18:44Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBPlot.plotFluxVariability">
<code class="descclassname">cbmpy.CBPlot.</code><code class="descname">plotFluxVariability</code><span class="sig-paren">(</span><em>fva_data</em>, <em>fva_names</em>, <em>fname</em>, <em>work_dir=None</em>, <em>title=None</em>, <em>ySlice=None</em>, <em>minHeight=None</em>, <em>maxHeight=None</em>, <em>roundec=None</em>, <em>autoclose=True</em>, <em>fluxval=True</em>, <em>type='png'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBPlot.html#plotFluxVariability"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBPlot.plotFluxVariability" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots and saves as an image the flux variability results as generated by CBSolver.FluxVariabilityAnalysis.</p>
<blockquote>
<div><ul class="simple">
<li><em>fva_data</em> FluxVariabilityAnalysis() FVA OUTPUT_ARRAY</li>
<li><em>fva_names</em> FluxVariabilityAnalysis() FVA OUTPUT_NAMES</li>
<li><em>fname</em> filename_base for the CSV output</li>
<li><em>work_dir</em> [default=None] if set the output directory for the csv files</li>
<li><em>title</em> [default=None] the user defined title for the graph</li>
<li><em>ySlice</em> [default=None] this sets an absolute (fixed) limit on the Y-axis (+- ySlice)</li>
<li><em>minHeight</em> [default=None] the minimum length that defined a span</li>
<li><em>maxHeight</em> [default=None] the maximum length a span can obtain, bar will be limited to maxHeight and coloured yellow</li>
<li><em>roundec</em> [default=None] an integer indicating at which decimal to round off output. Default is no rounding.</li>
<li><em>autoclose</em> [default=True] autoclose plot after save</li>
<li><em>fluxval</em> [default=True] plot the flux value</li>
<li><em>type</em> [default=’png’] the output format, depends on matplotlib backend e.g. ‘png’, ‘pdf’, ‘eps’</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBRead"></span><div class="section" id="cbmpy-cbread-module">
<h2>CBMPy: CBRead module<a class="headerlink" href="#cbmpy-cbread-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBRead.py 648 2018-05-23 20:01:54Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBRead.loadModel">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">loadModel</code><span class="sig-paren">(</span><em>sbmlfile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#loadModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.loadModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads any SBML model in COBRA, FAME (SBML2FBA), SBML3FBCv1, SBML3FBCv2 format.</p>
<blockquote>
<div><ul class="simple">
<li><em>sbmlfile</em> an SBML model file</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBRead.readCOBRASBML">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">readCOBRASBML</code><span class="sig-paren">(</span><em>fname</em>, <em>work_dir=None</em>, <em>return_sbml_model=False</em>, <em>delete_intermediate=False</em>, <em>fake_boundary_species_search=False</em>, <em>output_dir=None</em>, <em>skip_genes=False</em>, <em>scan_notes_gpr=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#readCOBRASBML"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.readCOBRASBML" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in a COBRA format SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm_mod, sbml_mod) pair if return_sbml_model=True</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>work_dir</em> is the working directory</li>
<li><em>delete_intermediate</em> [default=False] delete the intermediate SBML Level 3 FBC file</li>
<li><em>fake_boundary_species_search</em> [default=False] after looking for the boundary_condition of a species search for overloaded id’s &lt;id&gt;_b</li>
<li><em>output_dir</em> [default=None] the directory to output the intermediate SBML L3 files (if generated) default to input directory</li>
<li><em>skip_genes</em> [default=False] do not load GPR data</li>
<li><em>scan_notes_gpr</em> [default=True] if the model is loaded and no genes are detected the scan the &lt;notes&gt; field for GPR associationa</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBRead.readExcel97Model">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">readExcel97Model</code><span class="sig-paren">(</span><em>xlname</em>, <em>write_sbml=True</em>, <em>sbml_level=3</em>, <em>return_dictionaries=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#readExcel97Model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.readExcel97Model" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a model encoded as an Excel97 workbook and returns it as a CBMPy model object and SBML file. Note the workbook must be formatted
exactly like those produced by cbm.writeModelToExcel97(). Note that reactions have to be defined in <strong>both</strong> the <em>reaction</em>
and <em>network_react</em> sheets to be included in the model.</p>
<blockquote>
<div><ul class="simple">
<li><em>xlpath</em> the filename of the Excel workbook</li>
<li><em>return_model</em> [default=True] construct and return the CBMPy model</li>
<li><em>write_sbml</em> [default=True] write the SBML file to fname</li>
<li><em>return_dictionaries</em> [default=False] return the dictionaries constructed when reading the Excel file (in place of the model)</li>
<li><em>sbml_level</em> [default=3] write the SBML file as either SBML L2 FBA or SBML L3 FBC file.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBRead.readSBML2FBA">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">readSBML2FBA</code><span class="sig-paren">(</span><em>fname</em>, <em>work_dir=None</em>, <em>return_sbml_model=False</em>, <em>fake_boundary_species_search=False</em>, <em>scan_notes_gpr=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#readSBML2FBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.readSBML2FBA" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in an SBML Level 2 file with FBA annotation where:</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>work_dir</em> is the working directory if None then only fname is used</li>
<li><em>return_sbml_model</em> [default=False] return a a (cbm_mod, sbml_mod) pair</li>
<li><em>fake_boundary_species_search</em> [default=False] after looking for the boundary_condition of a species search for overloaded id’s &lt;id&gt;_b</li>
<li><em>scan_notes_gpr</em> [default=True] if the model is loaded and no genes are detected the scan the &lt;notes&gt; field for GPR associationa</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBRead.readSBML3FBC">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">readSBML3FBC</code><span class="sig-paren">(</span><em>fname</em>, <em>work_dir=None</em>, <em>return_sbml_model=False</em>, <em>xoptions={'validate': False}</em>, <em>scan_notes_gpr=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#readSBML3FBC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.readSBML3FBC" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in an SBML Level 3 file with FBC annotation where and return a CBM model object</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>work_dir</em> is the working directory</li>
<li><em>return_sbml_model</em> deprecated and ignored please update code</li>
<li><em>xoptions</em> special load options, enable with option=True except for nmatrix_type which has a type.<ul>
<li><em>nogenes</em> do not load/process genes</li>
<li><em>noannot</em> do not load/process any annotations</li>
<li><em>validate</em> validate model and display errors and warnings before loading</li>
<li><em>readcobra</em> read the cobra annotation</li>
<li><em>read_model_string</em> [default=False] read the model from a string (instead of a filename) containing an SBML document</li>
<li><em>nmatrix_type</em> [default=’normal’] define the type of stoichiometrich matrix to be built<ul>
<li>‘numpy’ dense numpy array (best performance)</li>
<li>‘scipy_csr’ scipy sparse matrix (lower performance, low memory)</li>
<li>‘sympy’ a sympy rational matrix (low performance, high memory, cast to dense to analyse)</li>
<li>None do not build matrix</li>
</ul>
</li>
</ul>
</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><em>scan_notes_gpr</em> [default=True] if the model is loaded and no genes are detected scan the &lt;notes&gt; field for GPR associationa</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBRead.readSK_FVA">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">readSK_FVA</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#readSK_FVA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.readSK_FVA" title="Permalink to this definition">¶</a></dt>
<dd><p>Read Stevens FVA results (opt.fva) file and return a list of dictionaries</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBRead.readSK_vertex">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">readSK_vertex</code><span class="sig-paren">(</span><em>fname</em>, <em>bigfile=True</em>, <em>fast_rational=False</em>, <em>nformat='%.14f'</em>, <em>compression=None</em>, <em>hdf5file=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#readSK_vertex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.readSK_vertex" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads in Stevens vertex analysis file:</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> the input filename (.all file that results from Stevens pipeline)</li>
<li><em>bigfile</em> [default=True] this option is now always true and is left in for backwards compatability</li>
<li><em>fast_rational</em> [default=False] by default off and uses SymPy for rational–&gt;float conversion, when on uses float decomposition with a slight (2th decimal) decrease in accuracy</li>
<li><em>nformat</em> [default=’%.14f’] the number format used in output files</li>
<li><em>compression</em> [default=None] compression to be used in hdf5 files can be one of [None, ‘lzf’, ‘gz?’, ‘szip’]</li>
<li><em>hdf5file</em> [default=None] if None then generic filename ‘_vtx_.tmp.hdf5’ is uses otherwise &lt;hdf5file&gt;.hdf5</li>
</ul>
</div></blockquote>
<p>and returns an hdf5 <em>filename</em> of the results with a single group named <strong>data</strong> which countains datasets</p>
<blockquote>
<div><ul class="simple">
<li>vertices</li>
<li>rays</li>
<li>lin</li>
</ul>
</div></blockquote>
<p>where all vectors are in terms of the column space of N.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBRead.readSK_vertexOld">
<code class="descclassname">cbmpy.CBRead.</code><code class="descname">readSK_vertexOld</code><span class="sig-paren">(</span><em>fname</em>, <em>bigfile=False</em>, <em>fast_rational=False</em>, <em>nformat='%.14f'</em>, <em>compresslevel=3</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBRead.html#readSK_vertexOld"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBRead.readSK_vertexOld" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads in Stevens vertex analysis file and returns, even more optimized for large datasets than the original.</p>
<blockquote>
<div><ul class="simple">
<li>a list of vertex vectors</li>
<li>a list of ray vectors</li>
<li>the basis of the lineality space as a list of vectors</li>
</ul>
</div></blockquote>
<p>all vectors in terms of the column space of N</p>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBReadtxt"></span><div class="section" id="cbmpy-cbreadtxt-module">
<h2>CBMPy: CBReadtxt module<a class="headerlink" href="#cbmpy-cbreadtxt-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBReadtxt.py 629 2017-10-24 22:01:14Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBReadtxt.readCSV">
<code class="descclassname">cbmpy.CBReadtxt.</code><code class="descname">readCSV</code><span class="sig-paren">(</span><em>model_file</em>, <em>bounds_file=None</em>, <em>biomass_flux=None</em>, <em>model_id='FBAModel'</em>, <em>reaction_prefix='R_'</em>, <em>has_header=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBReadtxt.html#readCSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBReadtxt.readCSV" title="Permalink to this definition">¶</a></dt>
<dd><p>This function loads a CSV file and translates it into a Python object:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span> <span class="o">*</span><span class="n">model_file</span><span class="o">*</span> <span class="n">the</span> <span class="n">name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">CSV</span> <span class="n">file</span> <span class="n">that</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">model</span>
<span class="o">-</span> <span class="o">*</span><span class="n">bounds_file</span><span class="o">*</span> <span class="n">the</span> <span class="n">name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">CSV</span> <span class="n">file</span> <span class="n">that</span> <span class="n">contains</span> <span class="n">the</span> <span class="n">flux</span> <span class="n">bounds</span>
<span class="o">-</span> <span class="o">*</span><span class="n">biomass_flux</span><span class="o">*</span> <span class="n">the</span> <span class="n">name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">reaction</span> <span class="n">that</span> <span class="ow">is</span> <span class="n">the</span> <span class="n">objective</span> <span class="n">function</span>
<span class="o">-</span> <span class="o">*</span><span class="n">reaction_prefix</span><span class="o">*</span> <span class="p">[</span><span class="n">default</span><span class="o">=</span><span class="s1">&#39;R _&#39;</span><span class="p">]</span> <span class="n">the</span> <span class="n">prefix</span> <span class="n">to</span> <span class="n">add</span> <span class="n">to</span> <span class="nb">input</span> <span class="n">reaction</span> <span class="n">ID</span><span class="s1">&#39;s</span>
<span class="o">-</span> <span class="o">*</span><span class="n">has_header</span><span class="o">*</span> <span class="p">[</span><span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">]</span> <span class="k">if</span> <span class="n">there</span> <span class="ow">is</span> <span class="n">a</span> <span class="n">header</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">csv</span> <span class="n">file</span>
</pre></div>
</div>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBSolver"></span><div class="section" id="cbmpy-cbsolver-module">
<h2>CBMPy: CBSolver module<a class="headerlink" href="#cbmpy-cbsolver-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBSolver.py 575 2017-04-13 12:18:44Z bgoli $)</p>
</div>
<span class="target" id="module-cbmpy.CBTools"></span><div class="section" id="cbmpy-cbtools-module">
<h2>CBMPy: CBTools module<a class="headerlink" href="#cbmpy-cbtools-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBTools.py 622 2017-09-13 12:56:30Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBTools.addFluxAsActiveObjective">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">addFluxAsActiveObjective</code><span class="sig-paren">(</span><em>f</em>, <em>reaction_id</em>, <em>osense</em>, <em>coefficient=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#addFluxAsActiveObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.addFluxAsActiveObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a flux as an active objective function</p>
<blockquote>
<div><ul class="simple">
<li><em>reaction_id</em> a string containing a reaction id</li>
<li><em>osense</em> objective sense must be <strong>maximize</strong> or <strong>minimize</strong></li>
<li><em>coefficient</em> the objective funtion coefficient [default=1]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.addGenesFromAnnotations">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">addGenesFromAnnotations</code><span class="sig-paren">(</span><em>fba</em>, <em>annotation_key='GENE ASSOCIATION'</em>, <em>gene_pattern=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#addGenesFromAnnotations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.addGenesFromAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>THIS METHOD IS DERPRECATED PLEASE USE cmod.createGeneAssociationsFromAnnotations()</p>
<p>Add genes to the model using the definitions stored in the annotation key</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> and fba object</li>
<li><em>annotation_key</em> the annotation dictionary key that holds the gene association for the protein/enzyme</li>
<li><em>gene_pattern</em> deprecated, not needed anymore</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.addSinkReaction">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">addSinkReaction</code><span class="sig-paren">(</span><em>fbam</em>, <em>species</em>, <em>lb=0.0</em>, <em>ub=1000.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#addSinkReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.addSinkReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a sink reactions that consumes a model <em>species</em> so that X –&gt;</p>
<blockquote>
<div><ul class="simple">
<li><em>fbam</em> an fba model object</li>
<li><em>species</em> a valid species name</li>
<li><em>lb</em> lower flux bound [default = 0.0]</li>
<li><em>ub</em> upper flux bound [default = 1000.0]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.addSourceReaction">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">addSourceReaction</code><span class="sig-paren">(</span><em>fbam</em>, <em>species</em>, <em>lb=0.0</em>, <em>ub=1000.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#addSourceReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.addSourceReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a source reactions that produces a model <em>species</em> so that –&gt; X</p>
<blockquote>
<div><ul class="simple">
<li><em>fbam</em> an fba model object</li>
<li><em>species</em> a valid species name</li>
<li><em>lb</em> lower flux bound [default = 0.0]</li>
<li><em>ub</em> upper flux bound [default = 1000.0]</li>
</ul>
</div></blockquote>
<p>Note reversiblity is determined by the lower bound, default 0 = irreversible. If
negative then reversible.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.addStoichToFBAModel">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">addStoichToFBAModel</code><span class="sig-paren">(</span><em>fm</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#addStoichToFBAModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.addStoichToFBAModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Build stoichiometry: this method has been refactored into the model class - cmod.buildStoichMatrix()</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.checkExchangeReactions">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">checkExchangeReactions</code><span class="sig-paren">(</span><em>fba</em>, <em>autocorrect=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#checkExchangeReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.checkExchangeReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan all reactions for exchange reactions (reactions containing a boundary species), return a list of
inconsistent reactions or correct automatically.</p>
<ul class="simple">
<li><em>fba</em> a CBMPy model</li>
<li><em>autocorrect</em> [default=True] correctly set the “is_exchange” attribute on a reaction</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.checkFluxBoundConsistency">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">checkFluxBoundConsistency</code><span class="sig-paren">(</span><em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#checkFluxBoundConsistency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.checkFluxBoundConsistency" title="Permalink to this definition">¶</a></dt>
<dd><p>Check flux bound consistency checks for multiply defined bounds, bounds without a reaction, inconsistent bounds with respect to each other
and reaction reversbility. Returns a dictionary of bounds/reactions where errors occur.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.checkIds">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">checkIds</code><span class="sig-paren">(</span><em>fba</em>, <em>items='all'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#checkIds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.checkIds" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the id’s of the specified model attributes to see if the name is legal and if there are duplicates.
Returns a list of items with errors.</p>
<ul class="simple">
<li><em>fba</em> a CBMPy model instance</li>
<li><em>items</em> [default=’all’] ‘all’ means ‘species,reactions,flux_bounds,objectives’ of which one or more can be specified</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.checkReactionBalanceElemental">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">checkReactionBalanceElemental</code><span class="sig-paren">(</span><em>f</em>, <em>Rid=None</em>, <em>zero_tol=1e-12</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#checkReactionBalanceElemental"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.checkReactionBalanceElemental" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the reaction is balanced using the chemical formula</p>
<blockquote>
<div><ul class="simple">
<li><em>f</em> the FBA object</li>
<li><em>Rid</em> [default = None] the reaction to check, defaults to all</li>
<li><em>zero_tol</em> [default=1.0e-12] the floating point zero used for elemental balancing</li>
</ul>
</div></blockquote>
<p>This function is derived from the code found here: <a class="reference external" href="http://pyparsing.wikispaces.com/file/view/chemicalFormulas.py">http://pyparsing.wikispaces.com/file/view/chemicalFormulas.py</a></p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.createTempFileName">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">createTempFileName</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#createTempFileName"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.createTempFileName" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a temporary filename</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.createZipArchive">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">createZipArchive</code><span class="sig-paren">(</span><em>zipname</em>, <em>files</em>, <em>move=False</em>, <em>compression='normal'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#createZipArchive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.createZipArchive" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a zip archive which contains one or more files</p>
<blockquote>
<div><ul class="simple">
<li><em>zipname</em> the name of the zip archive to create (fully qualified)</li>
<li><em>files</em> either a valid filename or a list of filenames (fully qualified)</li>
<li><em>move</em> [default=False] attempt to delete input files after zip-archive creation</li>
<li><em>compression</em> [default=’normal’] normal zip compression, set as None for no compression only store files (zlib not required)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.deSerialize">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">deSerialize</code><span class="sig-paren">(</span><em>s</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#deSerialize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.deSerialize" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserializes a serialised object contained in a string</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.deSerializeFromDisk">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">deSerializeFromDisk</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#deSerializeFromDisk"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.deSerializeFromDisk" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads a serialised Python pickle from <em>filename</em> returns the Python object(s)</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportArray2CSV">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportArray2CSV</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportArray2CSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportArray2CSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an array to fname.csv</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>fname</em> the output filename</li>
<li><em>sep</em> [default=’,’] the column separator</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportArray2TXT">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportArray2TXT</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportArray2TXT"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportArray2TXT" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an array to fname.txt</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>fname</em> the output filename</li>
<li><em>sep</em> [default=’,’] the column separator</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportLabelledArray">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportLabelledArray</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em>, <em>names=None</em>, <em>sep='</em>, <em>'</em>, <em>fmt='%f'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportLabelledArray"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportLabelledArray" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a 2D array type object to file</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>names</em> [default=None] the list of row names</li>
<li><em>fname</em> the output filename</li>
<li><em>sep</em> [default=’,’] the column separator</li>
<li><em>fmt</em> [default=’%s’] the output number format</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportLabelledArray2CSV">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportLabelledArray2CSV</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em>, <em>names=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportLabelledArray2CSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportLabelledArray2CSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an array with row names to fname.csv</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>fname</em> the output filename</li>
<li><em>names</em> [default=None] the list of row names</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportLabelledArray2TXT">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportLabelledArray2TXT</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em>, <em>names=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportLabelledArray2TXT"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportLabelledArray2TXT" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an array with row names to fname.txt</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>names</em> [default=None] the list of row names</li>
<li><em>fname</em> the output filename</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportLabelledArrayWithHeader">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportLabelledArrayWithHeader</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em>, <em>names=None</em>, <em>header=None</em>, <em>sep='</em>, <em>'</em>, <em>fmt='%f'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportLabelledArrayWithHeader"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportLabelledArrayWithHeader" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an array with row names and header</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>names</em> [default=None] the list of row names</li>
<li><em>header</em> [default=None] the list of column names</li>
<li><em>fname</em> the output filename</li>
<li><em>sep</em> [default=’,’] the column separator</li>
<li><em>fmt</em> [default=’%s’] the output number format</li>
<li><em>appendlist</em> [default=False] if True append the array to <em>fname</em> otherwise create a new file</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportLabelledArrayWithHeader2CSV">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportLabelledArrayWithHeader2CSV</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em>, <em>names=None</em>, <em>header=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportLabelledArrayWithHeader2CSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportLabelledArrayWithHeader2CSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an array with row names and header to fname.csv</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>fname</em> the output filename</li>
<li><em>names</em> [default=None] the list of row names</li>
<li><em>header</em> [default=None] the list of column names</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportLabelledArrayWithHeader2TXT">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportLabelledArrayWithHeader2TXT</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em>, <em>names=None</em>, <em>header=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportLabelledArrayWithHeader2TXT"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportLabelledArrayWithHeader2TXT" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an array with row names and header to fname.txt</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the an array like object</li>
<li><em>names</em> the list of row names</li>
<li><em>header</em> the list of column names</li>
<li><em>fname</em> the output filename</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.exportLabelledLinkedList">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">exportLabelledLinkedList</code><span class="sig-paren">(</span><em>arr</em>, <em>fname</em>, <em>names=None</em>, <em>sep='</em>, <em>'</em>, <em>fmt='%s'</em>, <em>appendlist=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#exportLabelledLinkedList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.exportLabelledLinkedList" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a 2D linked list [[…],[…],[…],[…]] and optionally a list of row labels to file:</p>
<blockquote>
<div><ul class="simple">
<li><em>arr</em> the linked list</li>
<li><em>fname</em> the output filename</li>
<li><em>names</em> [default=None] the list of row names</li>
<li><em>sep</em> [default=’,’] the column separator</li>
<li><em>fmt</em> [default=’%s’] the output number format</li>
<li><em>appendlist</em> [default=False] if True append the array to <em>fname</em> otherwise create a new file</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.findDeadEndMetabolites">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">findDeadEndMetabolites</code><span class="sig-paren">(</span><em>fbam</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#findDeadEndMetabolites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.findDeadEndMetabolites" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds dead-end (single reaction) metabolites rows in N with a single entry), returns a list of (metabolite, reaction) ids</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.findDeadEndReactions">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">findDeadEndReactions</code><span class="sig-paren">(</span><em>fbam</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#findDeadEndReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.findDeadEndReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds dead-end (single substrate/product) reactions (cols in N with a single entry), returns a list of (metabolite, reaction) ids</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.fixReversibility">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">fixReversibility</code><span class="sig-paren">(</span><em>fbam</em>, <em>auto_correct=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#fixReversibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.fixReversibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Set fluxbound lower bound from reactions reversibility information.</p>
<blockquote>
<div><ul class="simple">
<li><em>fbam</em> and FBAModel instance</li>
<li><em>auto_correct</em> (default=False) if True automatically sets lower bound to zero if required, otherwise prints a warning if false.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.getBoundsDict">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">getBoundsDict</code><span class="sig-paren">(</span><em>fbamod</em>, <em>substring=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#getBoundsDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.getBoundsDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary of reactions&amp;bounds</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.getExchBoundsDict">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">getExchBoundsDict</code><span class="sig-paren">(</span><em>fbamod</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#getExchBoundsDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.getExchBoundsDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary of all exchange reactions (as determined by the is_exchange attribute of Reaction)</p>
<ul class="simple">
<li><em>fbamod</em> a CBMPy model</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.getModelGenesPerReaction">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">getModelGenesPerReaction</code><span class="sig-paren">(</span><em>fba</em>, <em>gene_pattern=None</em>, <em>gene_annotation_key='GENE ASSOCIATION'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#getModelGenesPerReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.getModelGenesPerReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a BiGG style gene annotation string using default gene_pattern=’((W*w*W*))’ or
(&lt;any non-alphanum&gt;&lt;any alphanum&gt;&lt;any non-alphanum&gt;)</p>
<p>Old eColi specific pattern ‘(bw*W)’</p>
<p>It is advisable to use the model methods directly rather than this function</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.loadObj">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">loadObj</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#loadObj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.loadObj" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads a serialised Python pickle from <em>filename</em>.dat returns the Python object(s)</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.merge2Models">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">merge2Models</code><span class="sig-paren">(</span><em>m1</em>, <em>m2</em>, <em>ignore=None</em>, <em>ignore_duplicate_ids=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#merge2Models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.merge2Models" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge 2 models, this method does a raw merge of model 2 into model 1 without any model checking.
Component id’s in ignore are ignored in both models and the first objective of model 1 is arbitrarily
set as active. Compartments are also merged and a new “OuterMerge” compartment is also created.</p>
<p>In all cases duplicate id’s are tracked and ignored, essentially using the object id encountered first -
usually that of model 1. Duplicate checking can be disabled by setting the <em>ignore_duplicate_ids</em> flag.</p>
<ul class="simple">
<li><em>m1</em> model 1</li>
<li><em>m2</em> model 2</li>
<li><em>ignore</em> [[]] do not merge these id’s</li>
<li><em>ignore_duplicate_ids</em> [False] default behaviour that can be enabled</li>
</ul>
<p>In development: merging genes and gpr’s.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.processBiGGannotationNote">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">processBiGGannotationNote</code><span class="sig-paren">(</span><em>fba</em>, <em>annotation_key='note'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#processBiGGannotationNote"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.processBiGGannotationNote" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the HTML formatted reaction information stored in the BiGG notes field.
This function is being deprecated and replaced by <cite>CBTools.processSBMLAnnotationNotes()</cite></p>
<blockquote>
<div><ul class="simple">
<li>requires an <em>annotation_key</em> which contains a BiGG HTML fragment</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.processBiGGchemFormula">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">processBiGGchemFormula</code><span class="sig-paren">(</span><em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#processBiGGchemFormula"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.processBiGGchemFormula" title="Permalink to this definition">¶</a></dt>
<dd><p>Disambiguates the overloaded BiGG name NAME_CHEMFORMULA into</p>
<blockquote>
<div><ul class="simple">
<li><em>species.name</em> NAME</li>
<li><em>species.chemFormula</em> CHEMFORMULA</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.processExchangeReactions">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">processExchangeReactions</code><span class="sig-paren">(</span><em>fba</em>, <em>key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#processExchangeReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.processExchangeReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract exchange reactions from model using <em>key</em> and return:</p>
<blockquote>
<div><ul class="simple">
<li>a dictionary of all exchange reactions without <em>medium</em> reactions</li>
<li>a dictionary of <em>medium</em> exchange reactions (negative lower bound)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.processSBMLAnnotationNotes">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">processSBMLAnnotationNotes</code><span class="sig-paren">(</span><em>fba</em>, <em>annotation_key='note'</em>, <em>level=3</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#processSBMLAnnotationNotes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.processSBMLAnnotationNotes" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the HTML formatted reaction information stored in the SBML notes field currently
processes BiGG and PySCeSCBM style annotations it looks for the the annotation indexed
with the <em>annotation_key</em></p>
<blockquote>
<div><ul class="simple">
<li><em>annotation_key</em> [default=’note’] which contains a HTML/XHTML fragment in BiGG/PySCeSCBM format (ignored in L3)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.removeFixedSpeciesReactions">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">removeFixedSpeciesReactions</code><span class="sig-paren">(</span><em>f</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#removeFixedSpeciesReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.removeFixedSpeciesReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is a hack that removes reactions which only have boundary species as reactants
and products. These are typically gene associations encoded in the Manchester style and there
is probably a better way of working around this problem …</p>
<blockquote>
<div><ul class="simple">
<li><em>f</em> an instantiated fba model object</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.roundOffWithSense">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">roundOffWithSense</code><span class="sig-paren">(</span><em>val</em>, <em>osense='max'</em>, <em>tol=1e-08</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#roundOffWithSense"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.roundOffWithSense" title="Permalink to this definition">¶</a></dt>
<dd><p>Round of a value in a way that takes into consideration the sense of the operation that generated it</p>
<blockquote>
<div><ul class="simple">
<li><em>val</em> the value</li>
<li><em>osense</em> [default=’max’] the sense</li>
<li><em>tol</em> [default=1e-8] the tolerance of the roundoff factor</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.scanForReactionDuplicates">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">scanForReactionDuplicates</code><span class="sig-paren">(</span><em>f</em>, <em>ignore_coefficients=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#scanForReactionDuplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.scanForReactionDuplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>This method uses uses a brute force apprach to finding reactions with matching
stoichiometry</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.scanForUnbalancedReactions">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">scanForUnbalancedReactions</code><span class="sig-paren">(</span><em>f</em>, <em>output='all'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#scanForUnbalancedReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.scanForUnbalancedReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Scan a model for unbalanced reactions, returns a tuple of dictionaries balanced and unbalanced:</p>
<blockquote>
<div><ul class="simple">
<li><em>f</em> an FBA model instance</li>
<li><em>output</em> [default=’all’] can be one of [‘all’,’charge’,’element’]</li>
<li><em>charge</em> return all charge <strong>un</strong> balanced reactions</li>
<li><em>element</em> return all element <strong>un</strong> balanced reactions</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.setSpeciesPropertiesFromAnnotations">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">setSpeciesPropertiesFromAnnotations</code><span class="sig-paren">(</span><em>fbam</em>, <em>overwriteCharge=False</em>, <em>overwriteChemFormula=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#setSpeciesPropertiesFromAnnotations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.setSpeciesPropertiesFromAnnotations" title="Permalink to this definition">¶</a></dt>
<dd><p>This will attempt to set the model Species properties from the annotation. With the default options
it will only replace missing data. With ChemicalFormula this is easy to detect however charge may
have an “unknown value” of 0. Setting the optional values to true will replace any existing value
with any valid annotation.</p>
<blockquote>
<div><ul class="simple">
<li><em>overwriteChemFormula</em> [default=False]</li>
<li><em>overwriteCharge</em> [default=False]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.splitReversibleReactions">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">splitReversibleReactions</code><span class="sig-paren">(</span><em>fba</em>, <em>selected_reactions=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#splitReversibleReactions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.splitReversibleReactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Split a (set of) reactions into reversible reactions returns a copy of the original model</p>
<p>R1: A = B
R1f: A -&gt; B
R1r: B -&gt; A</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instantiated CBMPy model object</li>
<li><em>selected_reactions</em> if a reversible reaction id is in here split it</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.splitSingleReversibleReaction">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">splitSingleReversibleReaction</code><span class="sig-paren">(</span><em>fba</em>, <em>rid</em>, <em>fwd_id=None</em>, <em>rev_id=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#splitSingleReversibleReaction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.splitSingleReversibleReaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Split a single reversible reaction into two irreversible reactions, returns the original reversible reaction and bounds
while deleting them from model.</p>
<p>R1: A = B
R1_fwd: A -&gt; B
R1_rev: B -&gt; A</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instantiated CBMPy model object</li>
<li><em>rid</em> a valid reaction id</li>
<li><em>fwd_id</em> [default=None] the new forward reaction id, defaults to rid_fwd</li>
<li><em>rev_id</em> [default=None] the new forward reaction id, defaults to rid_rev</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.storeObj">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">storeObj</code><span class="sig-paren">(</span><em>obj</em>, <em>filename</em>, <em>compress=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#storeObj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.storeObj" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores a Python <em>obj</em> as a serialised binary object in <em>filename</em>.dat</p>
<ul class="simple">
<li><em>obj</em> a python object</li>
<li><em>filename</em> the base filename</li>
<li><em>compress</em> [False] use gzip compression not <em>implemented</em></li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBTools.stringReplace">
<code class="descclassname">cbmpy.CBTools.</code><code class="descname">stringReplace</code><span class="sig-paren">(</span><em>fbamod</em>, <em>old</em>, <em>new</em>, <em>target</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBTools.html#stringReplace"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBTools.stringReplace" title="Permalink to this definition">¶</a></dt>
<dd><p>This is alpha stuff, target can be:</p>
<blockquote>
<div><ul class="simple">
<li>‘species’</li>
<li>‘reactions’</li>
<li>‘constraints’</li>
<li>‘objectives’</li>
<li>‘all’</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBWrite"></span><div class="section" id="cbmpy-cbwrite-module">
<h2>CBMPy: CBWrite module<a class="headerlink" href="#cbmpy-cbwrite-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBWrite.py 648 2018-05-23 20:01:54Z bgoli $)</p>
<dl class="function">
<dt id="cbmpy.CBWrite.BuildHformatFluxBounds">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">BuildHformatFluxBounds</code><span class="sig-paren">(</span><em>fba</em>, <em>infinity_replace=None</em>, <em>use_rational=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#BuildHformatFluxBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.BuildHformatFluxBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Build and return a csio that contains the flux bounds in H format</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> a PySCeS-CBM FBA object</li>
<li><em>infinity_replace</em> [default=None] if defined this is the abs(value) of +-&lt;infinity&gt;</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.BuildLPConstraints">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">BuildLPConstraints</code><span class="sig-paren">(</span><em>fba</em>, <em>use_rational=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#BuildLPConstraints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.BuildLPConstraints" title="Permalink to this definition">¶</a></dt>
<dd><p>Build and return a csio that contains constraint constructed from
the StoichiometeryLP object</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object which has a stoichiometry</li>
<li><em>use_rational</em> write rational number output [default=False]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.BuildLPConstraintsMath">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">BuildLPConstraintsMath</code><span class="sig-paren">(</span><em>fba</em>, <em>use_rational=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#BuildLPConstraintsMath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.BuildLPConstraintsMath" title="Permalink to this definition">¶</a></dt>
<dd><p>Build and return a csio that contains the constaints in LP format
Strict refers to dS/dt =&gt; 0 and dS/dt &lt;= 0</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.BuildLPConstraintsRelaxed">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">BuildLPConstraintsRelaxed</code><span class="sig-paren">(</span><em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#BuildLPConstraintsRelaxed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.BuildLPConstraintsRelaxed" title="Permalink to this definition">¶</a></dt>
<dd><p>Build and return a csio that contains the constaints in LP format
Relaxed refers to dS/dt &gt;= 0</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.BuildLPConstraintsStrict">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">BuildLPConstraintsStrict</code><span class="sig-paren">(</span><em>fba</em>, <em>use_rational=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#BuildLPConstraintsStrict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.BuildLPConstraintsStrict" title="Permalink to this definition">¶</a></dt>
<dd><p>Build and return a csio that contains the constaints in LP format
Strict refers to dS/dt = 0</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.BuildLPFluxBounds">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">BuildLPFluxBounds</code><span class="sig-paren">(</span><em>fba</em>, <em>use_rational=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#BuildLPFluxBounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.BuildLPFluxBounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Build and return a csio that contains the flux bounds in LP format</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.BuildLPUserConstraints">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">BuildLPUserConstraints</code><span class="sig-paren">(</span><em>fba</em>, <em>use_rational=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#BuildLPUserConstraints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.BuildLPUserConstraints" title="Permalink to this definition">¶</a></dt>
<dd><p>Build and return a csio that contains constraint constructed from
the StoichiometeryLP object</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object which has a stoichiometry</li>
<li><em>use_rational</em> write rational number output [default=False]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.WriteFVAdata">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">WriteFVAdata</code><span class="sig-paren">(</span><em>fva</em>, <em>names</em>, <em>fname</em>, <em>work_dir=None</em>, <em>roundec=None</em>, <em>scale_min=False</em>, <em>appendfile=False</em>, <em>info=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#WriteFVAdata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.WriteFVAdata" title="Permalink to this definition">¶</a></dt>
<dd><p>INFO: this method will be deprecated please update your scripts to use “writeFVAdata()”</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.WriteFVAtoCSV">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">WriteFVAtoCSV</code><span class="sig-paren">(</span><em>id</em>, <em>fva</em>, <em>names</em>, <em>Dir=None</em>, <em>fbaObj=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#WriteFVAtoCSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.WriteFVAtoCSV" title="Permalink to this definition">¶</a></dt>
<dd><p>INFO: this method will be deprecated please update your scripts to use “writeFVAtoCSV()”</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.WriteModelHFormatFBA">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">WriteModelHFormatFBA</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em>, <em>use_rational=False</em>, <em>fullLP=True</em>, <em>format='%s'</em>, <em>infinity_replace=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#WriteModelHFormatFBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.WriteModelHFormatFBA" title="Permalink to this definition">¶</a></dt>
<dd><p>INFO: this method will be deprecated please update your scripts to use “writeModelHFormatFBA2()”</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.WriteModelHFormatFBA2">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">WriteModelHFormatFBA2</code><span class="sig-paren">(</span><em>fba</em>, <em>fname=None</em>, <em>work_dir=None</em>, <em>use_rational=False</em>, <em>fullLP=True</em>, <em>format='%s'</em>, <em>infinity_replace=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#WriteModelHFormatFBA2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.WriteModelHFormatFBA2" title="Permalink to this definition">¶</a></dt>
<dd><p>INFO: this method will be deprecated please update your scripts to use “writeModelHFormatFBA2()”</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.WriteModelLP">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">WriteModelLP</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em>, <em>fname=None</em>, <em>multisymb=' '</em>, <em>format='%s'</em>, <em>use_rational=False</em>, <em>constraint_mode=None</em>, <em>quiet=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#WriteModelLP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.WriteModelLP" title="Permalink to this definition">¶</a></dt>
<dd><p>INFO: this method will be deprecated please update your scripts to use “writeModelLP()”</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.WriteModelLPOld">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">WriteModelLPOld</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em>, <em>multisymb=' '</em>, <em>lpt=True</em>, <em>constraint_mode='strict'</em>, <em>use_rational=False</em>, <em>format='%s'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#WriteModelLPOld"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.WriteModelLPOld" title="Permalink to this definition">¶</a></dt>
<dd><p>INFO: this method will be deprecated please update your scripts to use “writeModelLPOld()”</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.WriteModelRaw">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">WriteModelRaw</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#WriteModelRaw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.WriteModelRaw" title="Permalink to this definition">¶</a></dt>
<dd><p>INFO: this method will be deprecated please update your scripts to use “writeModelRaw()”</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.convertExcelToFloat">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">convertExcelToFloat</code><span class="sig-paren">(</span><em>num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#convertExcelToFloat"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.convertExcelToFloat" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an Excel “number” to a float</p>
<blockquote>
<div><ul class="simple">
<li><em>num</em> a number</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.convertFloatToExcel">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">convertFloatToExcel</code><span class="sig-paren">(</span><em>num</em>, <em>roundoff</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#convertFloatToExcel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.convertFloatToExcel" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a float to Excel compatible “number”</p>
<blockquote>
<div><ul class="simple">
<li><em>num</em> a number</li>
<li><em>roundoff</em> the number of roundoff digits for round()</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.exportModel">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">exportModel</code><span class="sig-paren">(</span><em>fba</em>, <em>fname=None</em>, <em>fmt='lp'</em>, <em>work_dir=None</em>, <em>use_rational='both'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#exportModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.exportModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Export the FBA model in different formats:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> the FBA model</li>
<li><em>fname</em> [default=None] the exported filename if None then <cite>fba.getId()</cite> is used</li>
<li><em>fmt</em> [default=’lp’] the export format can be one of: ‘lp’ (CPLEX), ‘hformat’ (Polyhedra), ‘all’ (both)</li>
<li><em>use_rational</em> [default=’both’] if <em>all</em> or <em>hformat</em> is specified should hformat files be written using rational math or not. The default <em>both</em> is the legacy behaviour and writes both.</li>
</ul>
</div></blockquote>
<p>Note that ‘hformat’ ignores ‘fname’ and only uses fba.getId() this is a legacy behaviour</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.generateBGID">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">generateBGID</code><span class="sig-paren">(</span><em>num</em>, <em>prefix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#generateBGID"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.generateBGID" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Create a BGID generator, which is &lt;prefix&gt;&lt;num&gt; where perfix is two letters num is padded to 6 figures</dt>
<dd><ul class="first last simple">
<li><em>num</em> the starting number</li>
<li><em>prefix</em> the two letter prefix</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.printFBASolution">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">printFBASolution</code><span class="sig-paren">(</span><em>fba</em>, <em>include_all=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#printFBASolution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.printFBASolution" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints the FBA optimal solution to the screen.</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an FBA model object</li>
<li><em>include_all</em> include all variables</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.saveModel">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">saveModel</code><span class="sig-paren">(</span><em>model</em>, <em>filename</em>, <em>compress=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#saveModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.saveModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves the model to an SBML file using the lates SBML3 FBC version.</p>
<ul class="simple">
<li><em>model</em> the CBMPy model</li>
<li><em>filename</em> the filename to write</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeCOBRASBML">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeCOBRASBML</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>directory=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeCOBRASBML"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeCOBRASBML" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and writes it to file as a COBRA compatible :</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> the model will be written as XML to <em>fname</em></li>
<li><em>directory</em> [default=None] if defined it is prepended to fname</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeFVAdata">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeFVAdata</code><span class="sig-paren">(</span><em>fvadata</em>, <em>names</em>, <em>fname</em>, <em>work_dir=None</em>, <em>roundec=None</em>, <em>scale_min=False</em>, <em>appendfile=False</em>, <em>info=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeFVAdata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeFVAdata" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method replaces the glpk/cplx_WriteFVAtoCSV methods. Data is output as a csv file
with columns: FluxName, FVA_MIN, FVA_MAX, OPT_VAL, SPAN</p>
<blockquote>
<div><ul class="simple">
<li><em>fvadata</em> FluxVariabilityAnalysis() FVA OUTPUT_ARRAY</li>
<li><em>names</em> FluxVariabilityAnalysis() FVA OUTPUT_NAMES</li>
<li><em>fname</em> filename_base for the CSV output</li>
<li><em>work_dir</em> [default=None] if set the output directory for the csv files</li>
<li><em>roundec</em> [default=None] an integer indicating at which decimal to round off output. Default is no rounding.</li>
<li><em>scale_min</em> [default=False] normalise each flux such that that FVA_MIN = 0.0</li>
<li><em>appendfile</em> [default=False] instead of opening a new file try and append the data</li>
<li><em>info</em> [default=None] a string added to the results as an extra column, useful with <cite>appendfile</cite></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeFVAtoCSV">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeFVAtoCSV</code><span class="sig-paren">(</span><em>fvadata</em>, <em>names</em>, <em>fname</em>, <em>Dir=None</em>, <em>fbaObj=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeFVAtoCSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeFVAtoCSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method replaces the glpk/cplx_WriteFVAtoCSV methods.</p>
<blockquote>
<div><ul class="simple">
<li><em>fvadata</em> FluxVariabilityAnalysis() OUTPUT_ARRAY</li>
<li><em>names</em> FluxVariabilityAnalysis() OUTPUT_NAMES</li>
<li><em>fname</em> filename_base for the CSV output</li>
<li><em>Dir</em> [default=None] if set the output directory for the csv files</li>
<li><em>fbaObj</em> [default=None] if supplied adds extra model information into the output tables</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeMinDistanceLPwithCost">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeMinDistanceLPwithCost</code><span class="sig-paren">(</span><em>fname</em>, <em>fbas</em>, <em>work_dir=None</em>, <em>ignoreDistance=[]</em>, <em>constraint_mode='strict'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeMinDistanceLPwithCost"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeMinDistanceLPwithCost" title="Permalink to this definition">¶</a></dt>
<dd><p>For backwards compatability only</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelHFormatFBA">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelHFormatFBA</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em>, <em>use_rational=False</em>, <em>fullLP=True</em>, <em>format='%s'</em>, <em>infinity_replace=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelHFormatFBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelHFormatFBA" title="Permalink to this definition">¶</a></dt>
<dd><p>Write an FBA-LP in polynomial H-Format file. This version has been replaced by <cite>writeModelHFormatFBA2()</cite>
but is kept for backwards compatability.</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> a PySCeS-CBM FBA object</li>
<li><em>Work_dir</em> [default=None] the output directory</li>
<li><em>use_rational</em> [default=false] use rational numbers in output (requires sympy)</li>
<li><em>fullLP</em> [default=True] include the default objective function as a maximization target</li>
<li><em>format</em> [default=’%s’] the number format string</li>
<li><em>infinity_replace</em> [default=None] if defined this is the abs(value) of +-&lt;infinity&gt;</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelHFormatFBA2">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelHFormatFBA2</code><span class="sig-paren">(</span><em>fba</em>, <em>fname=None</em>, <em>work_dir=None</em>, <em>use_rational=False</em>, <em>fullLP=True</em>, <em>format='%s'</em>, <em>infinity_replace=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelHFormatFBA2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelHFormatFBA2" title="Permalink to this definition">¶</a></dt>
<dd><p>Write an FBA-LP in polynomial H-Format file. This is an improved version of <cite>WriteModelHFormatFBA()</cite>
which it replaces. Note that if a SymPy matrix is used as input then use_rational is automatically enabled.</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> a PySCeS-CBM FBA object</li>
<li><em>fname</em> [default=None] the output filename, fba.getId() if not defined</li>
<li><em>Work_dir</em> [default=None] the output directory</li>
<li><em>use_rational</em> [default=false] use rational numbers in output (requires sympy)</li>
<li><em>fullLP</em> [default=True] include the default objective function as a maximization target</li>
<li><em>format</em> [default=’%s’] the number format string</li>
<li><em>infinity_replace</em> [default=None] if defined this is the abs(value) of +-&lt;infinity&gt;</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelInfoToFile">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelInfoToFile</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>Dir=None</em>, <em>separator='</em>, <em>'</em>, <em>only_exchange=False</em>, <em>met_type='all'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelInfoToFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelInfoToFile" title="Permalink to this definition">¶</a></dt>
<dd><p>This function writes a CBModel to file</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instance of an PySCeSCBM model</li>
<li><em>fname</em> the output filename</li>
<li><em>Dir</em> [default=None] use directory if not None</li>
<li><em>separator</em> [default=’,’] the column separator</li>
<li><em>only_exchange</em> [default=False] only output fluxes labelled as exchange reactions</li>
<li><em>type</em> [default=’all’] only output certain type of species: ‘all’,’boundary’ or ‘variable’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelLP">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelLP</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em>, <em>fname=None</em>, <em>multisymb=' '</em>, <em>format='%s'</em>, <em>use_rational=False</em>, <em>constraint_mode=None</em>, <em>quiet=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelLP"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelLP" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes an FBA object as an LP in CPLEX LP format</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instantiated FBAmodel instance</li>
<li><em>work_dir</em> directory designated for output</li>
<li><em>fname</em> the file name [default=fba.getId()]</li>
<li><em>multisymb</em> the multiplication symbol (default: &lt;space&gt;)</li>
<li><em>format</em> the number format of the output</li>
<li><em>use_rational</em> output rational numbers [default=False]</li>
<li><em>quiet</em> [default=False] supress information messages</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelLPOld">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelLPOld</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em>, <em>multisymb=' '</em>, <em>lpt=True</em>, <em>constraint_mode='strict'</em>, <em>use_rational=False</em>, <em>format='%s'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelLPOld"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelLPOld" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a fba as an LP/LPT</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instantiated FBAmodel instance</li>
<li><em>work_dir</em> directory designated for output</li>
<li><em>multisymb</em> the multiplication symbol (default: &lt;space&gt;)</li>
<li><em>lpt</em> the file format (default: True for lpt) or False for lp</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelRaw">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelRaw</code><span class="sig-paren">(</span><em>fba</em>, <em>work_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelRaw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelRaw" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a fba (actually just dumps it) to a text file.</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instantiated FBAmodel instance</li>
<li><em>work_dir</em> directory designated for output</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelToCOMBINEarchive">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelToCOMBINEarchive</code><span class="sig-paren">(</span><em>mod</em>, <em>fname=None</em>, <em>directory=None</em>, <em>sbmlname=None</em>, <em>withExcel=True</em>, <em>vc_given='CBMPy'</em>, <em>vc_family='Software'</em>, <em>vc_email='None'</em>, <em>vc_org='cbmpy.sourceforge.net'</em>, <em>add_cbmpy_annot=True</em>, <em>add_cobra_annot=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelToCOMBINEarchive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelToCOMBINEarchive" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a model in SBML and Excel format to a COMBINE archive using the following information:</p>
<ul class="simple">
<li><em>mod</em> a model object</li>
<li><em>fname</em> the output base filename, archive will be &lt;fname&gt;.zip</li>
<li><em>directory</em> [default=None] created the combine archive ‘directory’</li>
<li><em>sbmlname</em> [default=’None’] If <em>sbmlname</em> is defined then SBML file is &lt;sbmlname&gt;.xml otherwise sbml will be &lt;fname&gt;.xml.</li>
<li><em>withExcel</em> [default=True] include a human readable Excel spreadsheet version of the model</li>
<li><em>vc_given</em> [default=’CBMPy’] first name</li>
<li><em>vc_family</em> [default=’Software’] family name</li>
<li><em>vc_email</em> [default=’None’] email</li>
<li><em>vc_org</em> [default=’None’] organisation</li>
<li><em>add_cbmpy_annot</em> [default=True] add CBMPy KeyValueData annotation. Replaces &lt;notes&gt;</li>
<li><em>add_cobra_annot</em> [default=True] add COBRA &lt;notes&gt; annotation</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeModelToExcel97">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeModelToExcel97</code><span class="sig-paren">(</span><em>fba</em>, <em>filename</em>, <em>roundoff=6</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeModelToExcel97"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeModelToExcel97" title="Permalink to this definition">¶</a></dt>
<dd><p>Exports the model as an Excel 97 spreadsheet</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> a CBMPy model instance</li>
<li><em>filename</em> the filename of the workbook</li>
<li><em>roundoff</em> [default=6] the number of digits to round off to</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeOptimalSolution">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeOptimalSolution</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>Dir=None</em>, <em>separator='</em>, <em>'</em>, <em>only_exchange=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeOptimalSolution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeOptimalSolution" title="Permalink to this definition">¶</a></dt>
<dd><p>This function writes the optimal solution to file</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instance of an PySCeSCBM model</li>
<li><em>fname</em> the output filename</li>
<li><em>Dir</em> [default=None] use current directory if not None</li>
<li><em>separator</em> [default=’,’] the column separator</li>
<li><em>only_exchange</em> [default=False] only output fluxes labelled as exchange reactions</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeProteinCostToCSV">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeProteinCostToCSV</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeProteinCostToCSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeProteinCostToCSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the protein costs ‘CBM_PEPTIDE_COST’ annotation toa csv file.</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instantiated FBA object</li>
<li><em>fname</em> the exported file name</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeReactionInfoToFile">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeReactionInfoToFile</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>Dir=None</em>, <em>separator='</em>, <em>'</em>, <em>only_exchange=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeReactionInfoToFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeReactionInfoToFile" title="Permalink to this definition">¶</a></dt>
<dd><p>This function writes a CBModel to file</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instance of an PySCeSCBM model</li>
<li><em>fname</em> the output filename</li>
<li><em>Dir</em> [default=None] use directory if not None</li>
<li><em>separator</em> [default=’,’] the column separator</li>
<li><em>only_exchange</em> [default=False] only output fluxes labelled as exchange reactions</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeSBML2FBA">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeSBML2FBA</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>directory=None</em>, <em>sbml_level_version=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeSBML2FBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeSBML2FBA" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and writes it to file as SBML L2 with FBA annotations.
Note if you want to write BiGG/FAME style annotations then you must use <em>sbml_level_version=(2,1)</em></p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> the model will be written as XML to <em>fname</em></li>
<li><em>sbml_level_version</em> [default=None] a tuple containing the SBML level and version e.g. (2,1)</li>
</ul>
</div></blockquote>
<p>This is a utility wrapper for the function <cite>CBXML.sbml_writeSBML2FBA</cite></p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeSBML3FBC">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeSBML3FBC</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>directory=None</em>, <em>gpr_from_annot=False</em>, <em>add_groups=True</em>, <em>add_cbmpy_annot=True</em>, <em>add_cobra_annot=False</em>, <em>xoptions={'compress_bounds': True</em>, <em>'fbc_version': 1</em>, <em>'validate': False}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeSBML3FBC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeSBML3FBC" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and writes it to file as SBML L3 FBC:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> the model will be written as XML to <em>fname</em></li>
<li><em>directory</em> [default=None] if defined it is prepended to fname</li>
<li><em>gpr_from_annot</em> [default=True] if enabled will attempt to add the gene protein associations from the annotations
if no gene protein association objects exist</li>
<li><em>add_groups</em> [default=True] add SBML3 groups (if supported by libSBML)</li>
<li><em>add_cbmpy_annot</em> [default=True] add CBMPy KeyValueData annotation. Replaces &lt;notes&gt;</li>
<li><em>add_cobra_annot</em> [default=True] add COBRA &lt;notes&gt; annotation</li>
<li><em>xoptions</em> extended options<ul>
<li><em>fbc_version</em> [default=1] write SBML3FBC using version 1 (2013) or version 2 (2015)</li>
<li><em>validate</em> [default=False] validate the output SBML file</li>
<li><em>compress_bounds</em> [default=False] try compress output flux bound parameters</li>
<li><em>zip_model</em> [default=False] compress the model using PKZIP encoding</li>
<li><em>return_model_string</em> [default=False] return the SBML XML file as a string</li>
</ul>
</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeSBML3FBCV2">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeSBML3FBCV2</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>directory=None</em>, <em>gpr_from_annot=False</em>, <em>add_groups=True</em>, <em>add_cbmpy_annot=True</em>, <em>add_cobra_annot=False</em>, <em>validate=False</em>, <em>compress_bounds=True</em>, <em>zip_model=False</em>, <em>return_model_string=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeSBML3FBCV2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeSBML3FBCV2" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and writes it to file as SBML L3 FBCv2 :</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> the model will be written as XML to <em>fname</em></li>
<li><em>directory</em> [default=None] if defined it is prepended to fname</li>
<li><em>gpr_from_annot</em> [default=False] if enabled will attempt to add the gene protein associations from the annotations</li>
<li><em>add_groups</em> [default=True] add SBML3 groups (if supported by libSBML)</li>
<li><em>add_cbmpy_annot</em> [default=True] add CBMPy KeyValueData annotation. Replaces &lt;notes&gt;</li>
<li><em>add_cobra_annot</em> [default=False] add COBRA &lt;notes&gt; annotation</li>
<li><em>validate</em> [default=False] validate the output SBML file</li>
<li><em>compress_bounds</em> [default=True] try compress output flux bound parameters</li>
<li><em>zip_model</em> [default=False] compress the model using ZIP encoding</li>
<li><em>return_model_string</em> [default=False] return the SBML XML file as a string</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeSensitivitiesToCSV">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeSensitivitiesToCSV</code><span class="sig-paren">(</span><em>sensitivities</em>, <em>fname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeSensitivitiesToCSV"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeSensitivitiesToCSV" title="Permalink to this definition">¶</a></dt>
<dd><p>Write out a sensitivity report using the objective sensitivities and
bound sensitivity dictionaries created by e.g. cplx_getSensitivities().</p>
<blockquote>
<div><ul class="simple">
<li><em>sensitivity</em> tuple containing</li>
</ul>
<blockquote>
<div><ul class="simple">
<li><em>obj_sens</em> dictionary of objective coefficient sensitivities (per flux)</li>
<li><em>rhs_sens</em> dictionary of constraint rhs sensitivities (per constraint)</li>
<li><em>bound_sens</em> dictionary of bound sensitivities (per flux)</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><em>fname</em> output filename e.g. fname.csv</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeSolutions">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeSolutions</code><span class="sig-paren">(</span><em>fname</em>, <em>sols=[]</em>, <em>sep='</em>, <em>'</em>, <em>extra_output=None</em>, <em>fba=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeSolutions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeSolutions" title="Permalink to this definition">¶</a></dt>
<dd><p>Write 2 or more solutions where a solution is a dictionary of flux:value pairs:</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> the export filename</li>
<li><em>sols</em> a list of dictionaries containing flux:value pairs (e.g. output by cmod.getReactionValues())</li>
<li><em>sep</em> [default=’,’] the column separator</li>
<li><em>extra_output</em> [default=None] add detailed information to output e.g. reaction names by giving a CBModel object as an argument to <em>extra_output</em>.</li>
<li><em>fba</em> an fba model that canbe used for extra_output</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeSpeciesInfoToFile">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeSpeciesInfoToFile</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>Dir=None</em>, <em>separator='</em>, <em>'</em>, <em>met_type='all'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeSpeciesInfoToFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeSpeciesInfoToFile" title="Permalink to this definition">¶</a></dt>
<dd><p>This function writes a CBModel to file</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an instance of an PySCeSCBM model</li>
<li><em>fname</em> the output filename</li>
<li><em>Dir</em> [default=None] use directory if not None</li>
<li><em>separator</em> [default=’,’] the column separator</li>
<li><em>met_type</em> [default=’all’] only output certain type of species: ‘all’,’boundary’ or ‘variable’</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBWrite.writeStoichiometricMatrix">
<code class="descclassname">cbmpy.CBWrite.</code><code class="descname">writeStoichiometricMatrix</code><span class="sig-paren">(</span><em>fba</em>, <em>fname=None</em>, <em>work_dir=None</em>, <em>use_rational=False</em>, <em>fullLP=True</em>, <em>format='%s'</em>, <em>infinity_replace=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBWrite.html#writeStoichiometricMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBWrite.writeStoichiometricMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Write an FBA-LP in polynomial H-Format file. This is an improved version of <cite>WriteModelHFormatFBA()</cite>
which it replaces but is kept for backwards compatability.</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> a PySCeS-CBM FBA object</li>
<li><em>fname</em> [default=None] the output filename, fba.getId() if not defined</li>
<li><em>Work_dir</em> [default=None] the output directory</li>
<li><em>use_rational</em> [default=false] use rational numbers in output (requires sympy)</li>
<li><em>fullLP</em> [default=True] include the default objective function as a maximization target</li>
<li><em>format</em> [default=’%s’] the number format string</li>
<li><em>infinity_replace</em> [default=None] if defined this is the abs(value) of +-&lt;infinity&gt;</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy.CBWx"></span><div class="section" id="cbmpy-cbwx-module">
<h2>CBMPy: CBWx module<a class="headerlink" href="#cbmpy-cbwx-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBWx.py 575 2017-04-13 12:18:44Z bgoli $)</p>
</div>
<span class="target" id="module-cbmpy.CBXML"></span><div class="section" id="cbmpy-cbxml-module">
<h2>CBMPy: CBXML module<a class="headerlink" href="#cbmpy-cbxml-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: CBXML.py 648 2018-05-23 20:01:54Z bgoli $)</p>
<dl class="class">
<dt id="cbmpy.CBXML.MLStripper">
<em class="property">class </em><code class="descclassname">cbmpy.CBXML.</code><code class="descname">MLStripper</code><a class="reference internal" href="_modules/cbmpy/CBXML.html#MLStripper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.MLStripper" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for stripping a string of HTML/XML used from:
<a class="reference external" href="http://stackoverflow.com/questions/753052/strip-html-from-strings-in-python">http://stackoverflow.com/questions/753052/strip-html-from-strings-in-python</a></p>
</dd></dl>

<dl class="data">
<dt id="cbmpy.CBXML.SBML_NS">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">SBML_NS</code><em class="property"> = [('http://www.sbml.org/sbml/level3/version1/fbc/version2', 'L3V1FBC2'), ('http://www.sbml.org/sbml/level3/version1/fbc/version1', 'L3V1FBC1'), ('http://www.sbml.org/sbml/level3/version2/fbc/version2', 'L3V2FBC2'), ('http://www.sbml.org/sbml/level3/version2/fbc/version1', 'L3V2FBC1'), ('http://www.sbml.org/sbml/level3/version2/core', 'L3V2core'), ('http://www.sbml.org/sbml/level3/version1/core', 'L3V1core'), ('http://www.sbml.org/sbml/level2/version4', 'L2'), ('http://www.sbml.org/sbml/level2', 'L2')]</em><a class="headerlink" href="#cbmpy.CBXML.SBML_NS" title="Permalink to this definition">¶</a></dt>
<dd><p>print libsbml.BQB_ENCODES            , 8  # “encodes”,
print libsbml.BQB_HAS_PART           , 1  # “hasPart”,
print libsbml.BQB_HAS_PROPERTY       , 10 # “hasProperty”,
print libsbml.BQB_HAS_VERSION        , 4  # “hasVersion”,
print libsbml.BQB_IS                 , 0  # “isA”,
print libsbml.BQB_IS_DESCRIBED_BY    , 6  # “isDescribedBy”,
print libsbml.BQB_IS_ENCODED_BY      , 7  # “isEncodedBy”,
print libsbml.BQB_IS_HOMOLOG_TO      , 5  # “isHomologTo”,
print libsbml.BQB_IS_PART_OF         , 2  # “isPartOf”,
print libsbml.BQB_IS_PROPERTY_OF     , 11 # “isPropertyOf”,
print libsbml.BQB_IS_VERSION_OF      , 3  # “isVersionOf”,
print libsbml.BQB_OCCURS_IN          , 9  # “occursIn”,
print libsbml.BQB_UNKNOWN            , 12 # None</p>
<p>print libsbml.BQM_IS                 , 0 # None
print libsbml.BQM_IS_DERIVED_FROM    , 2 # None
print libsbml.BQM_IS_DESCRIBED_BY    , 1 # None
print libsbml.BQM_UNKNOWN            , 3 # None</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_convertCOBRASBMLtoFBC">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_convertCOBRASBMLtoFBC</code><span class="sig-paren">(</span><em>fname</em>, <em>outname=None</em>, <em>work_dir=None</em>, <em>output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_convertCOBRASBMLtoFBC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_convertCOBRASBMLtoFBC" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in a COBRA SBML Level 2 file and return the name of the created SBML Level 3 with FBC
file that is created in the output directory</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>outname</em> the name of the output file. If not specified then &lt;filename&gt;.l3fbc.xml is used as default</li>
<li><em>work_dir</em> [default=None] is the working directory</li>
<li><em>output_dir</em> [default=None] is the output directory (default is work_dir)</li>
</ul>
</div></blockquote>
<p>This method is based on code from libSBML (<a class="reference external" href="http://sbml.org">http://sbml.org</a>) in the file “convertCobra.py”
written by Frank T. Bergmann.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_convertSBML3FBCToCOBRA">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_convertSBML3FBCToCOBRA</code><span class="sig-paren">(</span><em>fname</em>, <em>outname=None</em>, <em>work_dir=None</em>, <em>output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_convertSBML3FBCToCOBRA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_convertSBML3FBCToCOBRA" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in a SBML Level 3 file and return the name of the created COBRA
file that is created in the output directory</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>outname</em> the name of the output file. If not specified then &lt;filename&gt;.cobra.xml is used as default</li>
<li><em>work_dir</em> [default=None] is the working directory</li>
<li><em>output_dir</em> [default=None] is the output directory (default is work_dir)</li>
</ul>
</div></blockquote>
<p>This method is based on code from libSBML (<a class="reference external" href="http://sbml.org">http://sbml.org</a>) in the file “convertFbcToCobra.py”
written by Frank T. Bergmann.</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_createAssociationFromAST">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_createAssociationFromAST</code><span class="sig-paren">(</span><em>node</em>, <em>out</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_createAssociationFromAST"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_createAssociationFromAST" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a GPR string ‘((g1 and g2) or g3)’ to an association via a Python AST.
In future I will get rid of all the string elements and work only with associations
and AST’s.</p>
<blockquote>
<div><ul class="simple">
<li><em>node</em> a Python AST note (e.g. body)</li>
<li><em>out</em> a new shiny FBC V2 GeneProductAssociation</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_createAssociationFromTreeV2">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_createAssociationFromTreeV2</code><span class="sig-paren">(</span><em>tree</em>, <em>out</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_createAssociationFromTreeV2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_createAssociationFromTreeV2" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a GPR tree to an association</p>
<blockquote>
<div><ul class="simple">
<li><em>tree</em> a GPR dict tree</li>
<li><em>out</em> a new shiny FBC V2 GeneProductAssociation</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_createModelL2">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_createModelL2</code><span class="sig-paren">(</span><em>fba</em>, <em>level=2</em>, <em>version=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_createModelL2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_createModelL2" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an SBML model and document:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> a PySCeSCBM model instance</li>
<li><em>level</em> always 2</li>
<li><em>version</em> always 1</li>
</ul>
</div></blockquote>
<p>and returns:</p>
<blockquote>
<div><ul class="simple">
<li><em>model</em> an SBML model</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_exportSBML2FBAModel">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_exportSBML2FBAModel</code><span class="sig-paren">(</span><em>document</em>, <em>filename</em>, <em>directory=None</em>, <em>return_doc=False</em>, <em>remove_note_body=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_exportSBML2FBAModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_exportSBML2FBAModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes an SBML model object to file. Note this is an internal SBML method use <cite>sbml_writeSBML2FBA()</cite> to write an FBA model:</p>
<blockquote>
<div><ul class="simple">
<li><em>model</em> a libSBML model instance</li>
<li><em>filename</em> the output filename</li>
<li><em>directory</em> [default=None] by default use filename otherwise join, &lt;dir&gt;&lt;filename&gt;</li>
<li><em>return_doc</em> [default=False] return the SBML document used to write the XML</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_fileFindVersion">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_fileFindVersion</code><span class="sig-paren">(</span><em>f</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_fileFindVersion"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_fileFindVersion" title="Permalink to this definition">¶</a></dt>
<dd><p>Try and find the SBML version and FBC support</p>
<blockquote>
<div><ul class="simple">
<li><em>f</em> the SBML file</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_fileValidate">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_fileValidate</code><span class="sig-paren">(</span><em>f</em>, <em>level='normal'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_fileValidate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_fileValidate" title="Permalink to this definition">¶</a></dt>
<dd><p>Validate an SBML file and model</p>
<blockquote>
<div><ul class="simple">
<li><em>f</em> the SBML file</li>
<li><em>level</em> [default=’normal’] the level of validation “normal” or “full”</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_getCVterms">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_getCVterms</code><span class="sig-paren">(</span><em>sb</em>, <em>model=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_getCVterms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_getCVterms" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the MIRIAM compliant CV terms and return a MIRIAMAnnotation or None</p>
<blockquote>
<div><ul class="simple">
<li><em>sb</em> a libSBML SBase derived object</li>
<li><em>model</em> is this a BQmodel term</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_getGPRasDictFBCv1">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_getGPRasDictFBCv1</code><span class="sig-paren">(</span><em>node</em>, <em>out</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_getGPRasDictFBCv1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_getGPRasDictFBCv1" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a GPR string ‘((g1 and g2) or g3)’ to a gprDict which is returned</p>
<blockquote>
<div><ul class="simple">
<li><em>node</em> a Python AST note (e.g. <cite>ast.parse(gprstring).body[0]</cite>)</li>
<li><em>out</em> a new dictionary that will be be created in place</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_getGPRasDictFBCv2">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_getGPRasDictFBCv2</code><span class="sig-paren">(</span><em>association</em>, <em>out</em>, <em>cntr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_getGPRasDictFBCv2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_getGPRasDictFBCv2" title="Permalink to this definition">¶</a></dt>
<dd><p>Walk through an SBML L3FBCV2 gene protein association and return a dictionary/tree representation</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_getGeneRefs">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_getGeneRefs</code><span class="sig-paren">(</span><em>association</em>, <em>out</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_getGeneRefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_getGeneRefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Walk through a gene association and extract GeneRefs inspired by Frank</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_getNotes">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_getNotes</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_getNotes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_getNotes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the SBML objects notes</p>
<ul class="simple">
<li><em>obj</em> an SBML object</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_readCOBRANote">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_readCOBRANote</code><span class="sig-paren">(</span><em>s</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_readCOBRANote"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_readCOBRANote" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses a COBRA style note from a XML string</p>
<blockquote>
<div><ul class="simple">
<li><em>s</em> an XML string</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_readCOBRASBML">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_readCOBRASBML</code><span class="sig-paren">(</span><em>fname</em>, <em>work_dir=None</em>, <em>return_sbml_model=False</em>, <em>delete_intermediate=False</em>, <em>fake_boundary_species_search=False</em>, <em>output_dir=None</em>, <em>speciesAnnotationFix=True</em>, <em>skip_genes=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_readCOBRASBML"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_readCOBRASBML" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in a COBRA format SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm_mod, sbml_mod) pair if return_sbml_model=True</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>work_dir</em> is the working directory</li>
<li><em>return_sbml_model</em> [default=False] return a a (cbm_mod, sbml_mod) pair</li>
<li><em>delete_intermediate</em> [default=False] delete the intermediate SBML Level 3 FBC file</li>
<li><em>fake_boundary_species_search</em> [default=False] after looking for the boundary_condition of a species search for overloaded id’s &lt;id&gt;_b</li>
<li><em>output_dir</em> [default=None] the directory to output the intermediate SBML L3 files (if generated) default to input directory</li>
<li><em>speciesAnnotationFix</em> [default=True]</li>
<li><em>skip_genes</em> [default=False] convert GPR associations</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_readKeyValueDataAnnotation">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_readKeyValueDataAnnotation</code><span class="sig-paren">(</span><em>annotations</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_readKeyValueDataAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_readKeyValueDataAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads KeyValueData annotation (<a class="reference external" href="http://pysces.sourceforge.net/KeyValueData">http://pysces.sourceforge.net/KeyValueData</a>) and returns a dictionary of key:value pairs</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_readSBML2FBA">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_readSBML2FBA</code><span class="sig-paren">(</span><em>fname</em>, <em>work_dir=None</em>, <em>return_sbml_model=False</em>, <em>fake_boundary_species_search=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_readSBML2FBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_readSBML2FBA" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in an SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm_mod, sbml_mod) pair if return_sbml_model=True</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>work_dir</em> is the working directory (only used if not None)</li>
<li><em>return_sbml_model</em> [default=False] return a a (cbm_mod, sbml_mod) pair</li>
<li><em>fake_boundary_species_search</em> [default=False] after looking for the boundary_condition of a species search for overloaded id’s &lt;id&gt;_b</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_readSBML3FBC">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_readSBML3FBC</code><span class="sig-paren">(</span><em>fname</em>, <em>work_dir=None</em>, <em>return_sbml_model=False</em>, <em>xoptions={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_readSBML3FBC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_readSBML3FBC" title="Permalink to this definition">¶</a></dt>
<dd><p>Read in an SBML Level 3 file with FBC annotation where and return either a CBM model object
or a (cbm_mod, sbml_mod) pair if return_sbml_model=True</p>
<blockquote>
<div><ul class="simple">
<li><em>fname</em> is the filename</li>
<li><em>work_dir</em> is the working directory</li>
<li><em>return_sbml_model</em> [default=False] return a a (cbm_mod, sbml_mod) pair</li>
<li><em>xoptions</em> special load options enable with option = True<ul>
<li><em>nogenes</em> do not load/process genes</li>
<li><em>noannot</em> do not load/process any annotations</li>
<li><em>validate</em> validate model and display errors and warnings before loading</li>
<li><em>readcobra</em> read the cobra annotation</li>
<li><em>read_model_string</em> [default=False] read the model from a string (instead of a filename) containing an SBML document</li>
<li><em>nmatrix_type</em> [default=’normal’] define the type of stoichiometrich matrix to be built<ul>
<li>‘numpy’ dense numpy array (best performance)</li>
<li>‘scipy_csr’ scipy sparse matrix (lower performance, low memory)</li>
<li>‘sympy’ a sympy rational matrix (low performance, high memory, cast to dense to analyse)</li>
<li>None do not build matrix</li>
</ul>
</li>
</ul>
</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setAnnotationsL3Fbc">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setAnnotationsL3Fbc</code><span class="sig-paren">(</span><em>cbmo</em>, <em>sbmlo</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setAnnotationsL3Fbc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setAnnotationsL3Fbc" title="Permalink to this definition">¶</a></dt>
<dd><p>Add CBMPy Fbase annotations to an SBML object, MIRIAM, SBO, Notes. Should
be called last when creating SBML objects.</p>
<blockquote>
<div><ul class="simple">
<li><em>cbmo</em> the CBMPy object</li>
<li><em>sbmlo</em> SBML object</li>
</ul>
<p>Note: this function should be used for new code, old code still needs to be
refactored.</p>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setCVterms">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setCVterms</code><span class="sig-paren">(</span><em>sb</em>, <em>uridict</em>, <em>model=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setCVterms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setCVterms" title="Permalink to this definition">¶</a></dt>
<dd><p>Add MIRIAM compliant CV terms to a sbml object from a CBM object</p>
<blockquote>
<div><ul class="simple">
<li><em>sb</em> a libSBML SBase derived object</li>
<li><em>uridict</em> a dictionary of uri’s as produced by getAllMIRIAMUris()</li>
<li><em>model</em> is this a BQmodel term [deprecated attribute, ignored and autodetected]</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setCompartmentsL3">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setCompartmentsL3</code><span class="sig-paren">(</span><em>model</em>, <em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setCompartmentsL3"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setCompartmentsL3" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the model compartments.</p>
<blockquote>
<div><ul class="simple">
<li><em>model</em> a libSBML model instance</li>
<li><em>fba</em> a PySCeSCBM model instance</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setDescription">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setDescription</code><span class="sig-paren">(</span><em>model</em>, <em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setDescription"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the model description as a &lt;note&gt; containing <cite>txt</cite> in an HTML paragraph on the model object.</p>
<blockquote>
<div><ul class="simple">
<li><em>model</em> a libSBML model instance</li>
<li><em>fba</em> a PySCeSCBM model instance</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setGroupsL3">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setGroupsL3</code><span class="sig-paren">(</span><em>cs</em>, <em>fba</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setGroupsL3"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setGroupsL3" title="Permalink to this definition">¶</a></dt>
<dd><p>add groups to the SBML model</p>
<blockquote>
<div><ul class="simple">
<li><em>cs</em> a CBMLtoSBML instance</li>
<li><em>fba</em> a CBMPy model instance</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setNotes3">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setNotes3</code><span class="sig-paren">(</span><em>obj</em>, <em>s</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setNotes3"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setNotes3" title="Permalink to this definition">¶</a></dt>
<dd><p>Formats the CBMPy notes as an SBML note and adds it to the SBMl object</p>
<blockquote>
<div><ul class="simple">
<li><em>obj</em> an SBML object</li>
<li><em>s</em> a string that should be added as a note</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setParametersL3Fbc">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setParametersL3Fbc</code><span class="sig-paren">(</span><em>fbcmod</em>, <em>add_cbmpy_anno=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setParametersL3Fbc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setParametersL3Fbc" title="Permalink to this definition">¶</a></dt>
<dd><p>Add non fluxbound related parameters to the model</p>
<blockquote>
<div><ul class="simple">
<li><em>fbcmod</em> a CBM2SBML instance</li>
<li><em>add_cbmpy_anno</em> [default=True] add CBMPy KeyValueData annotation.</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setReactionsL2">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setReactionsL2</code><span class="sig-paren">(</span><em>model</em>, <em>fba</em>, <em>return_dict=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setReactionsL2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setReactionsL2" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the FBA instance reactions to the SBML model</p>
<blockquote>
<div><ul class="simple">
<li><em>model</em> an SBML model instance</li>
<li><em>fba</em> a PySCeSCBM model instance</li>
<li><em>return_dict</em> [default=False] if True do not add reactions to SBML document instead return a dictionary description of the reactions</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setReactionsL3Fbc">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setReactionsL3Fbc</code><span class="sig-paren">(</span><em>fbcmod</em>, <em>return_dict=False</em>, <em>add_cobra_anno=False</em>, <em>add_cbmpy_anno=True</em>, <em>fbc_version=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setReactionsL3Fbc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setReactionsL3Fbc" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the FBA instance reactions to the SBML model</p>
<blockquote>
<div><ul class="simple">
<li><em>fbcmod</em> a CBM2SBML instance</li>
<li><em>return_dict</em> [default=False] if True do not add reactions to SBML document instead return a dictionary description of the reactions</li>
<li><em>add_cbmpy_anno</em> [default=True] add CBMPy KeyValueData annotation. Replaces &lt;notes&gt;</li>
<li><em>add_cobra_anno</em> [default=False] add COBRA &lt;notes&gt; annotation</li>
<li><em>fbc_version</em> [default=1] writes either FBC v1 (2013) or v2 (2015)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setSpeciesL2">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setSpeciesL2</code><span class="sig-paren">(</span><em>model</em>, <em>fba</em>, <em>return_dicts=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setSpeciesL2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setSpeciesL2" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the species definitions to the SBML object:</p>
<blockquote>
<div><ul class="simple">
<li><em>model</em> [default=’‘] a libSBML model instance or can be None if <em>return_dicts</em> == True</li>
<li><em>fba</em> a PySCeSCBM model instance</li>
<li><em>return_dicts</em> [default=False] only returns the compartment and species dictionaries without updated the SBML</li>
</ul>
</div></blockquote>
<p>returns:</p>
<blockquote>
<div><ul class="simple">
<li><em>compartments</em> a dictionary of compartments (except when give <em>return_dicts</em> argument)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setSpeciesL3">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setSpeciesL3</code><span class="sig-paren">(</span><em>model</em>, <em>fba</em>, <em>return_dicts=False</em>, <em>add_cobra_anno=False</em>, <em>add_cbmpy_anno=True</em>, <em>substance_units=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setSpeciesL3"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setSpeciesL3" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the species definitions to the SBML object:</p>
<blockquote>
<div><ul class="simple">
<li><em>model</em> and SBML model instance or can be None if <em>return_dicts</em> == True</li>
<li><em>fba</em> a PySCeSCBM model instance</li>
<li><em>return_dicts</em> [default=False] only returns the compartment and species dictionaries without updating the SBML</li>
<li><em>add_cbmpy_anno</em> [default=True] add CBMPy KeyValueData annotation. Replaces &lt;notes&gt;</li>
<li><em>add_cobra_anno</em> [default=False] add COBRA &lt;notes&gt; annotation</li>
<li><em>substance_units</em> [default=True] defines the species in amounts rather than concentrations (necessary for default mmol/gdw.h)</li>
</ul>
</div></blockquote>
<p>returns:</p>
<blockquote>
<div><ul class="simple">
<li><em>compartments</em> a dictionary of compartments (except when given <em>return_dicts</em> argument)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setUnits">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setUnits</code><span class="sig-paren">(</span><em>model</em>, <em>units=None</em>, <em>give_default=False</em>, <em>L3=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setUnits"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setUnits" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds units to the model:</p>
<ul class="simple">
<li><em>model</em> a libSBML model instance</li>
<li><em>units</em> [default=None] a dictionary of units, if None default units are used</li>
<li><em>give_default</em> [default=False] if true method returns the default unit dictionary</li>
<li><em>L3</em> [default=True] use the L3 defaults</li>
</ul>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_setValidationOptions">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_setValidationOptions</code><span class="sig-paren">(</span><em>D</em>, <em>level</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_setValidationOptions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_setValidationOptions" title="Permalink to this definition">¶</a></dt>
<dd><p>set the validation level of an SBML document</p>
<blockquote>
<div><ul class="simple">
<li><em>D</em> an SBML document</li>
<li><em>level</em> the level of consistency check can be either one of:</li>
</ul>
<blockquote>
<div><ul class="simple">
<li>‘normal’ basic id checking only</li>
<li>‘full’ all checks enabled</li>
<li>None disable all validation</li>
</ul>
</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_validateDocument">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_validateDocument</code><span class="sig-paren">(</span><em>D</em>, <em>fullmsg=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_validateDocument"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_validateDocument" title="Permalink to this definition">¶</a></dt>
<dd><p>Validates and SBML document returns three dictionaries, errors, warnings, other and a boolean indicating an invalid document:</p>
<blockquote>
<div><ul class="simple">
<li><em>D</em> and SBML document</li>
<li><em>fullmsg</em> [default=False] optionally display the full error message</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_writeAnnotationsAsCOBRANote">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_writeAnnotationsAsCOBRANote</code><span class="sig-paren">(</span><em>annotations</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_writeAnnotationsAsCOBRANote"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_writeAnnotationsAsCOBRANote" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the annotations dictionary as a COBRA compatible SBML &lt;note&gt;</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_writeCOBRASBML">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_writeCOBRASBML</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>directory=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_writeCOBRASBML"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_writeCOBRASBML" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and writes it to file as a COBRA compatible :</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> the model will be written as XML to <em>fname</em></li>
<li><em>directory</em> [default=None] if defined it is prepended to fname</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_writeKeyValueDataAnnotation">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_writeKeyValueDataAnnotation</code><span class="sig-paren">(</span><em>annotations</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_writeKeyValueDataAnnotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_writeKeyValueDataAnnotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the key:value annotations as a KeyValueData annotation (http://pysces.sourceforge.net/KeyValueData)</p>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_writeSBML2FBA">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_writeSBML2FBA</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>directory=None</em>, <em>sbml_level_version=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_writeSBML2FBA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_writeSBML2FBA" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and writes it to file as SBML L3 FBA:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> the model will be written as XML to <em>fname</em></li>
<li><em>directory</em> [default=None] if defined it is prepended to fname</li>
<li><em>sbml_level_version</em> [default=None] a tuple containing the SBML level and version e.g. (2,4) (ignored)</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.sbml_writeSBML3FBC">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">sbml_writeSBML3FBC</code><span class="sig-paren">(</span><em>fba</em>, <em>fname</em>, <em>directory=None</em>, <em>sbml_level_version=(3</em>, <em>1)</em>, <em>autofix=True</em>, <em>return_fbc=False</em>, <em>gpr_from_annot=False</em>, <em>add_groups=False</em>, <em>add_cbmpy_annot=True</em>, <em>add_cobra_annot=False</em>, <em>xoptions={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#sbml_writeSBML3FBC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.sbml_writeSBML3FBC" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and writes it to file as SBML L3 FBC:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> the model will be written as XML to <em>fname</em></li>
<li><em>directory</em> [default=None] if defined it is prepended to fname</li>
<li><em>sbml_level_version</em> [default=(3,1)] a tuple containing the SBML level and version e.g. (3,1)</li>
<li><em>autofix</em> convert &lt;&gt; to &lt;=&gt;=</li>
<li><em>return_fbc</em> return the FBC converter instance</li>
<li><em>gpr_from_annot</em> [default=False] if enabled will attempt to add the gene protein associations from the annotations
if no gene protein association objects exist</li>
<li><em>add_cbmpy_annot</em> [default=True] add CBMPy KeyValueData annotation. Replaces &lt;notes&gt;</li>
<li><em>add_cobra_annot</em> [default=True] add COBRA &lt;notes&gt; annotation</li>
<li><em>xoptions</em> extended options<ul>
<li><em>fbc_version</em> [default=1] write SBML3FBC using version 1 (2013) or version 2 (2015)</li>
<li><em>validate</em> [default=False] validate the output SBML file</li>
<li><em>compress_bounds</em> [default=False] try compress output flux bound parameters</li>
<li><em>zip_model</em> [default=False] compress the model using ZIP encoding</li>
<li><em>return_model_string</em> [default=False] return the SBML XML file as a string</li>
</ul>
</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.setCBSBOterm">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">setCBSBOterm</code><span class="sig-paren">(</span><em>sbo</em>, <em>obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#setCBSBOterm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.setCBSBOterm" title="Permalink to this definition">¶</a></dt>
<dd><p>Given an SBOterm from libSBML, add it to a CBMPy object</p>
<blockquote>
<div><ul class="simple">
<li><em>sbo</em> the sbo term string</li>
<li><em>obj</em> the CBMPy Fbase derived object</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_addSBML2FBAFluxBound">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_addSBML2FBAFluxBound</code><span class="sig-paren">(</span><em>document</em>, <em>rid</em>, <em>operator</em>, <em>value</em>, <em>fbid=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_addSBML2FBAFluxBound"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_addSBML2FBAFluxBound" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an SBML3FBA flux bound to the document:</p>
<blockquote>
<div><ul class="simple">
<li><em>document</em> a minidom XML document created by xml_createSBML2FBADoc</li>
<li><em>rid</em> the reaction id</li>
<li><em>operator</em> one of [‘greater’,’greaterEqual’,’less’,’lessEqual’,’equal’,’&gt;’,’&gt;=’,’&lt;’,’&lt;=’,’=’]</li>
<li><em>value</em> a float which will be cast to a string using str(value)</li>
<li><em>fbid</em> the flux bound id, autogenerated by default</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_addSBML2FBAObjective">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_addSBML2FBAObjective</code><span class="sig-paren">(</span><em>document</em>, <em>objective</em>, <em>active=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_addSBML2FBAObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_addSBML2FBAObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an objective element to the documents listOfObjectives and sets the active attribute:</p>
<blockquote>
<div><ul class="simple">
<li><em>document</em> a minidom XML document created by <cite>xml_createSBML2FBADoc</cite></li>
<li><em>objective</em> a minidom XML objective element created with <cite>xml_createSBML2FBAObjective</cite></li>
<li><em>active</em> [default=True] a boolean flag specifiying whether this objective is active</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_createListOfFluxObjectives">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_createListOfFluxObjectives</code><span class="sig-paren">(</span><em>document</em>, <em>fluxObjectives</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_createListOfFluxObjectives"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_createListOfFluxObjectives" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a list of fluxObjectives to add to an Objective:</p>
<blockquote>
<div><ul class="simple">
<li><em>document</em> a minidom XML document created by xml_createSBML2FBADoc</li>
<li><em>fluxobjs</em> a list of (rid, coefficient) tuples</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_createSBML2FBADoc">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_createSBML2FBADoc</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_createSBML2FBADoc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_createSBML2FBADoc" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a ‘document’ to store the SBML2FBA annotation, returns:</p>
<blockquote>
<div><ul class="simple">
<li><em>DOC</em> a minidom document</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_createSBML2FBAObjective">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_createSBML2FBAObjective</code><span class="sig-paren">(</span><em>document</em>, <em>oid</em>, <em>sense</em>, <em>fluxObjectives</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_createSBML2FBAObjective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_createSBML2FBAObjective" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a list of fluxObjectives to add to an Objective:</p>
<blockquote>
<div><ul class="simple">
<li><em>document</em> a minidom XML document created by xml_createSBML2FBADoc</li>
<li><em>oid</em> the objective id</li>
<li><em>sense</em> a string containing the objective sense either: <strong>maximize</strong> or <strong>minimize</strong></li>
<li><em>fluxObjectives</em> a list of (rid, coefficient) tuples</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_getSBML2FBAannotation">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_getSBML2FBAannotation</code><span class="sig-paren">(</span><em>fba</em>, <em>fname=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_getSBML2FBAannotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_getSBML2FBAannotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes an FBA model object and returns the SBML3FBA annotation as an XML string:</p>
<blockquote>
<div><ul class="simple">
<li><em>fba</em> an fba model object</li>
<li><em>fname</em> [default=None] if supplied the XML will be written to file <em>fname</em></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_stripTags">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_stripTags</code><span class="sig-paren">(</span><em>html</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_stripTags"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_stripTags" title="Permalink to this definition">¶</a></dt>
<dd><p>Strip a string of HTML/XML, returns a string</p>
<blockquote>
<div><ul class="simple">
<li><em>html</em> the string containing html</li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cbmpy.CBXML.xml_viewSBML2FBAXML">
<code class="descclassname">cbmpy.CBXML.</code><code class="descname">xml_viewSBML2FBAXML</code><span class="sig-paren">(</span><em>document</em>, <em>fname=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cbmpy/CBXML.html#xml_viewSBML2FBAXML"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cbmpy.CBXML.xml_viewSBML2FBAXML" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a minidom XML document to screen or file, arguments:</p>
<blockquote>
<div><ul class="simple">
<li><em>document</em> a minidom XML document</li>
<li><em>fname</em> [default=None] by default print to screen or write to file fname</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
<span class="target" id="module-cbmpy._multicorefva"></span><div class="section" id="cbmpy-multicorefva-module">
<h2>CBMPy: MultiCoreFVA module<a class="headerlink" href="#cbmpy-multicorefva-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: _multicorefva.py 575 2017-04-13 12:18:44Z bgoli $)</p>
</div>
<span class="target" id="module-cbmpy._multicoreenvfva"></span><div class="section" id="cbmpy-multicoreenvfva-module">
<h2>CBMPy: MultiCoreEnvFVA module<a class="headerlink" href="#cbmpy-multicoreenvfva-module" title="Permalink to this headline">¶</a></h2>
<p>PySCeS Constraint Based Modelling (<a class="reference external" href="http://cbmpy.sourceforge.net">http://cbmpy.sourceforge.net</a>)
Copyright (C) 2009-2017 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands</p>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>
<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>
<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;<a class="reference external" href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>&gt;</p>
<p>Author: Brett G. Olivier
Contact email: <a class="reference external" href="mailto:bgoli&#37;&#52;&#48;users&#46;sourceforge&#46;net">bgoli<span>&#64;</span>users<span>&#46;</span>sourceforge<span>&#46;</span>net</a>
Last edit: $Author: bgoli $ ($Id: _multicoreenvfva.py 575 2017-04-13 12:18:44Z bgoli $)</p>
</div>
<span class="target" id="module-cbmpy.miriamids"></span></div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="cbmpy.html">
              <img class="logo" src="_static/pysces_cbm1_head.jpg" alt="Logo"/>
            </a></p>
  <h3><a href="cbmpy.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">CBMPy Module Reference</a><ul>
<li><a class="reference internal" href="#cbmpy-cbcommon-module">CBMPy: CBCommon module</a></li>
<li><a class="reference internal" href="#cbmpy-cbconfig-module">CBMPy: CBConfig module</a></li>
<li><a class="reference internal" href="#cbmpy-cbcplex-module">CBMPy: CBCPLEX module</a></li>
<li><a class="reference internal" href="#cbmpy-cbdatastruct-module">CBMPy: CBDataStruct module</a></li>
<li><a class="reference internal" href="#cbmpy-cbglpk-module">CBMPy: CBGLPK module</a></li>
<li><a class="reference internal" href="#cbmpy-cbgui-module">CBMPy: CBGUI module</a></li>
<li><a class="reference internal" href="#cbmpy-cbmodel-module">CBMPy: CBModel module</a></li>
<li><a class="reference internal" href="#cbmpy-cbmodeltools-module">CBMPy: CBModelTools module</a></li>
<li><a class="reference internal" href="#cbmpy-cbmulticore-module">CBMPy: CBMultiCore module</a></li>
<li><a class="reference internal" href="#cbmpy-cbmultienv-module">CBMPy: CBMultiEnv module</a></li>
<li><a class="reference internal" href="#cbmpy-cbnetdb-module">CBMPy: CBNetDB module</a></li>
<li><a class="reference internal" href="#cbmpy-cbplot-module">CBMPy: CBPlot module</a></li>
<li><a class="reference internal" href="#cbmpy-cbread-module">CBMPy: CBRead module</a></li>
<li><a class="reference internal" href="#cbmpy-cbreadtxt-module">CBMPy: CBReadtxt module</a></li>
<li><a class="reference internal" href="#cbmpy-cbsolver-module">CBMPy: CBSolver module</a></li>
<li><a class="reference internal" href="#cbmpy-cbtools-module">CBMPy: CBTools module</a></li>
<li><a class="reference internal" href="#cbmpy-cbwrite-module">CBMPy: CBWrite module</a></li>
<li><a class="reference internal" href="#cbmpy-cbwx-module">CBMPy: CBWx module</a></li>
<li><a class="reference internal" href="#cbmpy-cbxml-module">CBMPy: CBXML module</a></li>
<li><a class="reference internal" href="#cbmpy-multicorefva-module">CBMPy: MultiCoreFVA module</a></li>
<li><a class="reference internal" href="#cbmpy-multicoreenvfva-module">CBMPy: MultiCoreEnvFVA module</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="cbmpy.html">Documentation overview</a><ul>
      <li>Previous: <a href="manual_doc.html" title="previous chapter">Introduction</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/modules_doc.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2010-2018, Brett G. Olivier.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/modules_doc.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>